--- ocaml_504/parsing/depend.ml	2025-11-07 15:50:07
+++ ocaml_505/parsing/depend.ml	2026-02-03 16:48:21
@@ -119,10 +119,13 @@
     let bv = open_module bv mod_ident.txt in
     add_type bv t
   | Ptyp_extension e -> handle_extension e
+  | Ptyp_functor (_, id, pt, t2) ->
+    add_package_type bv pt;
+    add_type (String.Map.add id.txt bound bv) t2
 
 and add_package_type bv ptyp =
   add bv ptyp.ppt_path;
-  List.iter (fun (_, ty) -> add_type bv ty) ptyp.ppt_cstrs
+  List.iter (fun (_, ty) -> add_type bv ty) ptyp.ppt_constraints
 
 let add_opt add_fn bv = function
     None -> ()
@@ -139,15 +142,17 @@
 let add_type_declaration bv td =
   List.iter
     (fun (ty1, ty2, _) -> add_type bv ty1; add_type bv ty2)
-    td.ptype_cstrs;
+    td.ptype_constraints;
   add_opt add_type bv td.ptype_manifest;
   let add_tkind = function
-    Ptype_abstract -> ()
-  | Ptype_variant cstrs ->
-      List.iter (add_constructor_decl bv) cstrs
-  | Ptype_record lbls ->
-      List.iter (fun pld -> add_type bv pld.pld_type) lbls
-  | Ptype_open -> () in
+    | Ptype_abstract -> ()
+    | Ptype_variant cstrs ->
+        List.iter (add_constructor_decl bv) cstrs
+    | Ptype_record lbls ->
+        List.iter (fun pld -> add_type bv pld.pld_type) lbls
+    | Ptype_open -> ()
+    | Ptype_external _ -> ()
+  in
   add_tkind td.ptype_kind
 
 let add_extension_constructor bv ext =
@@ -187,7 +192,8 @@
   | Ppat_variant(_, op) -> add_opt add_pattern bv op
   | Ppat_type li -> add bv li
   | Ppat_lazy p -> add_pattern bv p
-  | Ppat_unpack id ->
+  | Ppat_unpack (id, ptyp) ->
+      add_opt add_package_type bv ptyp;
       Option.iter
         (fun name -> pattern_bv := String.Map.add name bound !pattern_bv) id.txt
   | Ppat_open ( m, p) -> let bv = open_module bv m.txt in add_pattern bv p
@@ -240,15 +246,6 @@
   | Pexp_new li -> add bv li
   | Pexp_setinstvar(_v, e) -> add_expr bv e
   | Pexp_override sel -> List.iter (fun (_s, e) -> add_expr bv e) sel
-  | Pexp_letmodule(id, m, e) ->
-      let b = add_module_binding bv m in
-      let bv =
-        match id.txt with
-        | None -> bv
-        | Some id -> String.Map.add id b bv
-      in
-      add_expr bv e
-  | Pexp_letexception(_, e) -> add_expr bv e
   | Pexp_assert (e) -> add_expr bv e
   | Pexp_lazy (e) -> add_expr bv e
   | Pexp_poly (e, t) -> add_expr bv e; add_opt add_type bv t
@@ -257,9 +254,6 @@
   | Pexp_newtype (_, e) -> add_expr bv e
   | Pexp_pack (m, opty) ->
       add_module_expr bv m; add_opt add_package_type bv opty
-  | Pexp_open (o, e) ->
-      let bv = open_declaration bv o in
-      add_expr bv e
   | Pexp_letop {let_; ands; body} ->
       let bv' = add_binding_op bv bv let_ in
       let bv' = List.fold_left (add_binding_op bv) bv' ands in
@@ -273,6 +267,9 @@
       end
   | Pexp_extension e -> handle_extension e
   | Pexp_unreachable -> ()
+  | Pexp_struct_item (si, e) ->
+      let bv, _ = add_struct_item (bv, String.Map.empty) si in
+      add_expr bv e
 
 and add_function_param bv param =
   match param.pparam_desc with
