--- ocaml_504/parsing/location.mli	2025-11-07 15:50:07
+++ ocaml_505/parsing/location.mli	2026-02-03 16:48:21
@@ -94,6 +94,7 @@
 
 val mknoloc : 'a -> 'a loc
 val mkloc : 'a -> t -> 'a loc
+val map : ('a -> 'b) -> 'a loc -> 'b loc
 
 
 (** {1 Input info} *)
@@ -254,7 +255,7 @@
     Format.formatter -> Format_doc.t -> unit;
 }
 (** A printer for [report]s, defined using open-recursion.
-    The goal is to make it easy to define new printers by re-using code from
+    The goal is to make it easy to define new printers by reusing code from
     existing ones.
 *)
 
