--- ocaml_505/typing/env.ml	2026-07-07 01:48:43.513022109 +0200
+++ ocaml_506_typing_recovery/typing/env.ml	2026-07-07 02:21:16.938995335 +0200
@@ -672,12 +672,23 @@
   | Illegal_value_name of Location.t * string
   | Lookup_error of Location.t * t * lookup_error
 
-exception Error of error
+module Error : sig
+  type exn += private In_context of error
 
-let error err = raise (Error err)
+  val log_or_raise : error -> unit
+  val log_and_raise : error -> 'a
+end = struct
+  type exn += In_context of error
+
+  let log_and_raise err =
+    Typing_recovery.log_and_raise (In_context err)
+
+  let log_or_raise err =
+    Typing_recovery.log_or_raise (In_context err)
+end
 
 let lookup_error loc env err =
-  error (Lookup_error(loc, env, err))
+  Error.log_and_raise (Lookup_error(loc, env, err))
 
 let same_type_declarations e1 e2 =
   e1.types == e2.types &&
@@ -906,11 +917,24 @@
     }
   }
 
+(* Maintain the same physical name for .cmi names even across type-checker
+   resets. In particular, this means that lazy thunks (e.g. in Matching) do not
+   need to be reset as persistent names in identifiers in those lambda blocks
+   will be physically identical across type-checker resets.
+   See also tests in testsuite/tests/tool-ocamlc-determinism *)
+let hashcons_name =
+  let names = String.Tbl.create 1023 in
+  fun name ->
+    try String.Tbl.find names name
+    with Not_found ->
+      String.Tbl.add names name name;
+      name
+
 let sign_of_cmi ~freshen { Persistent_env.Persistent_signature.cmi; _ } =
   let name = cmi.cmi_name in
   let sign = cmi.cmi_sign in
   let flags = cmi.cmi_flags in
-  let id = Ident.create_persistent name in
+  let id = Ident.create_persistent (hashcons_name name) in
   let path = Pident id in
   let alerts =
     List.fold_left (fun acc -> function Alerts s -> s | _ -> acc)
@@ -1385,10 +1409,11 @@
 let normalize_module_path oloc env path =
   try normalize_module_path ~lax:(oloc = None) env path
   with Not_found ->
-    match oloc with None -> assert false
-    | Some loc ->
-        error (Missing_module(loc, path,
-                              normalize_module_path ~lax:true env path))
+  match oloc with
+  | None -> assert false
+  | Some loc ->
+      Error.log_and_raise
+        (Missing_module(loc, path, normalize_module_path ~lax:true env path))
 
 let rec normalize_path_prefix oloc env path =
   match path with
@@ -1773,194 +1798,9 @@
   | Mp_present ->
       Lazy_backtrack.create_forced (Aident id)
 
-let rec components_of_module_maker
-          {cm_env; cm_prefixing_subst;
-           cm_path; cm_addr; cm_mty; cm_shape} : _ result =
-  match scrape_alias cm_env cm_mty with
-    MtyL_signature sg ->
-      let c =
-        { comp_values = NameMap.empty;
-          comp_constrs = NameMap.empty;
-          comp_labels = NameMap.empty; comp_types = NameMap.empty;
-          comp_modules = NameMap.empty; comp_modtypes = NameMap.empty;
-          comp_classes = NameMap.empty; comp_cltypes = NameMap.empty }
-      in
-      let items_and_paths, sub =
-        prefix_idents cm_path cm_prefixing_subst sg
-      in
-      let env = ref cm_env in
-      let pos = ref 0 in
-      let next_address () =
-        let addr : address_unforced =
-          Projection { parent = cm_addr; pos = !pos }
-        in
-        incr pos;
-        Lazy_backtrack.create addr
-      in
-      List.iter (fun ((item : Subst.Lazy.signature_item), path) ->
-        match item with
-          SigL_value(id, decl, _) ->
-            let decl' = Subst.value_description sub decl in
-            let addr =
-              match decl.val_kind with
-              | Val_prim _ -> Lazy_backtrack.create_failed Not_found
-              | _ -> next_address ()
-            in
-            let vda_shape = Shape.proj cm_shape (Shape.Item.value id) in
-            let vda =
-              { vda_description = decl'; vda_address = addr; vda_shape }
-            in
-            c.comp_values <- NameMap.add (Ident.name id) vda c.comp_values;
-        | SigL_type(id, decl, _, _) ->
-            let final_decl = Subst.type_declaration sub decl in
-            Btype.set_static_row_name final_decl
-              (Subst.type_path sub (Path.Pident id));
-            let descrs =
-              match decl.type_kind with
-              | Type_variant (_,repr) ->
-                  let cstrs = List.map snd
-                    (Datarepr.constructors_of_type path final_decl
-                        ~current_unit:(get_current_unit ()))
-                  in
-                  List.iter
-                    (fun descr ->
-                      let cda_shape = Shape.leaf descr.cstr_uid in
-                      let cda = {
-                        cda_description = descr;
-                        cda_address = None;
-                        cda_shape }
-                      in
-                      c.comp_constrs <-
-                        add_to_tbl descr.cstr_name cda c.comp_constrs
-                    ) cstrs;
-                 Type_variant (cstrs, repr)
-              | Type_record (_, repr) ->
-                  let lbls = List.map snd
-                    (Datarepr.labels_of_type path final_decl)
-                  in
-                  List.iter
-                    (fun descr ->
-                      c.comp_labels <-
-                        add_to_tbl descr.lbl_name descr c.comp_labels)
-                    lbls;
-                  Type_record (lbls, repr)
-              | Type_abstract r -> Type_abstract r
-              | Type_open -> Type_open
-              | Type_external name -> Type_external name
-            in
-            let shape = Shape.proj cm_shape (Shape.Item.type_ id) in
-            let tda =
-              { tda_declaration = final_decl;
-                tda_descriptions = descrs;
-                tda_shape = shape; }
-            in
-            c.comp_types <- NameMap.add (Ident.name id) tda c.comp_types;
-            env := store_type_infos ~tda_shape:shape id decl !env
-        | SigL_typext(id, ext, _, _) ->
-            let ext' = Subst.extension_constructor sub ext in
-            let descr =
-              Datarepr.extension_descr ~current_unit:(get_current_unit ()) path
-                ext'
-            in
-            let addr = next_address () in
-            let cda_shape =
-              Shape.proj cm_shape (Shape.Item.extension_constructor id)
-            in
-            let cda =
-              { cda_description = descr; cda_address = Some addr; cda_shape }
-            in
-            c.comp_constrs <- add_to_tbl (Ident.name id) cda c.comp_constrs
-        | SigL_module(id, pres, md, _, _) ->
-            let md' =
-              (* The prefixed items get the same scope as [cm_path], which is
-                 the prefix. *)
-              Subst.Lazy.module_decl
-                (Subst.Rescope (Path.scope cm_path)) sub md
-            in
-            let addr =
-              match pres with
-              | Mp_absent -> begin
-                  match md.mdl_type with
-                  | MtyL_alias path ->
-                      Lazy_backtrack.create (ModAlias {env = !env; path})
-                  | _ -> assert false
-                end
-              | Mp_present -> next_address ()
-            in
-            let alerts =
-              Builtin_attributes.alerts_of_attrs md.mdl_attributes
-            in
-            let shape = Shape.proj cm_shape (Shape.Item.module_ id) in
-            let comps =
-              components_of_module ~alerts ~uid:md.mdl_uid !env
-                sub path addr md.mdl_type shape
-            in
-            let mda =
-              { mda_declaration = md';
-                mda_components = comps;
-                mda_address = addr;
-                mda_shape = shape; }
-            in
-            c.comp_modules <-
-              NameMap.add (Ident.name id) mda c.comp_modules;
-            env :=
-              store_module ~update_summary:false ~check:None
-                id addr pres md shape !env
-        | SigL_modtype(id, decl, _) ->
-            let final_decl =
-              (* The prefixed items get the same scope as [cm_path], which is
-                 the prefix. *)
-              Subst.Lazy.modtype_decl (Rescope (Path.scope cm_path))
-                sub decl
-            in
-            let shape = Shape.proj cm_shape (Shape.Item.module_type id) in
-            let mtda =
-              { mtda_declaration = final_decl;
-                mtda_shape = shape; }
-            in
-            c.comp_modtypes <-
-              NameMap.add (Ident.name id) mtda c.comp_modtypes;
-            env := store_modtype ~update_summary:false id decl shape !env
-        | SigL_class(id, decl, _, _) ->
-            let decl' = Subst.class_declaration sub decl in
-            let addr = next_address () in
-            let shape = Shape.proj cm_shape (Shape.Item.class_ id) in
-            let clda =
-              { clda_declaration = decl';
-                clda_address = addr;
-                clda_shape = shape; }
-            in
-            c.comp_classes <- NameMap.add (Ident.name id) clda c.comp_classes
-        | SigL_class_type(id, decl, _, _) ->
-            let decl' = Subst.cltype_declaration sub decl in
-            let shape = Shape.proj cm_shape (Shape.Item.class_type id) in
-            let cltda = { cltda_declaration = decl'; cltda_shape = shape } in
-            c.comp_cltypes <-
-              NameMap.add (Ident.name id) cltda c.comp_cltypes)
-        items_and_paths;
-        Ok (Structure_comps c)
-  | MtyL_functor(arg, ty_res) ->
-      let sub = cm_prefixing_subst in
-      let scoping = Subst.Rescope (Path.scope cm_path) in
-      let open Subst.Lazy in
-        Ok (Functor_comps {
-          (* fcomp_arg and fcomp_res must be prefixed eagerly, because
-             they are interpreted in the outer environment *)
-          fcomp_arg =
-            (match arg with
-            | Unit -> Unit
-            | Named (param, ty_arg) ->
-              Named (param, force_modtype (modtype scoping sub ty_arg)));
-          fcomp_res = force_modtype (modtype scoping sub ty_res);
-          fcomp_shape = cm_shape;
-          fcomp_cache = Hashtbl.create 17;
-          fcomp_subst_cache = Hashtbl.create 17 })
-  | MtyL_ident _ -> Error No_components_abstract
-  | MtyL_alias p -> Error (No_components_alias p)
-
 (* Insertion of bindings by identifier + path *)
 
-and check_usage loc id uid warn tbl =
+let check_usage loc id uid warn tbl =
   if not loc.Location.loc_ghost &&
      Uid.for_actual_declaration uid &&
      Warnings.is_active (warn "")
@@ -1973,9 +1813,9 @@
     then
       !add_delayed_check_forward
         (fun () -> if not !used then Location.prerr_warning loc (warn name))
-  end;
+  end
 
-and check_value_name name loc =
+let check_value_name name loc =
   (* Note: we could also check here general validity of the
      identifier, to protect against bad identifiers forged by -pp or
      -ppx preprocessors. *)
@@ -1983,10 +1823,10 @@
        (Utf8_lexeme.starts_like_a_valid_identifier name) then
     for i = 1 to String.length name - 1 do
       if name.[i] = '#' then
-        error (Illegal_value_name(loc, name))
+        Error.log_or_raise (Illegal_value_name(loc, name))
     done
 
-and store_value ?check id addr decl shape env =
+let store_value ?check id addr decl shape env =
   check_value_name (Ident.name id) decl.val_loc;
   Builtin_attributes.mark_alerts_used decl.val_attributes;
   Option.iter
@@ -2001,7 +1841,7 @@
     values = IdTbl.add id (Val_bound vda) env.values;
     summary = Env_value(env.summary, id, decl) }
 
-and store_constructor ~check type_decl type_id cstr_id cstr env =
+let store_constructor ~check type_decl type_id cstr_id cstr env =
   Builtin_attributes.warning_scope cstr.cstr_attributes (fun () ->
   if check && not type_decl.type_loc.Location.loc_ghost
      && Warnings.is_active (Warnings.Unused_constructor ("", Unused))
@@ -2036,7 +1876,7 @@
         { cda_description = cstr; cda_address = None; cda_shape } env.constrs;
   }
 
-and store_label ~check type_decl type_id lbl_id lbl env =
+let store_label ~check type_decl type_id lbl_id lbl env =
   Builtin_attributes.warning_scope lbl.lbl_attributes (fun () ->
   if check && not type_decl.type_loc.Location.loc_ghost
      && Warnings.is_active (Warnings.Unused_field ("", Unused))
@@ -2068,7 +1908,7 @@
     labels = TycompTbl.add lbl_id lbl env.labels;
   }
 
-and store_type ~check id info shape env =
+let store_type ~check id info shape env =
   let loc = info.type_loc in
   if check then
     check_usage loc id info.type_uid
@@ -2107,7 +1947,7 @@
     types = IdTbl.add id tda env.types;
     summary = Env_type(env.summary, id, info) }
 
-and store_type_infos ~tda_shape id info env =
+let store_type_infos ~tda_shape id info env =
   (* Simplified version of store_type that doesn't compute and store
      constructor and label infos, but simply record the arity and
      manifest-ness of the type.  Used in components_of_module to
@@ -2124,7 +1964,7 @@
     types = IdTbl.add id tda env.types;
     summary = Env_type(env.summary, id, info) }
 
-and store_extension ~check ~rebind id addr ext shape env =
+let store_extension ~check ~rebind id addr ext shape env =
   let loc = ext.ext_loc in
   let cstr =
     Datarepr.extension_descr
@@ -2164,7 +2004,7 @@
     constrs = TycompTbl.add id cda env.constrs;
     summary = Env_extension(env.summary, id, ext) }
 
-and store_module ?(update_summary=true) ~check
+let store_module ?(update_summary=true) ~check
                  id addr presence md shape env =
   let open Subst.Lazy in
   let loc = md.mdl_loc in
@@ -2189,7 +2029,7 @@
     modules = IdTbl.add id (Mod_local mda) env.modules;
     summary }
 
-and store_modtype ?(update_summary=true) id info shape env =
+let store_modtype ?(update_summary=true) id info shape env =
   Builtin_attributes.mark_alerts_used info.Subst.Lazy.mtdl_attributes;
   let mtda = { mtda_declaration = info; mtda_shape = shape } in
   let summary =
@@ -2199,7 +2039,7 @@
     modtypes = IdTbl.add id mtda env.modtypes;
     summary }
 
-and store_class id addr desc shape env =
+let store_class id addr desc shape env =
   Builtin_attributes.mark_alerts_used desc.cty_attributes;
   let clda =
     { clda_declaration = desc;
@@ -2210,13 +2050,198 @@
     classes = IdTbl.add id clda env.classes;
     summary = Env_class(env.summary, id, desc) }
 
-and store_cltype id desc shape env =
+let store_cltype id desc shape env =
   Builtin_attributes.mark_alerts_used desc.clty_attributes;
   let cltda = { cltda_declaration = desc; cltda_shape = shape } in
   { env with
     cltypes = IdTbl.add id cltda env.cltypes;
     summary = Env_cltype(env.summary, id, desc) }
 
+let components_of_module_maker
+          {cm_env; cm_prefixing_subst;
+           cm_path; cm_addr; cm_mty; cm_shape} : _ result =
+  match scrape_alias cm_env cm_mty with
+    MtyL_signature sg ->
+      let c =
+        { comp_values = NameMap.empty;
+          comp_constrs = NameMap.empty;
+          comp_labels = NameMap.empty; comp_types = NameMap.empty;
+          comp_modules = NameMap.empty; comp_modtypes = NameMap.empty;
+          comp_classes = NameMap.empty; comp_cltypes = NameMap.empty }
+      in
+      let items_and_paths, sub =
+        prefix_idents cm_path cm_prefixing_subst sg
+      in
+      let env = ref cm_env in
+      let pos = ref 0 in
+      let next_address () =
+        let addr : address_unforced =
+          Projection { parent = cm_addr; pos = !pos }
+        in
+        incr pos;
+        Lazy_backtrack.create addr
+      in
+      List.iter (fun ((item : Subst.Lazy.signature_item), path) ->
+        match item with
+          SigL_value(id, decl, _) ->
+            let decl' = Subst.value_description sub decl in
+            let addr =
+              match decl.val_kind with
+              | Val_prim _ -> Lazy_backtrack.create_failed Not_found
+              | _ -> next_address ()
+            in
+            let vda_shape = Shape.proj cm_shape (Shape.Item.value id) in
+            let vda =
+              { vda_description = decl'; vda_address = addr; vda_shape }
+            in
+            c.comp_values <- NameMap.add (Ident.name id) vda c.comp_values;
+        | SigL_type(id, decl, _, _) ->
+            let final_decl = Subst.type_declaration sub decl in
+            Btype.set_static_row_name final_decl
+              (Subst.type_path sub (Path.Pident id));
+            let descrs =
+              match decl.type_kind with
+              | Type_variant (_,repr) ->
+                  let cstrs = List.map snd
+                    (Datarepr.constructors_of_type path final_decl
+                        ~current_unit:(get_current_unit ()))
+                  in
+                  List.iter
+                    (fun descr ->
+                      let cda_shape = Shape.leaf descr.cstr_uid in
+                      let cda = {
+                        cda_description = descr;
+                        cda_address = None;
+                        cda_shape }
+                      in
+                      c.comp_constrs <-
+                        add_to_tbl descr.cstr_name cda c.comp_constrs
+                    ) cstrs;
+                 Type_variant (cstrs, repr)
+              | Type_record (_, repr) ->
+                  let lbls = List.map snd
+                    (Datarepr.labels_of_type path final_decl)
+                  in
+                  List.iter
+                    (fun descr ->
+                      c.comp_labels <-
+                        add_to_tbl descr.lbl_name descr c.comp_labels)
+                    lbls;
+                  Type_record (lbls, repr)
+              | Type_abstract r -> Type_abstract r
+              | Type_open -> Type_open
+              | Type_external name -> Type_external name
+            in
+            let shape = Shape.proj cm_shape (Shape.Item.type_ id) in
+            let tda =
+              { tda_declaration = final_decl;
+                tda_descriptions = descrs;
+                tda_shape = shape; }
+            in
+            c.comp_types <- NameMap.add (Ident.name id) tda c.comp_types;
+            env := store_type_infos ~tda_shape:shape id decl !env
+        | SigL_typext(id, ext, _, _) ->
+            let ext' = Subst.extension_constructor sub ext in
+            let descr =
+              Datarepr.extension_descr ~current_unit:(get_current_unit ()) path
+                ext'
+            in
+            let addr = next_address () in
+            let cda_shape =
+              Shape.proj cm_shape (Shape.Item.extension_constructor id)
+            in
+            let cda =
+              { cda_description = descr; cda_address = Some addr; cda_shape }
+            in
+            c.comp_constrs <- add_to_tbl (Ident.name id) cda c.comp_constrs
+        | SigL_module(id, pres, md, _, _) ->
+            let md' =
+              (* The prefixed items get the same scope as [cm_path], which is
+                 the prefix. *)
+              Subst.Lazy.module_decl
+                (Subst.Rescope (Path.scope cm_path)) sub md
+            in
+            let addr =
+              match pres with
+              | Mp_absent -> begin
+                  match md.mdl_type with
+                  | MtyL_alias path ->
+                      Lazy_backtrack.create (ModAlias {env = !env; path})
+                  | _ -> assert false
+                end
+              | Mp_present -> next_address ()
+            in
+            let alerts =
+              Builtin_attributes.alerts_of_attrs md.mdl_attributes
+            in
+            let shape = Shape.proj cm_shape (Shape.Item.module_ id) in
+            let comps =
+              components_of_module ~alerts ~uid:md.mdl_uid !env
+                sub path addr md.mdl_type shape
+            in
+            let mda =
+              { mda_declaration = md';
+                mda_components = comps;
+                mda_address = addr;
+                mda_shape = shape; }
+            in
+            c.comp_modules <-
+              NameMap.add (Ident.name id) mda c.comp_modules;
+            env :=
+              store_module ~update_summary:false ~check:None
+                id addr pres md shape !env
+        | SigL_modtype(id, decl, _) ->
+            let final_decl =
+              (* The prefixed items get the same scope as [cm_path], which is
+                 the prefix. *)
+              Subst.Lazy.modtype_decl (Rescope (Path.scope cm_path))
+                sub decl
+            in
+            let shape = Shape.proj cm_shape (Shape.Item.module_type id) in
+            let mtda =
+              { mtda_declaration = final_decl;
+                mtda_shape = shape; }
+            in
+            c.comp_modtypes <-
+              NameMap.add (Ident.name id) mtda c.comp_modtypes;
+            env := store_modtype ~update_summary:false id decl shape !env
+        | SigL_class(id, decl, _, _) ->
+            let decl' = Subst.class_declaration sub decl in
+            let addr = next_address () in
+            let shape = Shape.proj cm_shape (Shape.Item.class_ id) in
+            let clda =
+              { clda_declaration = decl';
+                clda_address = addr;
+                clda_shape = shape; }
+            in
+            c.comp_classes <- NameMap.add (Ident.name id) clda c.comp_classes
+        | SigL_class_type(id, decl, _, _) ->
+            let decl' = Subst.cltype_declaration sub decl in
+            let shape = Shape.proj cm_shape (Shape.Item.class_type id) in
+            let cltda = { cltda_declaration = decl'; cltda_shape = shape } in
+            c.comp_cltypes <-
+              NameMap.add (Ident.name id) cltda c.comp_cltypes)
+        items_and_paths;
+        Ok (Structure_comps c)
+  | MtyL_functor(arg, ty_res) ->
+      let sub = cm_prefixing_subst in
+      let scoping = Subst.Rescope (Path.scope cm_path) in
+      let open Subst.Lazy in
+        Ok (Functor_comps {
+          (* fcomp_arg and fcomp_res must be prefixed eagerly, because
+             they are interpreted in the outer environment *)
+          fcomp_arg =
+            (match arg with
+            | Unit -> Unit
+            | Named (param, ty_arg) ->
+              Named (param, force_modtype (modtype scoping sub ty_arg)));
+          fcomp_res = force_modtype (modtype scoping sub ty_res);
+          fcomp_shape = cm_shape;
+          fcomp_cache = Hashtbl.create 17;
+          fcomp_subst_cache = Hashtbl.create 17 })
+  | MtyL_ident _ -> Error No_components_abstract
+  | MtyL_alias p -> Error (No_components_alias p)
+
 let scrape_alias env mty = scrape_alias env mty
 
 (* Compute the components of a functor application in a path. *)
@@ -2638,7 +2663,7 @@
 
 (* Save a signature to a file *)
 let save_signature_with_transform cmi_transform ~alerts sg cmi_info =
-  Btype.cleanup_abbrev ();
+  Btype.cleanup_abbrev_memo ();
   Subst.reset_for_saving ();
   let sg = Subst.signature Make_local (Subst.for_saving Subst.identity) sg in
   let cmi =
@@ -3330,8 +3355,14 @@
   lookup_cltype ~errors:true ~use ~loc lid env
 
 let lookup_all_constructors ?(use=true) ~loc usage lid env =
-  match lookup_all_constructors ~errors:true ~use ~loc usage lid env with
-  | exception Error(Lookup_error(loc', env', err)) ->
+  match
+    (* We don't want errors to be logged here, as the caller will process them
+       (and log them if they see fit). *)
+    Typing_recovery.uncatch_errors (fun () ->
+      lookup_all_constructors ~errors:true ~use ~loc usage lid env
+    )
+  with
+  | exception Error.In_context (Lookup_error(loc', env', err)) ->
       (Error(loc', env', err) : _ result)
   | cstrs -> Ok cstrs
 
@@ -3342,8 +3373,14 @@
   lookup_all_constructors_from_type ~use ~loc usage ty_path env
 
 let lookup_all_labels ?(use=true) ~loc usage lid env =
-  match lookup_all_labels ~errors:true ~use ~loc usage lid env with
-  | exception Error(Lookup_error(loc', env', err)) ->
+  match
+    (* We don't want errors to be logged here, as the caller will process them
+       (and log them if they see fit). *)
+    Typing_recovery.uncatch_errors (fun () ->
+      lookup_all_labels ~errors:true ~use ~loc usage lid env
+    )
+  with
+  | exception Error.In_context (Lookup_error(loc', env', err)) ->
       (Error(loc', env', err) : _ result)
   | lbls -> Ok lbls
 
@@ -3873,7 +3910,7 @@
 let () =
   Location.register_error_of_exn
     (function
-      | Error err ->  Some (report_error_doc err)
+      | Error.In_context err ->  Some (report_error_doc err)
       | _ ->
           None
     )
