--- ocaml_505/typing/env.mli	2026-07-07 01:48:43.513022109 +0200
+++ ocaml_506_typing_recovery/typing/env.mli	2026-07-07 02:21:16.939555176 +0200
@@ -403,7 +403,7 @@
   signature * t
 
 (* Same as [enter_signature] but also extends the shape map ([parent_shape])
-   with all the the items from the signature, their shape being a projection
+   with all the items from the signature, their shape being a projection
    from the given shape. *)
 val enter_signature_and_shape: scope:int -> parent_shape:Shape.Map.t ->
   Shape.t -> signature -> t -> signature * Shape.Map.t * t
@@ -485,7 +485,12 @@
   | Illegal_value_name of Location.t * string
   | Lookup_error of Location.t * t * lookup_error
 
-exception Error of error
+module Error : sig
+  type exn += private In_context of error
+
+  val log_or_raise : error -> unit
+  val log_and_raise : error -> 'a
+end
 
 val in_signature: bool -> t -> t
 
