--- ocaml_505/typing/errortrace_report.ml	2026-07-07 01:48:43.513538859 +0200
+++ ocaml_506_typing_recovery/typing/errortrace_report.ml	2026-07-07 02:21:16.943728378 +0200
@@ -37,7 +37,7 @@
 module Fmt = Format_doc
 module Style = Misc.Style
 
-type 'a diff = 'a Out_type.diff = Same of 'a | Diff of 'a * 'a
+type 'a diff = Same of 'a | Diff of 'a * 'a
 
 let trees_of_trace mode =
   List.map (Errortrace.map_diff (trees_of_type_expansion mode))
@@ -92,7 +92,7 @@
 let prepare_trace f tr =
   prepare_any_trace printing_status (Errortrace.map f tr)
 
-(** Keep elements that are [Diff _ ] and split the the last element if it is
+(** Keep elements that are [Diff _ ] and split the last element if it is
     optionally elidable, require a prepared trace *)
 let rec filter_trace = function
   | [] -> [], None
@@ -243,12 +243,19 @@
         pp_doc pre pp_path (Path.Pident id)
     )
   | Errortrace.Equation Errortrace.{ty = _; expanded = t} ->
+      let[@manual.ref "s:gadts-type-inference"] manual_ref =
+        [ 7; 2 ]
+      in
       Variable_names.reserve t;
       Some(
-        doc_printf "%a@ @[<hov>This instance of %a is ambiguous:@ %s@]"
+        doc_printf "%a@ @[<hov>This instance of %a is ambiguous:@ %s@]\
+                    @,@[%a A type annotation may resolve \
+                    the ambiguity,@,either on this \
+                    expression or the whole function.@]"
           pp_doc pre
           (Style.as_inline_code type_expr_with_reserved_names) t
           "it would escape the scope of its equation"
+          Misc.print_manual_hint manual_ref
       )
   | Errortrace.Self ->
       Some (doc_printf "%a@,Self type cannot escape its class" pp_doc pre)
@@ -269,6 +276,13 @@
       Some (doc_printf
               "@,Self type cannot be unified with a closed object type"
            )
+  | Errortrace.Kind_differ (f, k1, k2) -> Some(
+      doc_printf
+        "@,@[The method %a is %s and was expected to be %s@]"
+          Style.inline_code f
+          (match k1 with Fpublic -> "public" | _ -> "private")
+          (match k2 with Fpublic -> "public" | _ -> "private")
+    )
 
 let explain_incompatible_fields name (diff: Types.type_expr Errortrace.diff) =
   Variable_names.reserve diff.got;
@@ -315,6 +329,52 @@
       Some(doc_printf "@,@[%a@]" Fmt.pp_doc pr)
   | Errortrace.Package_coercion pr ->
       Some(doc_printf "@,@[%a@]" Fmt.pp_doc pr)
+  | Errortrace.Constraint_on_missing_type (position,name) ->
+      let name = String.concat "." name in
+      let expl = match position with
+        | First ->
+            doc_printf "@,@[There is no type %a in the first module type.@]"
+              Style.inline_code name
+        | Second ->
+            doc_printf "@,@[There is no type %a in the second module type.@]"
+              Style.inline_code name
+      in
+      Some expl
+  | Errortrace.Constraint_with_deps (position,name) ->
+      let name = String.concat "." name in
+      let expl = match position with
+        | First ->
+            doc_printf
+              "@,@[The type %a depends on internal types@ in@ the@ \
+               first@ module@ type.@]"
+              Style.inline_code name
+        | Second ->
+            doc_printf
+              "@,@[The type %a depends on internal types@ in@ the@ \
+               second@ module@ type.@]"
+              Style.inline_code name
+      in
+      Some expl
+  | Errortrace.Constraint_on_mismatched_type {pos; decl; lhs }  ->
+      let name = String.concat "." lhs in
+      let id = Ident.create_local name in
+      let expl = match pos with
+        | First ->
+            doc_printf
+              "@,@[The constraint on %a in the second module type@ \
+               is not compatible@ with the declaration of@;<1 2>@[%a@]@ in \
+               the first module type.@]"
+              Style.inline_code name
+              (Printtyp.Doc.type_declaration id) decl
+        | Second ->
+            doc_printf
+              "@,@[The constraint on %a in the first module type@ \
+               is not compatible@ with the declaration of@;<1 2>@[%a@]@ in \
+               the second module type.@]"
+              Style.inline_code name
+              (Printtyp.Doc.type_declaration id) decl
+      in
+      Some expl
 
 let explain_univar prev = function
   | Errortrace.Var_mismatch { diff; order} ->
@@ -456,7 +516,7 @@
               "@,@[<hov>Type %a was considered abstract@ when checking\
                @ constraints@ in this@ recursive module definition.@]"
               pp_path p
-        | Definition | Existential _ -> ()
+        | Equation _ | Definition | Existential _ -> ()
       end
   | _ -> ()
 
@@ -480,6 +540,62 @@
       warn_on_missing_def env ppf te1;
       warn_on_missing_def env ppf te2
 
+let pp_print_list_comma_and elt ppf l =
+  match List.rev l with
+  | [] -> ()
+  | [ single ] ->
+      fprintf ppf "%a" elt single
+  | fst :: rest ->
+      fprintf
+        ppf
+        "%a@ and %a"
+        (pp_print_list ~pp_sep:comma elt) (List.rev rest)
+        elt fst
+
+let quoted_ident ppf t =
+  Style.as_inline_code !Oprint.out_ident ppf t
+
+let pp_plural (singular, plural) ppf l =
+  match l with
+  | [ _ ] -> pp_print_string ppf singular
+  | _ -> pp_print_string ppf plural
+
+let explain_names env ppf =
+  let explanations = Internal_names.explain env in
+  List.iter
+    (function
+      | _, Internal_names.Equation { lhs; rhs; } ->
+          add_type_to_preparation lhs;
+          add_type_to_preparation rhs;
+      | _, Internal_names.Existential _ ->
+          ()
+    ) explanations;
+  List.iter
+    (fun (paths, explanation) ->
+       let paths = List.map tree_of_path paths in
+       match explanation with
+       | Internal_names.Equation { lhs; rhs; } ->
+           let rhseq = tree_of_typexp Type_scheme rhs in
+           let lhseq = tree_of_typexp Type_scheme lhs in
+           fprintf ppf
+             "@ @[<2>@{<hint>Hint@}:@ %a@ %a@ \
+              introduced in the equation@ %a = %a@]"
+             (pp_print_list_comma_and quoted_ident) paths
+             (pp_plural ("is a type variable", "are type variables")) paths
+             (Style.as_inline_code !Oprint.out_type)
+             lhseq
+             (Style.as_inline_code !Oprint.out_type)
+             rhseq
+       | Internal_names.Existential { constructor } ->
+           fprintf ppf
+             "@ @[<2>@{<hint>Hint@}:@ %a@ %a@ \
+              bound by the constructor@ %a.@]"
+             (pp_print_list_comma_and quoted_ident) paths
+             (pp_plural ("is an existential type", "are existential types"))
+             paths
+             Style.inline_code constructor
+    ) explanations
+
 (* [subst] comes out of equality, and is [[]] otherwise *)
 let error trace_format mode subst env tr txt1 ppf txt2 ty_expect_explanation =
   reset ();
@@ -510,7 +626,7 @@
       in
       fprintf ppf
         "@[<v>\
-          @[%a%a@]%a%a\
+         @[%a%a@]%a%a\
          @]"
         pp_doc head_error
         pp_doc ty_expect_explanation
@@ -518,7 +634,7 @@
         (pp_print_option pp_doc) mis;
       if env <> Env.empty
       then warn_on_missing_defs env ppf head;
-      Internal_names.print_explanations env ppf;
+      explain_names env ppf;
       Ident_conflicts.err_print ppf
     )
 
@@ -631,9 +747,6 @@
 
 let subtype = Subtype.error
 
-let quoted_ident ppf t =
-  Style.as_inline_code !Oprint.out_ident ppf t
-
 let type_path_expansion ppf = function
   | Same p -> quoted_ident ppf p
   | Diff(p,p') ->
