--- ocaml_505/typing/errortrace.mli	2026-07-07 01:48:43.513538859 +0200
+++ ocaml_506_typing_recovery/typing/errortrace.mli	2026-07-07 02:21:16.942120719 +0200
@@ -87,11 +87,20 @@
   | Abstract_row : position -> _ obj
   (* Unification *)
   | Self_cannot_be_closed : unification obj
+  (* Equality & Moregen *)
+  | Kind_differ : string * field_kind_view * field_kind_view -> comparison obj
 
 type first_class_module =
     | Package_cannot_scrape of Path.t
     | Package_inclusion of Format_doc.doc
     | Package_coercion of Format_doc.doc
+    | Constraint_on_missing_type of position * string list
+    | Constraint_with_deps of position * string list
+    | Constraint_on_mismatched_type of {
+        pos:position;
+        decl:Types.type_declaration;
+        lhs:string list
+      }
 
 type univar =
   | Var_mismatch of { order:order; diff:type_expr diff }
@@ -118,6 +127,10 @@
 
 val map : ('a -> 'b) -> ('a, 'variety) t -> ('b, 'variety) t
 
+val map_types :
+  (type_expr -> type_expr) ->
+  (expanded_type, 'variety) t -> (expanded_type, 'variety) t
+
 val incompatible_fields :
   name:string -> got:type_expr -> expected:type_expr -> (type_expr, _) elt
 
@@ -182,4 +195,6 @@
     trace:error_trace -> unification_trace:unification_error_trace -> error
 
   val map : ('a -> 'b) -> 'a t -> 'b t
+
+  val map_types : (type_expr -> type_expr) -> expanded_type t -> expanded_type t
 end
