--- ocaml_505/typing/ident.ml	2026-07-07 01:48:43.513538859 +0200
+++ ocaml_506_typing_recovery/typing/ident.ml	2026-07-07 02:21:16.948572274 +0200
@@ -43,6 +43,9 @@
     | Ulink u -> get_desc u
 
   let name us = (get_desc us).name
+  let print ppf us =
+    let d = get_desc us in
+    Format.fprintf ppf "%s/%d" d.name d.stamp
 
   let refresh us = create (get_desc us).name
 
@@ -287,7 +290,7 @@
         pp_stamped (name, stamp)
   | Unscoped us ->
       let Unscoped.{ name; stamp } = Unscoped.get_desc us in
-      fprintf ppf "U%a"
+      fprintf ppf "U:%a"
         pp_stamped (name, stamp)
   | Scoped { name; stamp; scope } ->
       fprintf ppf "%a%s"
