--- ocaml_505/typing/includecore.ml	2026-07-07 01:48:43.514538858 +0200
+++ ocaml_506_typing_recovery/typing/includecore.ml	2026-07-07 02:21:16.951046045 +0200
@@ -112,7 +112,7 @@
          included into (usually numbered with "2" in this file).  In this case,
          the abstract row variable has been substituted for an object or variant
          type. *)
-      begin match get_desc (Ctype.expand_head env ty) with
+      begin match get_desc (Ctype.expand_head_nolink env ty) with
       | Tobject _|Tvariant _ -> true
       | _ -> false
       end
@@ -412,7 +412,7 @@
       pr "The constructor %a is only present in %s %s."
         Style.inline_code name (choose ord first second) decl
   | Presence s ->
-      pr "The tag %a is present in the %s %s,@ but might not be in the %s"
+      pr "The tag %a is present in %s %s,@ but might not be in %s"
         (Style.as_inline_code pp_tag) s second decl first
   | Incompatible_types_for s -> pr "Types for tag `%s are incompatible" s
   | Types err ->
@@ -782,7 +782,7 @@
         when Option.is_some decl2.type_manifest -> begin
           match decl1.type_manifest with
           | Some ty1 -> begin
-            let ty1 = Ctype.expand_head env ty1 in
+            let ty1 = Ctype.expand_head_nolink env ty1 in
             match get_desc ty1 with
             | Tvariant row when Btype.is_constr_row ~allow_ident:true
                                   (row_more row) ->
@@ -894,7 +894,8 @@
   end
 
 let type_manifest env ty1 params1 ty2 params2 priv2 kind2 =
-  let ty1' = Ctype.expand_head env ty1 and ty2' = Ctype.expand_head env ty2 in
+  let ty1' = Ctype.expand_head_nolink env ty1
+  and ty2' = Ctype.expand_head_nolink env ty2 in
   match get_desc ty1', get_desc ty2' with
   | Tvariant row1, Tvariant row2
     when is_absrow env (row_more row2) -> begin
@@ -932,7 +933,8 @@
         else
           Ctype.equal env true (params1 @ [ty1]) (params2 @ [ty2])
       with
-      | exception Ctype.Equality err -> Some (Manifest err)
+      | exception Ctype.Equality err ->
+          Some (Manifest err)
       | () -> None
     end
 
