--- ocaml_505/typing/predef.mli	2026-07-07 01:48:43.515538857 +0200
+++ ocaml_506_typing_recovery/typing/predef.mli	2026-07-07 02:21:16.971448391 +0200
@@ -33,6 +33,7 @@
   | `Floatarray
   | `Iarray
   | `Atomic_loc
+  | `Todo_info
 ]
 type data_type_constr = [
   | `Bool
@@ -70,6 +71,7 @@
 val type_extension_constructor: type_expr
 val type_floatarray: type_expr
 val type_atomic_loc: type_expr -> type_expr
+val type_todo_info: type_expr
 
 val path_int: Path.t
 val path_char: Path.t
@@ -91,10 +93,12 @@
 val path_extension_constructor: Path.t
 val path_floatarray: Path.t
 val path_continuation: Path.t
+val path_todo_info: Path.t
 
 val path_match_failure: Path.t
 val path_assert_failure : Path.t
 val path_undefined_recursive_module : Path.t
+val path_todo : Path.t
 
 val ident_false : Ident.t
 val ident_true : Ident.t
