--- ocaml_505/typing/primitive.ml	2026-07-07 01:48:43.515538857 +0200
+++ ocaml_506_typing_recovery/typing/primitive.ml	2026-07-07 02:21:16.972150771 +0200
@@ -16,7 +16,6 @@
 (* Description of primitive functions *)
 
 open Misc
-open Parsetree
 
 type boxed_integer = Pnativeint | Pint32 | Pint64
 
@@ -81,10 +80,10 @@
    prim_native_repr_args = native_repr_args;
    prim_native_repr_res = native_repr_res}
 
-let parse_declaration valdecl ~native_repr_args ~native_repr_res =
+let parse_description ~native_repr_args ~native_repr_res ~prim ~attrs ~loc =
   let arity = List.length native_repr_args in
   let name, native_name, old_style_noalloc, old_style_float =
-    match valdecl.pval_prim with
+    match prim with
     | name :: "noalloc" :: name2 :: "float" :: _ -> (name, name2, true, true)
     | name :: "noalloc" :: name2 :: _ -> (name, name2, true, false)
     | name :: name2 :: "float" :: _ -> (name, name2, false, true)
@@ -92,34 +91,30 @@
     | name :: name2 :: _ -> (name, name2, false, false)
     | name :: _ -> (name, "", false, false)
     | [] ->
-        fatal_error "Primitive.parse_declaration"
+        fatal_error "Primitive.parse_description"
   in
   let noalloc_attribute =
-    Attr_helper.has_no_payload_attribute "noalloc" valdecl.pval_attributes
+    Attr_helper.has_no_payload_attribute "noalloc" attrs
   in
   if old_style_float &&
      not (List.for_all is_ocaml_repr native_repr_args &&
           is_ocaml_repr native_repr_res) then
-    raise (Error (valdecl.pval_loc,
-                  Old_style_float_with_native_repr_attribute));
+    raise (Error (loc, Old_style_float_with_native_repr_attribute));
   if old_style_noalloc && noalloc_attribute then
-    raise (Error (valdecl.pval_loc,
-                  Old_style_noalloc_with_noalloc_attribute));
+    raise (Error (loc, Old_style_noalloc_with_noalloc_attribute));
   (* The compiler used to assume "noalloc" with "float", we just make this
      explicit now (GPR#167): *)
   let old_style_noalloc = old_style_noalloc || old_style_float in
   if old_style_float then
-    Location.deprecated valdecl.pval_loc
+    Location.deprecated loc
       "[@@unboxed] + [@@noalloc] should be used\n\
        instead of \"float\""
   else if old_style_noalloc then
-    Location.deprecated valdecl.pval_loc
-      "[@@noalloc] should be used instead of \"noalloc\"";
+    Location.deprecated loc "[@@noalloc] should be used instead of \"noalloc\"";
   if native_name = "" &&
      not (List.for_all is_ocaml_repr native_repr_args &&
           is_ocaml_repr native_repr_res) then
-    raise (Error (valdecl.pval_loc,
-                  No_native_primitive_with_repr_attribute));
+    raise (Error (loc, No_native_primitive_with_repr_attribute));
   let noalloc = old_style_noalloc || noalloc_attribute in
   let native_repr_args, native_repr_res =
     if old_style_float then
