--- ocaml_505/typing/primitive.mli	2026-07-07 01:48:43.516538856 +0200
+++ ocaml_506_typing_recovery/typing/primitive.mli	2026-07-07 02:21:16.972286889 +0200
@@ -49,10 +49,12 @@
   -> native_repr_res: native_repr
   -> description
 
-val parse_declaration
-  :  Parsetree.value_description
-  -> native_repr_args:native_repr list
+val parse_description
+  :  native_repr_args:native_repr list
   -> native_repr_res:native_repr
+  -> prim:string list
+  -> attrs:Parsetree.attributes
+  -> loc:Location.t
   -> description
 
 val print
