--- ocaml_505/typing/printtyped.ml	2026-07-07 01:48:43.517538855 +0200
+++ ocaml_506_typing_recovery/typing/printtyped.ml	2026-07-07 02:21:16.977096890 +0200
@@ -490,8 +490,24 @@
   line i ppf "value_description %a %a\n" fmt_ident x.val_id fmt_location
        x.val_loc;
   attributes i ppf x.val_attributes;
-  core_type (i+1) ppf x.val_desc;
-  list (i+1) string ppf x.val_prim;
+  core_type (i+1) ppf x.val_desc
+
+and primitive_description i ppf x =
+  line i ppf "primitive_description %a %a\n" fmt_ident x.prim_id fmt_location
+       x.prim_loc;
+  attributes i ppf x.prim_attributes;
+  let i = i + 1 in
+  match x.prim_kind with
+  | Tprim_decl (typ, prim) ->
+    line i ppf "Tprim_decl\n";
+    let i = i + 1 in
+    core_type i ppf typ;
+    list i string ppf prim
+  | Tprim_alias (typ, path, _) ->
+    line i ppf "Tprim_alias\n";
+    let i = i + 1 in
+    option i core_type ppf typ;
+    line i ppf "%a\n" fmt_path path
 
 and binding_op i ppf x =
   line i ppf "binding_op %a %a\n" fmt_path x.bop_op_path
@@ -771,6 +787,9 @@
   | Tsig_value vd ->
       line i ppf "Tsig_value\n";
       value_description i ppf vd;
+  | Tsig_primitive pd ->
+      line i ppf "Tsig_primitive\n";
+      primitive_description i ppf pd;
   | Tsig_type (rf, l) ->
       line i ppf "Tsig_type %a\n" fmt_rec_flag rf;
       list i type_declaration ppf l;
@@ -895,9 +914,9 @@
   | Tstr_value (rf, l) ->
       line i ppf "Tstr_value %a\n" fmt_rec_flag rf;
       list i (value_binding rf) ppf l;
-  | Tstr_primitive vd ->
+  | Tstr_primitive pd ->
       line i ppf "Tstr_primitive\n";
-      value_description i ppf vd;
+      primitive_description i ppf pd;
   | Tstr_type (rf, l) ->
       line i ppf "Tstr_type %a\n" fmt_rec_flag rf;
       list i type_declaration ppf l;
