--- ocaml_505/typing/tast_mapper.ml	2026-07-07 01:48:43.517538855 +0200
+++ ocaml_506_typing_recovery/typing/tast_mapper.ml	2026-07-07 02:21:16.989934616 +0200
@@ -69,6 +69,8 @@
     value_bindings: mapper -> (rec_flag * value_binding list) ->
       (rec_flag * value_binding list);
     value_description: mapper -> value_description -> value_description;
+    primitive_description:
+      mapper -> primitive_description -> primitive_description;
     with_constraint: mapper -> with_constraint -> with_constraint;
   }
 
@@ -166,7 +168,7 @@
     | Tstr_value (rec_flag, list) ->
         let (rec_flag, list) = sub.value_bindings sub (rec_flag, list) in
         Tstr_value (rec_flag, list)
-    | Tstr_primitive v -> Tstr_primitive (sub.value_description sub v)
+    | Tstr_primitive p -> Tstr_primitive (sub.primitive_description sub p)
     | Tstr_type (rec_flag, list) ->
         let (rec_flag, list) = sub.type_declarations sub (rec_flag, list) in
         Tstr_type (rec_flag, list)
@@ -197,6 +199,18 @@
   let val_attributes = sub.attributes sub x.val_attributes in
   {x with val_loc; val_name; val_desc; val_attributes}
 
+let primitive_description sub x =
+  let prim_loc = sub.location sub x.prim_loc in
+  let prim_name = map_loc sub x.prim_name in
+  let prim_kind =
+    match x.prim_kind with
+    | Tprim_decl (typ, prim) -> Tprim_decl (sub.typ sub typ, prim)
+    | Tprim_alias (typ, path, lid) ->
+      Tprim_alias (Option.map (sub.typ sub) typ, path, map_loc_lid sub lid)
+  in
+  let prim_attributes = sub.attributes sub x.prim_attributes in
+  {x with prim_loc; prim_name; prim_kind; prim_attributes}
+
 let label_decl sub x =
   let ld_loc = sub.location sub x.ld_loc in
   let ld_name = map_loc sub x.ld_name in
@@ -529,6 +543,8 @@
     match x.sig_desc with
     | Tsig_value v ->
         Tsig_value (sub.value_description sub v)
+    | Tsig_primitive p ->
+        Tsig_primitive (sub.primitive_description sub p)
     | Tsig_type (rec_flag, list) ->
         let (rec_flag, list) = sub.type_declarations sub (rec_flag, list) in
         Tsig_type (rec_flag, list)
@@ -920,5 +936,6 @@
     value_binding;
     value_bindings;
     value_description;
+    primitive_description;
     with_constraint;
   }
