--- ocaml_505/typing/types.mli	2026-07-07 01:48:43.519538853 +0200
+++ ocaml_506_typing_recovery/typing/types.mli	2026-07-07 02:21:17.022026068 +0200
@@ -111,18 +111,6 @@
   | Tnil
   (** [Tnil] ==> [<...; >] *)
 
-  | Tlink of type_expr
-  (** Indirection used by unification engine. *)
-
-  | Tsubst of type_expr * type_expr option
-  (** [Tsubst] is used temporarily to store information in low-level
-      functions manipulating representation of types, such as
-      instantiation or copy.
-      The first argument contains a copy of the original node.
-      The second is available only when the first is the row variable of
-      a polymorphic variant.  It then contains a copy of the whole variant.
-      This constructor should not appear outside of these cases. *)
-
   | Tvariant of row_desc
   (** Representation of polymorphic variants, see [row_desc]. *)
 
@@ -141,6 +129,22 @@
   | Tfunctor of arg_label * Ident.Unscoped.t * package * type_expr
   (** Type of a dependent arrow *)
 
+  | Texpand of type_expr * Path.t * type_expr list
+  (** [Texpand] is like [Tlink] but the result of an expansion;
+      [Path.t] and [type_expr list] remember the original declaration. *)
+
+  | Tlink of type_expr
+  (** Indirection used by unification engine. *)
+
+  | Tsubst of type_expr * type_expr option
+  (** [Tsubst] is used temporarily to store information in low-level
+      functions manipulating representation of types, such as
+      instantiation or copy.
+      The first argument contains a copy of the original node.
+      The second is available only when the first is the row variable of
+      a polymorphic variant.  It then contains a copy of the whole variant.
+      This constructor should not appear outside of these cases. *)
+
 (** [package] corresponds to the type of a first-class module *)
 and package =
   { pack_path : Path.t;
@@ -166,17 +170,21 @@
     tail.
 
     Note on marshalling: [abbrev_memo] must not appear in saved types.
-    [Btype], with [cleanup_abbrev] and [memo], takes care of tracking and
+    [Btype], with [cleanup_abbrev_memo] and [memo], takes care of tracking and
     removing abbreviations.
 *)
 and abbrev_memo =
   | Mnil (** No known abbreviation *)
 
-  | Mcons of private_flag * Path.t * type_expr * type_expr * abbrev_memo
+  | Mcons of
+      { privacy : private_flag;
+        path : Path.t;
+        abbreviation : type_expr;
+        expansion : type_expr;
+        rem : abbrev_memo }
   (** Found one abbreviation.
       A valid abbreviation should be at least as visible and reachable by the
-      same path.
-      The first expression is the abbreviation and the second the expansion. *)
+      same path. *)
 
   | Mlink of abbrev_memo ref
   (** Abbreviations can be found after this indirection *)
@@ -258,6 +266,12 @@
 
            Return false if it was already marked *)
 
+(** Handle kept abbreviations *)
+val get_abbrev: type_expr -> (Path.t * type_expr list) option
+val iter_abbrev: (Path.t -> type_expr list -> unit) -> type_expr -> unit
+val forget_abbrev: type_expr -> unit
+val ignore_abbrev: type_expr -> type_expr
+
 (** Transient [type_expr].
     Should only be used immediately after [Transient_expr.repr] *)
 type transient_expr = private
@@ -286,9 +300,12 @@
           Fail if already instantiated. *)
 
   val try_mark_node: type_mark -> transient_expr -> bool
+
+  val eq : transient_expr -> transient_expr -> bool
 end
 
-val create_expr: type_desc -> level: int -> scope: int -> id: int -> type_expr
+val create_expr:
+    type_desc -> level: int -> scope: int -> id: int -> type_expr
 
 (** Functions and definitions moved from Btype *)
 
@@ -479,6 +496,7 @@
 and class_signature =
   { csig_self: type_expr;
     mutable csig_self_row: type_expr;
+    mutable csig_dummy_method : field_kind;
     mutable csig_vars: (mutable_flag * virtual_flag * type_expr) Vars.t;
     mutable csig_meths: (method_privacy * virtual_flag * type_expr) Meths.t; }
 
@@ -587,6 +605,7 @@
   | Rec_check_regularity       (* See Typedecl.transl_type_decl *)
   | Approx_recmod
   | Existential of string
+  | Equation of type_expr * type_expr
 
 and record_representation =
     Record_regular                      (* All fields are boxed / tagged *)
@@ -747,25 +766,34 @@
         (* A snapshot for backtracking *)
 val snapshot: unit -> snapshot
         (* Make a snapshot for later backtracking. Costs nothing *)
-val backtrack: cleanup_abbrev:(unit -> unit) -> snapshot -> unit
+val backtrack: cleanup:(unit -> unit) -> snapshot -> unit
         (* Backtrack to a given snapshot. Only possible if you have
            not already backtracked to a previous snapshot.
-           Calls [cleanup_abbrev] internally *)
+           Calls [cleanup] internally *)
 val undo_first_change_after: snapshot -> unit
         (* Backtrack only the first change after a snapshot.
            Does not update the list of changes *)
 val undo_compress: snapshot -> unit
         (* Backtrack only path compression. Only meaningful if you have
            not already backtracked to a previous snapshot.
-           Does not call [cleanup_abbrev] *)
+           Does not call [cleanup] *)
 
 (** Functions to use when modifying a type (only Ctype?).
     The old values are logged and reverted on backtracking.
  *)
 
+val link_expand: type_expr -> type_expr -> unit
+        (* Set the desc field of [t1] to [Texpand (t2, p)],
+           assuming that [t1] is [Tconstr (p, args, _)], and
+           logging the old values if there is an active snapshot.
+           To be used only when expanding a type abbreviation. *)
 val link_type: type_expr -> type_expr -> unit
         (* Set the desc field of [t1] to [Tlink t2], logging the old
-           value if there is an active snapshot *)
+           values if there is an active snapshot.
+           To be used during unification.
+           Must only be used if
+           - [t1] is [Tvar], or
+           - [t1] and [t2] have the same [type_desc] constructor. *)
 val set_type_desc: type_expr -> type_desc -> unit
         (* Set directly the desc field, without sharing *)
 val set_level: type_expr -> int -> unit
@@ -780,3 +808,5 @@
 val link_kind: inside:field_kind -> field_kind -> unit
 val link_commu: inside:commutable -> commutable -> unit
 val set_commu_ok: commutable -> unit
+
+val reset : unit -> unit
