--- ocaml_505/typing/typeopt.mli	2026-07-07 01:48:43.519538853 +0200
+++ ocaml_506_typing_recovery/typing/typeopt.mli	2026-07-07 02:21:17.020661319 +0200
@@ -30,11 +30,19 @@
       Env.t -> Types.type_expr -> Lambda.bigarray_kind * Lambda.bigarray_layout
 val value_kind : Env.t -> Types.type_expr -> Lambda.value_kind
 
-val classify_lazy_argument : Typedtree.expression ->
-                             [ `Constant_or_function
-                             | `Float_that_cannot_be_shortcut
-                             | `Identifier of [`Forward_value | `Other]
-                             | `Other]
+(** [lazy_summary] describes how the expression [lazy e] must be compiled. *)
+type lazy_summary =
+  | Lazy_thunk
+    (** Evaluation of [e] must be delayed inside a thunk. *)
+  | Eager of forward_repr
+    (** Evaluation of [e] can be done eagerly. *)
+and forward_repr =
+  | Forward
+    (** The value must be placed inside a [Forward] block. *)
+  | Shortcut
+    (** The value can be injected directly without wrapping. *)
+
+val classify_lazy_argument : Typedtree.expression -> lazy_summary
 
 val value_kind_union :
       Lambda.value_kind -> Lambda.value_kind -> Lambda.value_kind
