--- ocaml_505/typing/typedecl_separability.ml	2026-07-07 01:48:43.518538854 +0200
+++ ocaml_506_typing_recovery/typing/typedecl_separability.ml	2026-07-07 02:21:17.008692098 +0200
@@ -53,7 +53,11 @@
   | Type_abstract _ ->
       begin match def.type_manifest with
       | None -> Abstract
-      | Some type_expr -> Synonym type_expr
+      | Some type_expr ->
+          if !Clflags.typing_recovery
+          && Typing_recovery.erroneous_type_check type_expr
+          then Abstract
+          else Synonym type_expr
       end
   | Type_external _ -> Abstract
 
@@ -153,13 +157,15 @@
       (* these should only occur under Tobject and not at the toplevel,
          but "better safe than sorry" *)
       immediate_subtypes_object_row [] ty
-  | Tlink _ | Tsubst _ -> assert false (* impossible due to Ctype.repr *)
   | Tvar _ | Tunivar _ -> []
   | Tpoly (pty, _) -> [pty]
   | Tconstr (_path, tys, _) -> tys
   | Tfunctor _ ->
     Misc.fatal_error
       "[Typedecl_separability.immediate_subtypes] invalid argument"
+  | Tlink _  | Texpand _ -> assert false
+        (* impossible cases due to [get_desc] and [Transient_expr.repr] *)
+  | Tsubst _ -> assert false
 
 and immediate_subtypes_object_row acc ty = match get_desc ty with
   | Tnil -> acc
@@ -400,8 +406,9 @@
     else
     let hyps = Hyps.add ty m hyps in
     match (get_desc ty, m) with
-    (* Impossible case due to the call to [Ctype.repr]. *)
-    | (Tlink _            , _      ) -> assert false
+    (* Impossible cases due to the call to [get_desc]. *)
+    | (Tlink _            , _      )
+    | (Texpand _          , _      ) -> assert false
     (* Impossible case (according to comment in [typing/types.mli]. *)
     | (Tsubst(_)          , _      ) -> assert false
     (* "Indifferent" case, the empty context is sufficient. *)
