--- ocaml_505/typing/typeclass.ml	2026-07-07 01:48:43.517538855 +0200
+++ ocaml_506_typing_recovery/typing/typeclass.ml	2026-07-07 02:21:16.992532053 +0200
@@ -65,12 +65,17 @@
   | Class_type
 
 type final =
-  | Final
-  | Not_final
+  | Definitely_final
+  | Maybe_final
+  | Definitely_not_final
 
 let kind_of_final = function
-  | Final -> Object
-  | Not_final -> Class
+  | Definitely_final -> Object
+  | Maybe_final | Definitely_not_final -> Class
+
+let inheriting_final = function
+  | Definitely_not_final -> Definitely_not_final
+  | Maybe_final | Definitely_final -> Maybe_final
 
 type error =
   | Unconsistent_constraint of Errortrace.unification_error
@@ -109,9 +114,22 @@
   | Closing_self_type of class_signature
   | Polymorphic_class_parameter
 
-exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
 
+module Error : sig
+  type exn += private In_context of Location.t * Env.t * error
+  val log_or_raise : Location.t -> Env.t -> error -> unit
+  val log_and_raise : Location.t -> Env.t -> error -> 'a
+end = struct
+  type exn += In_context of Location.t * Env.t * error
+
+  let log_and_raise loc env err =
+    Typing_recovery.log_and_raise (In_context (loc, env, err))
+
+  let log_or_raise loc env err =
+    Typing_recovery.log_or_raise (In_context (loc, env, err))
+end
+
 open Typedtree
 
 let type_open_descr :
@@ -147,28 +165,20 @@
   node
 
 let update_class_signature loc env ~warn_implicit_public virt kind sign =
-  let implicit_public, implicit_declared =
-    Ctype.update_class_signature env sign
-  in
-  if implicit_declared <> [] then begin
+  let implicitly_public = Ctype.update_implicitly_public_methods sign in
+  if warn_implicit_public && implicitly_public <> [] then begin
+    Location.prerr_warning
+      loc (Warnings.Implicit_public_methods implicitly_public)
+  end;
+  let implicitly_declared = Ctype.update_implicitly_declared_methods env sign in
+  if implicitly_declared <> [] then begin
     match virt with
     | Virtual -> () (* Should perhaps emit warning 17 here *)
     | Concrete ->
-        raise (Error(loc, env, Undeclared_methods(kind, implicit_declared)))
-  end;
-  if warn_implicit_public && implicit_public <> [] then begin
-    Location.prerr_warning
-      loc (Warnings.Implicit_public_methods implicit_public)
+        Error.log_and_raise loc env
+          (Undeclared_methods (kind, implicitly_declared))
   end
 
-let complete_class_signature loc env virt kind sign =
-  update_class_signature loc env ~warn_implicit_public:false virt kind sign;
-  Ctype.hide_private_methods env sign
-
-let complete_class_type loc env virt kind typ =
-  let sign = Btype.signature_of_class_type typ in
-  complete_class_signature loc env virt kind sign
-
 let check_virtual loc env virt kind sign =
   match virt with
   | Virtual -> ()
@@ -176,7 +186,7 @@
       match Btype.virtual_methods sign, Btype.virtual_instance_vars sign with
       | [], [] -> ()
       | meths, vars ->
-          raise(Error(loc, env, Virtual_class(kind, meths, vars)))
+          Error.log_and_raise loc env (Virtual_class(kind, meths, vars))
 
 let rec check_virtual_clty loc env virt kind clty =
   match clty with
@@ -196,6 +206,13 @@
       let ty = Ctype.newmono ty in
       Ctype.newty (Tarrow (l, ty, constructor_type constr cty, commu_ok))
 
+let add_dummy_method env final ~scope sign =
+  match final with
+  | Definitely_not_final -> Ctype.add_dummy_method env ~scope sign
+  | Maybe_final | Definitely_final ->
+      (* Don't prevent closing the self type if it might be final *)
+      ()
+
                 (***********************************)
                 (*  Primitives for typing classes  *)
                 (***********************************)
@@ -203,21 +220,22 @@
 let raise_add_method_failure loc env label sign failure =
   match (failure : Ctype.add_method_failure) with
   | Ctype.Unexpected_method ->
-      raise(Error(loc, env, Unexpected_field (sign.Types.csig_self, label)))
+      Error.log_and_raise loc env
+        (Unexpected_field (sign.Types.csig_self, label))
   | Ctype.Type_mismatch trace ->
-      raise(Error(loc, env, Field_type_mismatch ("method", label, trace)))
+      Error.log_and_raise loc env (Field_type_mismatch ("method", label, trace))
 
 let raise_add_instance_variable_failure loc env label failure =
   match (failure : Ctype.add_instance_variable_failure) with
   | Ctype.Mutability_mismatch mut ->
-      raise (Error(loc, env, Mutability_mismatch(label, mut)))
+      Error.log_and_raise loc env (Mutability_mismatch (label, mut))
   | Ctype.Type_mismatch trace ->
-      raise (Error(loc, env,
-        Field_type_mismatch("instance variable", label, trace)))
+      Error.log_and_raise loc env
+        (Field_type_mismatch("instance variable", label, trace))
 
 let raise_inherit_class_signature_failure loc env sign = function
   | Ctype.Self_type_mismatch trace ->
-      raise(Error(loc, env, Self_clash trace))
+      Error.log_and_raise loc env (Self_clash trace)
   | Ctype.Method(label, failure) ->
       raise_add_method_failure loc env label sign failure
   | Ctype.Instance_variable(label, failure) ->
@@ -246,7 +264,7 @@
     match Btype.scrape_class_type cty2 with
     | Cty_signature sign2 -> sign2
     | _ ->
-      raise(Error(loc, env, Structure_expected cty2))
+        Error.log_and_raise loc env (Structure_expected cty2)
   in
   inherit_class_signature ~strict loc env sign1 sign2
 
@@ -254,7 +272,7 @@
   match Ctype.unify env ty expected_ty with
   | () -> ()
   | exception Ctype.Unify trace ->
-      raise(Error(loc, env, Field_type_mismatch ("method", label, trace)))
+      Error.log_and_raise loc env (Field_type_mismatch ("method", label, trace))
 
 let type_constraint val_env sty sty' loc =
   let cty  = transl_simple_type val_env ~closed:false sty in
@@ -263,7 +281,7 @@
   let ty' = cty'.ctyp_type in
   begin
     try Ctype.unify val_env ty ty' with Ctype.Unify err ->
-        raise(Error(loc, val_env, Unconsistent_constraint err));
+      Error.log_and_raise loc val_env (Unconsistent_constraint err);
   end;
   (cty, cty')
 
@@ -298,8 +316,9 @@
       mkctf_with_attrs
         (fun () ->
           let parent = class_type env Virtual self_scope sparent in
-          complete_class_type parent.cltyp_loc
-            env Virtual Class_type parent.cltyp_type;
+          let parent_sign = Btype.signature_of_class_type parent.cltyp_type in
+          update_class_signature parent.cltyp_loc ~warn_implicit_public:false
+            env Virtual Class_type parent_sign;
           inherit_class_type ~strict:false loc env sign parent.cltyp_type;
           Tctf_inherit parent)
   | Pctf_val ({txt=lab}, mut, virt, sty) ->
@@ -358,7 +377,7 @@
   begin try
     Ctype.unify env self_type sign.csig_self
   with Ctype.Unify _ ->
-    raise(Error(sty.ptyp_loc, env, Pattern_type_clash self_type))
+    Error.log_and_raise sty.ptyp_loc env (Pattern_type_clash self_type)
   end;
 
   (* Class type fields *)
@@ -389,28 +408,30 @@
   | Pcty_constr (lid, styl) ->
       let (path, decl) = Env.lookup_cltype ~loc:scty.pcty_loc lid.txt env in
       if Path.same decl.clty_path unbound_class then
-        raise(Error(scty.pcty_loc, env, Unbound_class_type_2 lid.txt));
+        Error.log_and_raise scty.pcty_loc env (Unbound_class_type_2 lid.txt);
       let (params, clty) =
         Ctype.instance_class decl.clty_params decl.clty_type
       in
+      let sign = Btype.signature_of_class_type clty in
       (* Adding a dummy method to the self type prevents it from being closed /
          escaping. *)
-      Ctype.add_dummy_method env ~scope:self_scope
-        (Btype.signature_of_class_type clty);
+      Ctype.add_dummy_method env ~scope:self_scope sign;
+      Ctype.reveal_private_methods env sign;
+      Ctype.set_object_name decl.clty_path params sign.csig_self;
       if List.length params <> List.length styl then
-        raise(Error(scty.pcty_loc, env,
-                    Parameter_arity_mismatch (lid.txt, List.length params,
-                                                   List.length styl)));
+        Error.log_and_raise scty.pcty_loc env
+          (Parameter_arity_mismatch
+             (lid.txt, List.length params, List.length styl));
       let ctys = List.map2
         (fun sty ty ->
           let cty' = transl_simple_type env ~closed:false sty in
           let ty' = cty'.ctyp_type in
           begin
-           try Ctype.unify env ty' ty with Ctype.Unify err ->
-                  raise(Error(sty.ptyp_loc, env, Parameter_mismatch err))
+            try Ctype.unify env ty' ty with Ctype.Unify err ->
+              Error.log_and_raise sty.ptyp_loc env (Parameter_mismatch err)
             end;
             cty'
-        )       styl params
+        ) styl params
       in
       let typ = Cty_constr (path, params, clty) in
       (* Check for unexpected virtual methods *)
@@ -571,7 +592,7 @@
     local_vals : VarSet.t;
     vars : Ident.t Vars.t; }
 
-let rec class_field_first_pass self_loc cl_num sign self_scope acc cf =
+let rec class_field_first_pass self_loc cl_num final sign self_scope acc cf =
   let { rev_fields; val_env; par_env; concrete_meths; concrete_vals;
         local_meths; local_vals; vars } = acc
   in
@@ -582,14 +603,15 @@
   | Pcf_inherit (override, sparent, super) ->
       with_attrs
         (fun () ->
+           let final = inheriting_final final in
            let parent =
-             class_expr cl_num val_env par_env
+             class_expr cl_num final val_env par_env
                Virtual self_scope sparent
            in
-           complete_class_type parent.cl_loc
-             par_env Virtual Class parent.cl_type;
-           inherit_class_type ~strict:true loc val_env sign parent.cl_type;
            let parent_sign = Btype.signature_of_class_type parent.cl_type in
+           update_class_signature parent.cl_loc ~warn_implicit_public:false
+             par_env Virtual Class parent_sign;
+           inherit_class_type ~strict:true loc val_env sign parent.cl_type;
            let new_concrete_meths = Btype.concrete_methods parent_sign in
            let new_concrete_vals = Btype.concrete_instance_vars parent_sign in
            let over_meths = MethSet.inter new_concrete_meths concrete_meths in
@@ -611,7 +633,7 @@
                       (cname :: VarSet.elements over_vals));
            | Override ->
                if MethSet.is_empty over_meths && VarSet.is_empty over_vals then
-                 raise (Error(loc, val_env, No_overriding ("","")))
+                 Error.log_and_raise loc val_env (No_overriding ("", ""))
            end;
            let concrete_vals = VarSet.union new_concrete_vals concrete_vals in
            let concrete_meths =
@@ -682,16 +704,16 @@
       with_attrs
         (fun () ->
            if VarSet.mem label.txt local_vals then
-             raise(Error(loc, val_env,
-                         Duplicate ("instance variable", label.txt)));
+             Error.log_and_raise loc val_env
+               (Duplicate ("instance variable", label.txt));
            if VarSet.mem label.txt concrete_vals then begin
              if override = Fresh then
                Location.prerr_warning label.loc
                  (Warnings.Instance_variable_override[label.txt])
            end else begin
              if override = Override then
-               raise(Error(loc, val_env,
-                           No_overriding ("instance variable", label.txt)))
+               Error.log_and_raise loc val_env
+                 (No_overriding ("instance variable", label.txt))
            end;
            let definition =
              Ctype.with_local_level_generalize_structure_if_principal
@@ -738,7 +760,7 @@
       with_attrs
         (fun () ->
            if MethSet.mem label.txt local_meths then
-             raise(Error(loc, val_env, Duplicate ("method", label.txt)));
+             Error.log_and_raise loc val_env (Duplicate ("method", label.txt));
            if MethSet.mem label.txt concrete_meths then begin
              if override = Fresh then begin
                  Location.prerr_warning loc
@@ -746,7 +768,8 @@
              end
            end else begin
              if override = Override then begin
-               raise(Error(loc, val_env, No_overriding("method", label.txt)))
+               Error.log_and_raise loc val_env
+                 (No_overriding("method", label.txt))
              end
            end;
            let expr =
@@ -782,8 +805,8 @@
                    type_approx val_env sbody ty1'
                | _ -> assert false
              with Ctype.Unify err ->
-               raise(Error(loc, val_env,
-                           Field_type_mismatch ("method", label.txt, err)))
+               Error.log_and_raise loc val_env
+                 (Field_type_mismatch ("method", label.txt, err))
            end;
            let sdefinition = make_method self_loc cl_num expr in
            let warning_state = Warnings.backup () in
@@ -825,7 +848,7 @@
   | Pcf_extension ext ->
       raise (Error_forward (Builtin_attributes.error_of_extension ext))
 
-and class_fields_first_pass self_loc cl_num sign self_scope
+and class_fields_first_pass self_loc cl_num final sign self_scope
       val_env par_env cfs =
   let rev_fields = [] in
   let concrete_meths = MethSet.empty in
@@ -842,7 +865,7 @@
     Builtin_attributes.warning_scope []
       (fun () ->
         List.fold_left
-          (class_field_first_pass self_loc cl_num sign self_scope)
+          (class_field_first_pass self_loc cl_num final sign self_scope)
           init_acc cfs)
   in
   List.rev acc.rev_fields, acc.vars
@@ -972,11 +995,9 @@
   let sign = Ctype.new_class_signature () in
 
   (* Adding a dummy method to the signature prevents it from being closed /
-     escaping. That isn't needed for objects though. *)
-  begin match final with
-  | Not_final -> Ctype.add_dummy_method val_env ~scope:self_scope sign;
-  | Final -> ()
-  end;
+     escaping. *)
+  add_dummy_method val_env final ~scope:self_scope sign;
+
 
   (* Self binder *)
   let (self_pat, self_pat_vars) = type_self_pattern val_env spat in
@@ -993,13 +1014,13 @@
   (* Check that the binder has a correct type *)
   begin try Ctype.unify val_env self_pat.pat_type sign.csig_self with
     Ctype.Unify _ ->
-      raise(Error(spat.ppat_loc, val_env,
-        Pattern_type_clash self_pat.pat_type))
+      Error.log_and_raise spat.ppat_loc val_env
+        (Pattern_type_clash self_pat.pat_type)
   end;
 
   (* Typing of class fields *)
   let (fields, vars) =
-    class_fields_first_pass self_loc cl_num sign self_scope
+    class_fields_first_pass self_loc cl_num final sign self_scope
            val_env par_env str
   in
   let kind = kind_of_final final in
@@ -1018,12 +1039,12 @@
       sign.csig_meths Meths.empty
   in
 
-  (* Close the signature if it is final *)
+  (* Close the signature if it is definitely final *)
   begin match final with
-  | Not_final -> ()
-  | Final ->
+  | Definitely_not_final | Maybe_final -> ()
+  | Definitely_final ->
       if not (Ctype.close_class_signature val_env sign) then
-        raise(Error(loc, val_env, Closing_self_type sign));
+        Error.log_and_raise loc val_env (Closing_self_type sign);
   end;
   (* Typing of method bodies *)
   Ctype.generalize_class_signature_spine sign;
@@ -1043,7 +1064,7 @@
     class_fields_second_pass cl_num sign met_env fields
   in
 
-  (* Update the class signature and warn about public methods made private *)
+  (* Update the class signature and warn about private methods made public *)
   update_class_signature loc val_env
     ~warn_implicit_public:true virt kind sign;
 
@@ -1057,16 +1078,16 @@
     cstr_type = sign;
     cstr_meths = meths; }
 
-and class_expr cl_num val_env met_env virt self_scope scl =
+and class_expr cl_num final val_env met_env virt self_scope scl =
   Builtin_attributes.warning_scope scl.pcl_attributes
-    (fun () -> class_expr_aux cl_num val_env met_env virt self_scope scl)
+    (fun () -> class_expr_aux cl_num final val_env met_env virt self_scope scl)
 
-and class_expr_aux cl_num val_env met_env virt self_scope scl =
+and class_expr_aux cl_num final val_env met_env virt self_scope scl =
   match scl.pcl_desc with
   | Pcl_constr (lid, styl) ->
       let (path, decl) = Env.lookup_class ~loc:scl.pcl_loc lid.txt val_env in
       if Path.same decl.cty_path unbound_class then
-        raise(Error(scl.pcl_loc, val_env, Unbound_class_2 lid.txt));
+        Error.log_and_raise scl.pcl_loc val_env (Unbound_class_2 lid.txt);
       let tyl = List.map
           (fun sty -> transl_simple_type val_env ~closed:false sty)
           styl
@@ -1075,19 +1096,21 @@
         Ctype.instance_class decl.cty_params decl.cty_type
       in
       let clty' = Btype.abbreviate_class_type path params clty in
+      let sign = Btype.signature_of_class_type clty' in
       (* Adding a dummy method to the self type prevents it from being closed /
          escaping. *)
-      Ctype.add_dummy_method val_env ~scope:self_scope
-        (Btype.signature_of_class_type clty');
+      add_dummy_method val_env final ~scope:self_scope sign;
+      Ctype.reveal_private_methods val_env sign;
+      Ctype.set_object_name decl.cty_path params sign.csig_self;
       if List.length params <> List.length tyl then
-        raise(Error(scl.pcl_loc, val_env,
-                    Parameter_arity_mismatch (lid.txt, List.length params,
-                                                   List.length tyl)));
+        Error.log_and_raise scl.pcl_loc val_env
+          (Parameter_arity_mismatch
+             (lid.txt, List.length params, List.length tyl));
       List.iter2
         (fun cty' ty ->
           let ty' = cty'.ctyp_type in
-           try Ctype.unify val_env ty' ty with Ctype.Unify err ->
-             raise(Error(cty'.ctyp_loc, val_env, Parameter_mismatch err)))
+          try Ctype.unify val_env ty' ty with Ctype.Unify err ->
+            Error.log_and_raise cty'.ctyp_loc val_env (Parameter_mismatch err))
         tyl params;
       (* Check for unexpected virtual methods *)
       check_virtual_clty scl.pcl_loc val_env virt Class clty';
@@ -1108,7 +1131,7 @@
          }
   | Pcl_structure cl_str ->
       let desc =
-        class_structure cl_num virt self_scope Not_final
+        class_structure cl_num virt self_scope final
           val_env met_env scl.pcl_loc cl_str
       in
       rc {cl_desc = Tcl_structure desc;
@@ -1119,7 +1142,7 @@
          }
   | Pcl_fun (l, Some default, spat, sbody) ->
       if has_poly_constraint spat then
-        raise(Error(spat.ppat_loc, val_env, Polymorphic_class_parameter));
+        Error.log_and_raise spat.ppat_loc val_env Polymorphic_class_parameter;
       let loc = default.pexp_loc in
       let open Ast_helper in
       let scases = [
@@ -1151,10 +1174,10 @@
           (* Note: we don't put the '#default' attribute, as it
              is not detected for class-level let bindings.  See #5975.*)
       in
-      class_expr cl_num val_env met_env virt self_scope sfun
+      class_expr cl_num final val_env met_env virt self_scope sfun
   | Pcl_fun (l, None, spat, scl') ->
       if has_poly_constraint spat then
-        raise(Error(spat.ppat_loc, val_env, Polymorphic_class_parameter));
+        Error.log_and_raise spat.ppat_loc val_env Polymorphic_class_parameter;
       let (pat, pv, val_env', met_env) =
         Ctype.with_local_level_generalize_structure_if_principal
           (fun () ->
@@ -1189,7 +1212,9 @@
       in
       let cl =
         Ctype.with_raised_nongen_level
-          (fun () -> class_expr cl_num val_env' met_env virt self_scope scl') in
+          (fun () ->
+             class_expr cl_num final val_env' met_env virt self_scope scl')
+      in
       if Btype.is_optional l && not_nolabel_function cl.cl_type then
         Location.prerr_warning pat.pat_loc
           Warnings.Unerasable_optional_argument;
@@ -1203,8 +1228,9 @@
   | Pcl_apply (scl', sargs) ->
       assert (sargs <> []);
       let cl =
-        Ctype.with_local_level_generalize_structure_if_principal
-          (fun () -> class_expr cl_num val_env met_env virt self_scope scl')
+        Ctype.with_local_level_generalize_structure_if_principal (fun () ->
+            class_expr cl_num final val_env met_env virt self_scope scl'
+          )
       in
       let rec nonopt_labels ls ty_fun =
         match ty_fun with
@@ -1264,7 +1290,8 @@
                     then
                       (sargs, eliminate_optional_arg ())
                     else
-                      raise(Error(sarg.pexp_loc, val_env, Apply_wrong_label l'))
+                      Error.log_and_raise sarg.pexp_loc val_env
+                        (Apply_wrong_label l')
               end else
                 match Btype.extract_label name sargs with
                 | Some (l', sarg, _, remaining_sargs) ->
@@ -1290,9 +1317,11 @@
             match sargs with
               (l, sarg0)::_ ->
                 if omitted <> [] then
-                  raise(Error(sarg0.pexp_loc, val_env, Apply_wrong_label l))
+                  Error.log_and_raise sarg0.pexp_loc val_env
+                    (Apply_wrong_label l)
                 else
-                  raise(Error(cl.cl_loc, val_env, Cannot_apply cl.cl_type))
+                  Error.log_and_raise cl.cl_loc val_env
+                    (Cannot_apply cl.cl_type)
             | [] ->
                 (List.rev args,
                  List.fold_left
@@ -1347,7 +1376,7 @@
           (let_bound_idents_full defs)
           ([], met_env)
       in
-      let cl = class_expr cl_num val_env met_env virt self_scope scl' in
+      let cl = class_expr cl_num final val_env met_env virt self_scope scl' in
       let defs = match rec_flag with
         | Recursive -> annotate_recursive_bindings val_env defs
         | Nonrecursive -> defs
@@ -1364,15 +1393,25 @@
           let self_scope = Ctype.get_current_level () in
           let cl =
             Typetexp.TyVarEnv.with_local_scope begin fun () ->
-              let cl = class_expr cl_num val_env met_env virt self_scope scl' in
-              complete_class_type cl.cl_loc val_env virt Class_type cl.cl_type;
+              let cl =
+                class_expr cl_num Definitely_not_final
+                  val_env met_env virt self_scope scl'
+              in
+              let sign = Btype.signature_of_class_type cl.cl_type in
+              update_class_signature cl.cl_loc val_env
+                ~warn_implicit_public:false virt Class sign;
+              Ctype.remove_dummy_method sign;
+              Ctype.hide_private_methods sign;
               cl
             end
           and clty =
             Typetexp.TyVarEnv.with_local_scope begin fun () ->
               let clty = class_type val_env virt self_scope scty in
-              complete_class_type
-                clty.cltyp_loc val_env virt Class clty.cltyp_type;
+              let sign = Btype.signature_of_class_type clty.cltyp_type in
+              update_class_signature clty.cltyp_loc val_env
+                ~warn_implicit_public:false virt Class_type sign;
+              Ctype.remove_dummy_method sign;
+              Ctype.hide_private_methods sign;
               clty
             end
           in
@@ -1389,14 +1428,14 @@
         Includeclass.class_types val_env cl.cl_type clty.cltyp_type
       with
         []    -> ()
-      | error -> raise(Error(cl.cl_loc, val_env, Class_match_failure error))
+      | error ->
+          Error.log_and_raise cl.cl_loc val_env (Class_match_failure error)
       end;
       let (vals, meths, concrs) = extract_constraints clty.cltyp_type in
       let ty = snd (Ctype.instance_class [] clty.cltyp_type) in
-      (* Adding a dummy method to the self type prevents it from being closed /
-         escaping. *)
-      Ctype.add_dummy_method val_env ~scope:self_scope
-        (Btype.signature_of_class_type ty);
+      let sign = Btype.signature_of_class_type ty in
+      add_dummy_method val_env final ~scope:self_scope sign;
+      Ctype.reveal_private_methods val_env sign;
       rc {cl_desc = Tcl_constraint (cl, Some clty, vals, meths, concrs);
           cl_loc = scl.pcl_loc;
           cl_type = ty;
@@ -1407,7 +1446,9 @@
       let used_slot = ref false in
       let (od, new_val_env) = !type_open_descr ~used_slot val_env pod in
       let ( _, new_met_env) = !type_open_descr ~used_slot met_env pod in
-      let cl = class_expr cl_num new_val_env new_met_env virt self_scope e in
+      let cl =
+        class_expr cl_num final new_val_env new_met_env virt self_scope e
+      in
       rc {cl_desc = Tcl_open (od, cl);
           cl_loc = scl.pcl_loc;
           cl_type = cl.cl_type;
@@ -1546,7 +1587,7 @@
           try
             (transl_type_param env sty, v)
           with Already_bound ->
-            raise(Error(sty.ptyp_loc, env, Repeated_parameter))
+            Error.log_and_raise sty.ptyp_loc env Repeated_parameter
         in
         List.map make_param cl.pci_params
       in
@@ -1586,19 +1627,20 @@
     begin try
       List.iter2 (Ctype.unify env) obj_params obj_params'
     with Ctype.Unify _ ->
-      raise(Error(cl.pci_loc, env,
-            Bad_parameters (obj_id, obj_params, obj_params')))
+      Error.log_and_raise cl.pci_loc env
+        (Bad_parameters (obj_id, obj_params, obj_params'))
     end;
     let ty = Btype.self_type obj_type in
     begin try
       Ctype.unify env ty constr
     with Ctype.Unify _ ->
-      raise(Error(cl.pci_loc, env,
-        Abbrev_type_clash (constr, ty, Ctype.expand_head env constr)))
+      let constr = Ctype.newconstr (Path.Pident obj_id) obj_params in
+      Error.log_and_raise cl.pci_loc env
+        (Abbrev_type_clash (constr, ty, Ctype.expand_head env constr))
     end
   end;
 
-  Ctype.set_object_name obj_id params (Btype.self_type typ);
+  Ctype.set_object_name (Path.Pident obj_id) params (Btype.self_type typ);
 
   (* Check the other temporary abbreviation (#-type) *)
   begin
@@ -1607,14 +1649,15 @@
     begin try
       List.iter2 (Ctype.unify env) cl_params cl_params'
     with Ctype.Unify _ ->
-      raise(Error(cl.pci_loc, env,
-            Bad_class_type_parameters (ty_id, cl_params, cl_params')))
+      Error.log_and_raise cl.pci_loc env
+        (Bad_class_type_parameters (ty_id, cl_params, cl_params'))
     end;
     begin try
       Ctype.unify env ty cl_ty
     with Ctype.Unify _ ->
       let ty_expanded = Ctype.object_fields ty in
-      raise(Error(cl.pci_loc, env, Abbrev_type_clash (ty, ty_expanded, cl_ty)))
+      Error.log_and_raise cl.pci_loc env
+        (Abbrev_type_clash (ty, ty_expanded, cl_ty))
     end
   end;
 
@@ -1624,8 +1667,8 @@
       (constructor_type constr obj_type)
       (Ctype.instance constr_type)
   with Ctype.Unify err ->
-    raise(Error(cl.pci_loc, env,
-                Constructor_type_mismatch (cl.pci_name.txt, err)))
+    Error.log_and_raise cl.pci_loc env
+      (Constructor_type_mismatch (cl.pci_name.txt, err))
   end;
 
   (* Class and class type temporary definitions *)
@@ -1702,7 +1745,7 @@
   let (cl_params, cl_ty) =
     Ctype.instance_parameterized_type params (Btype.self_type typ)
   in
-  Ctype.set_object_name obj_id cl_params cl_ty;
+  Ctype.set_object_name (Path.Pident obj_id) cl_params cl_ty;
   let cl_abbr =
     { cl_td with
      type_params = cl_params;
@@ -1726,7 +1769,8 @@
 let collapse_conj_class_params env (cl, id, clty, _, _, _, _, _, _, _, _, _) =
   try Ctype.collapse_conj_params env clty.cty_params
   with Ctype.Unify err ->
-    raise(Error(cl.pci_loc, env, Non_collapsable_conjunction (id, clty, err)))
+    Error.log_and_raise cl.pci_loc env
+      (Non_collapsable_conjunction (id, clty, err))
 
 let final_decl env define_class
     (cl, id, clty, ty_id, cltydef, obj_id, obj_abbr, ci_params,
@@ -1734,8 +1778,8 @@
   Ctype.nongen_vars_in_class_declaration clty
   |> Option.iter (fun vars ->
       let nongen_vars = Btype.TypeSet.elements vars in
-      raise(Error(cl.pci_loc, env
-                 , Non_generalizable_class { id; clty; nongen_vars }));
+      Error.log_and_raise cl.pci_loc env
+        (Non_generalizable_class { id; clty; nongen_vars });
     );
   begin match
     Ctype.closed_class clty.cty_params
@@ -1751,7 +1795,7 @@
           Format_doc.doc_printf "%a"
             (Printtyp.Doc.cltype_declaration id) cltydef
       in
-      raise(Error(cl.pci_loc, env, Unbound_type_var(printer, reason)))
+      Error.log_and_raise cl.pci_loc env (Unbound_type_var(printer, reason))
   end;
   { id; clty; ty_id; cltydef; obj_id; obj_abbr; arity;
     pub_meths; coe;
@@ -1825,10 +1869,10 @@
       in
       begin try Ctype.subtype env cl_ty obj_ty ()
       with Ctype.Subtype err ->
-        raise(Typecore.Error(loc, env, Typecore.Not_subtype err))
+        Typecore.Error.log_and_raise loc env (Typecore.Not_subtype err)
       end;
       if not (Ctype.opened_object cl_ty) then
-        raise(Error(loc, env, Cannot_coerce_self obj_ty))
+        Error.log_and_raise loc env (Cannot_coerce_self obj_ty)
   end;
   {cls_id = id;
    cls_id_loc = id_loc;
@@ -1874,7 +1918,7 @@
   let decls =
     try Typedecl_variance.update_class_decls env decls
     with Typedecl_variance.Error(loc, err) ->
-      raise (Typedecl.Error(loc, Typedecl.Variance err))
+      Typedecl.Error.log_and_raise loc (Typedecl.Variance err)
   in
   let res = List.map2 merge_type_decls res decls in
   let env = List.fold_left (final_env define_class) env res in
@@ -1886,15 +1930,24 @@
   incr class_num;
   let self_scope = Ctype.get_current_level () in
   let expr =
-    class_expr (Int.to_string !class_num) env env virt self_scope sexpr
+    class_expr (Int.to_string !class_num) Definitely_not_final
+      env env virt self_scope sexpr
   in
-  complete_class_type expr.cl_loc env virt Class expr.cl_type;
+  let sign = Btype.signature_of_class_type expr.cl_type in
+  update_class_signature expr.cl_loc env
+    ~warn_implicit_public:false virt Class sign;
+  Ctype.remove_dummy_method sign;
+  Ctype.hide_private_methods sign;
   (expr, expr.cl_type)
 
 let class_description env virt sexpr =
   let self_scope = Ctype.get_current_level () in
   let expr = class_type env virt self_scope sexpr in
-  complete_class_type expr.cltyp_loc env virt Class_type expr.cltyp_type;
+  let sign = Btype.signature_of_class_type expr.cltyp_type in
+  update_class_signature expr.cltyp_loc env
+    ~warn_implicit_public:false virt Class_type sign;
+  Ctype.remove_dummy_method sign;
+  Ctype.hide_private_methods sign;
   (expr, expr.cltyp_type)
 
 let class_declarations env cls =
@@ -1933,9 +1986,9 @@
   incr class_num;
   let desc =
     class_structure (Int.to_string !class_num)
-      Concrete Btype.lowest_level Final env env loc s
+      Concrete Btype.lowest_level Definitely_final env env loc s
   in
-  complete_class_signature loc env Concrete Object desc.cstr_type;
+  Ctype.hide_private_methods desc.cstr_type;
   let meths = Btype.public_methods desc.cstr_type in
   (desc, meths)
 
@@ -1951,7 +2004,7 @@
       begin try
         ignore (Env.lookup_cltype ~use:false ~loc:lid.loc lid.txt env)
       with
-      | Env.Error
+      | Env.Error.In_context
           (Lookup_error
              (location, env,
               Illegal_reference_to_recursive_module { container; unbound; })) ->
@@ -2223,7 +2276,7 @@
 let () =
   Location.register_error_of_exn
     (function
-      | Error (loc, env, err) ->
+      | Error.In_context (loc, env, err) ->
         Some (Location.error_of_printer ~loc (report_error_doc env) err)
       | Error_forward err ->
         Some err
