--- ocaml_505/typing/typedtree.ml	2026-07-07 01:48:43.519538853 +0200
+++ ocaml_506_typing_recovery/typing/typedtree.ml	2026-07-07 02:21:17.014209942 +0200
@@ -306,7 +306,7 @@
 and structure_item_desc =
     Tstr_eval of expression * attributes
   | Tstr_value of rec_flag * value_binding list
-  | Tstr_primitive of value_description
+  | Tstr_primitive of primitive_description
   | Tstr_type of rec_flag * type_declaration list
   | Tstr_typext of type_extension
   | Tstr_exception of type_exception
@@ -385,6 +385,7 @@
 
 and signature_item_desc =
     Tsig_value of value_description
+  | Tsig_primitive of primitive_description
   | Tsig_type of rec_flag * type_declaration list
   | Tsig_typesubst of type_declaration list
   | Tsig_typext of type_extension
@@ -523,10 +524,22 @@
     val_name: string loc;
     val_desc: core_type;
     val_val: Types.value_description;
-    val_prim: string list;
     val_loc: Location.t;
-    val_attributes: attribute list;
-    }
+    val_attributes: attributes;
+  }
+
+and primitive_description =
+  { prim_id: Ident.t;
+    prim_name: string loc;
+    prim_kind: primitive_kind;
+    prim_val: Types.value_description;
+    prim_loc: Location.t;
+    prim_attributes: attributes;
+  }
+
+and primitive_kind =
+  | Tprim_decl of core_type * string list
+  | Tprim_alias of core_type option * Path.t * Longident.t loc
 
 and type_declaration =
   { typ_id: Ident.t;
@@ -674,6 +687,7 @@
 
 type item_declaration =
   | Value of value_description
+  | Primitive of primitive_description
   | Value_binding of value_binding
   | Type of type_declaration
   | Constructor of constructor_declaration
