--- ocaml_505/typing/untypeast.mli	2026-07-07 01:48:43.519538853 +0200
+++ ocaml_506_typing_recovery/typing/untypeast.mli	2026-07-07 02:21:17.029067660 +0200
@@ -72,6 +72,8 @@
   type_kind: mapper -> Typedtree.type_kind -> type_kind;
   value_binding: mapper -> Typedtree.value_binding -> value_binding;
   value_description: mapper -> Typedtree.value_description -> value_description;
+  primitive_description:
+    mapper -> Typedtree.primitive_description -> primitive_description;
   with_constraint:
     mapper -> (Path.t * Longident.t Location.loc * Typedtree.with_constraint)
     -> with_constraint;
