--- ocaml_505/typing/value_rec_check.ml	2026-07-07 01:48:43.519538853 +0200
+++ ocaml_506_typing_recovery/typing/value_rec_check.ml	2026-07-07 02:21:17.030526829 +0200
@@ -202,20 +202,15 @@
     | Texp_function _ ->
         Static
     | Texp_lazy e ->
-      (* The code below was copied (in part) from translcore.ml *)
       begin match Typeopt.classify_lazy_argument e with
-      | `Constant_or_function ->
-        (* A constant expr (of type <> float if [Config.flat_float_array] is
-           true) gets compiled as itself. *)
+      | Eager Shortcut ->
+          (* compiled to [e] directly. *)
           classify_expression env e
-      | `Float_that_cannot_be_shortcut
-      | `Identifier `Forward_value ->
-          (* Forward blocks *)
+      | Eager Forward ->
+          (* [e] is placed inside a Forward block *)
           Static
-      | `Identifier `Other ->
-          classify_expression env e
-      | `Other ->
-          (* other cases compile to a lazy block holding a function *)
+      | Lazy_thunk ->
+          (* [e] is placed inside a Lazy thunk. *)
           Static
       end
 
@@ -440,7 +435,7 @@
 
   val equal : t -> t -> bool
 end = struct
-  module M = Map.Make(Ident)
+  module M = Ident.Map
 
   (** A "t" maps each rec-bound variable to an access status *)
   type t = Mode.t M.t
@@ -453,10 +448,8 @@
   let empty = M.empty
 
   let join (x: t) (y: t) =
-    M.fold
-      (fun (id: Ident.t) (v: Mode.t) (tbl: t) ->
-         let v' = find id tbl in
-         M.add id (Mode.join v v') tbl)
+    M.union
+      (fun _id v1 v2 -> Some (Mode.join v1 v2))
       x y
 
   let join_list li = List.fold_left join empty li
@@ -899,11 +892,12 @@
         G |- lazy e: m
       *)
       let lazy_mode = match Typeopt.classify_lazy_argument e with
-        | `Constant_or_function
-        | `Identifier _
-        | `Float_that_cannot_be_shortcut ->
+        | Eager _ ->
+          (* We pessimize [Eager Forward] into [Return] instead of [Guarded],
+             so that this check remains robust to adding more shortcutting
+             in the future. *)
           Return
-        | `Other ->
+        | Lazy_thunk ->
           Delay
       in
       expression e << lazy_mode
