--- ocaml_502/typing/cmt2annot.ml	2024-06-27 15:42:08.724127244 +0200
+++ ocaml_503/typing/cmt2annot.ml	2024-09-17 01:15:58.292566923 +0200
@@ -100,10 +100,12 @@
         bind_bindings exp.exp_loc bindings
     | Texp_let (Nonrecursive, bindings, body) ->
         bind_bindings body.exp_loc bindings
-    | Texp_match (_, f1, _) ->
-        bind_cases f1
-    | Texp_try (_, f) ->
-        bind_cases f
+    | Texp_match (_, f1, f2, _) ->
+        bind_cases f1;
+        bind_cases f2
+    | Texp_try (_, f1, f2) ->
+        bind_cases f1;
+        bind_cases f2
     | Texp_function (params, _) ->
         List.iter (bind_function_param exp.exp_loc) params
     | Texp_letmodule (_, modname, _, _, body ) ->
