--- ocaml_502/typing/envaux.mli	2024-06-27 15:42:08.727460578 +0200
+++ ocaml_503/typing/envaux.mli	2024-09-17 01:15:58.292566923 +0200
@@ -14,8 +14,6 @@
 (*                                                                        *)
 (**************************************************************************)
 
-open Format
-
 (* Convert environment summaries to environments *)
 
 val env_from_summary : Env.summary -> Subst.t -> Env.t
@@ -33,4 +31,5 @@
 
 exception Error of error
 
-val report_error: formatter -> error -> unit
+val report_error: error Format_doc.format_printer
+val report_error_doc: error Format_doc.printer
