--- ocaml_502/typing/includeclass.mli	2024-06-27 15:42:08.727460578 +0200
+++ ocaml_503/typing/includeclass.mli	2024-09-17 01:15:58.292566923 +0200
@@ -17,7 +17,6 @@
 
 open Types
 open Ctype
-open Format
 
 val class_types:
         Env.t -> class_type -> class_type -> class_match_failure list
@@ -30,4 +29,6 @@
   class_match_failure list
 
 val report_error :
-  Printtyp.type_or_scheme -> formatter -> class_match_failure list -> unit
+  Out_type.type_or_scheme -> class_match_failure list Format_doc.format_printer
+val report_error_doc :
+  Out_type.type_or_scheme -> class_match_failure list Format_doc.printer
