--- ocaml_502/typing/predef.mli	2024-06-27 15:42:08.727460578 +0200
+++ ocaml_503/typing/predef.mli	2024-09-17 01:15:58.292566923 +0200
@@ -25,6 +25,8 @@
 val type_bool: type_expr
 val type_unit: type_expr
 val type_exn: type_expr
+val type_eff: type_expr -> type_expr
+val type_continuation: type_expr -> type_expr -> type_expr
 val type_array: type_expr -> type_expr
 val type_list: type_expr -> type_expr
 val type_option: type_expr -> type_expr
@@ -43,6 +45,7 @@
 val path_bool: Path.t
 val path_unit: Path.t
 val path_exn: Path.t
+val path_eff: Path.t
 val path_array: Path.t
 val path_list: Path.t
 val path_option: Path.t
@@ -52,6 +55,7 @@
 val path_lazy_t: Path.t
 val path_extension_constructor: Path.t
 val path_floatarray: Path.t
+val path_continuation: Path.t
 
 val path_match_failure: Path.t
 val path_assert_failure : Path.t
