--- ocaml_502/typing/typemod.mli	2024-06-27 15:42:08.730793912 +0200
+++ ocaml_503/typing/typemod.mli	2024-09-17 01:15:58.295900254 +0200
@@ -43,8 +43,6 @@
   Typedtree.implementation
 val type_interface:
         Env.t -> Parsetree.signature -> Typedtree.signature
-val transl_signature:
-        Env.t -> Parsetree.signature -> Typedtree.signature
 val check_nongen_signature:
         Env.t -> Types.signature -> unit
         (*
@@ -137,6 +135,7 @@
   | Invalid_type_subst_rhs
   | Unpackable_local_modtype_subst of Path.t
   | With_cannot_remove_packed_modtype of Path.t * module_type
+  | Cannot_alias of Path.t
 
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
