--- ocaml_502/typing/typeopt.ml	2024-06-27 15:42:08.730793912 +0200
+++ ocaml_503/typing/typeopt.ml	2024-09-17 01:15:58.295900254 +0200
@@ -24,7 +24,7 @@
 let scrape_ty env ty =
   match get_desc ty with
   | Tconstr _ ->
-      let ty = Ctype.expand_head_opt env (Ctype.correct_levels ty) in
+      let ty = Ctype.expand_head_opt env ty in
       begin match get_desc ty with
       | Tconstr (p, _, _) ->
           begin match Env.find_type p env with
