--- ocaml_502/typing/typetexp.mli	2024-06-27 15:42:08.730793912 +0200
+++ ocaml_503/typing/typetexp.mli	2024-09-17 01:15:58.295900254 +0200
@@ -95,7 +95,8 @@
 
 exception Error of Location.t * Env.t * error
 
-val report_error: Env.t -> Format.formatter -> error -> unit
+val report_error: Env.t -> error Format_doc.format_printer
+val report_error_doc: Env.t -> error Format_doc.printer
 
 (* Support for first-class modules. *)
 val transl_modtype_longident:  (* from Typemod *)
