--- ocaml_504/file_formats/cmt_format.mli	2025-11-07 15:50:07
+++ ocaml_505/file_formats/cmt_format.mli	2026-02-03 16:48:51
@@ -50,11 +50,10 @@
   | Partial_signature_item of signature_item
   | Partial_module_type of module_type
 
-type dependency_kind = Definition_to_declaration | Declaration_to_declaration
 type cmt_infos = {
   cmt_modname : modname;
   cmt_annots : binary_annots;
-  cmt_declaration_dependencies : (dependency_kind * Uid.t * Uid.t) list;
+  cmt_declaration_dependencies : (Uid.Deps.kind * Uid.t * Uid.t) list;
   cmt_comments : (string * Location.t) list;
   cmt_args : string array;
     (** {!Sys.argv} from the compiler invocation which created the file.
@@ -65,7 +64,7 @@
   cmt_source_digest : string option;
   cmt_initial_env : Env.t;
   cmt_imports : crcs;
-  cmt_interface_digest : Digest.t option;
+  cmt_interface_digest : Digest.BLAKE128.t option;
   cmt_use_summaries : bool;
   cmt_uid_to_decl : item_declaration Shape.Uid.Tbl.t;
   cmt_impl_shape : Shape.t option; (* None for mli *)
@@ -111,8 +110,6 @@
 val get_saved_types : unit -> binary_part list
 val set_saved_types : binary_part list -> unit
 
-val record_declaration_dependency: dependency_kind * Uid.t * Uid.t -> unit
-
 (*
 
   val is_magic_number : string -> bool
