--- ocaml_502/utils/config.fixed.ml	2024-06-27 15:42:08.730793912 +0200
+++ ocaml_503/utils/config.fixed.ml	2024-09-17 01:19:03.395759607 +0200
@@ -27,10 +27,10 @@
 let c_output_obj = ""
 let c_has_debug_prefix_map = false
 let as_has_debug_prefix_map = false
-let ocamlc_cflags = ""
-let ocamlc_cppflags = ""
-let ocamlopt_cflags = ""
-let ocamlopt_cppflags = ""
+let bytecode_cflags = ""
+let bytecode_cppflags = ""
+let native_cflags = ""
+let native_cppflags = ""
 let bytecomp_c_libraries = ""
 let bytecomp_c_compiler = ""
 let native_c_compiler = c_compiler
