--- ocaml_502/utils/warnings.mli	2024-06-27 15:42:08.730793912 +0200
+++ ocaml_503/utils/warnings.mli	2024-09-17 01:19:03.395759607 +0200
@@ -57,7 +57,7 @@
   | Implicit_public_methods of string list  (* 15 *)
   | Unerasable_optional_argument            (* 16 *)
   | Undeclared_virtual_method of string     (* 17 *)
-  | Not_principal of string                 (* 18 *)
+  | Not_principal of Format_doc.t           (* 18 *)
   | Non_principal_labels of string          (* 19 *)
   | Ignored_extra_argument                  (* 20 *)
   | Nonreturning_statement                  (* 21 *)
@@ -116,6 +116,7 @@
   | Unused_tmc_attribute                    (* 71 *)
   | Tmc_breaks_tailcall                     (* 72 *)
   | Generative_application_expects_unit     (* 73 *)
+  | Degraded_to_partial_match               (* 74 *)
 
 type alert = {kind:string; message:string; def:loc; use:loc}
 
