--- ocaml_505/parsing/ast_helper.ml	2026-07-07 01:48:43.508960791 +0200
+++ ocaml_506_typing_recovery/parsing/ast_helper.ml	2026-07-07 02:21:01.008609623 +0200
@@ -283,6 +283,7 @@
   let mk ?(loc = !default_loc) d = {psig_desc = d; psig_loc = loc}
 
   let value ?loc a = mk ?loc (Psig_value a)
+  let primitive ?loc a = mk ?loc (Psig_primitive a)
   let type_ ?loc rec_flag a = mk ?loc (Psig_type (rec_flag, a))
   let type_subst ?loc a = mk ?loc (Psig_typesubst a)
   let type_extension ?loc a = mk ?loc (Psig_typext a)
@@ -310,6 +311,7 @@
 
   let eval ?loc ?(attrs = []) a = mk ?loc (Pstr_eval (a, attrs))
   let value ?loc a b = mk ?loc (Pstr_value (a, b))
+  let val_ ?loc a = mk ?loc (Pstr_val a)
   let primitive ?loc a = mk ?loc (Pstr_primitive a)
   let type_ ?loc rec_flag a = mk ?loc (Pstr_type (rec_flag, a))
   let type_extension ?loc a = mk ?loc (Pstr_typext a)
@@ -421,13 +423,32 @@
 
 module Val = struct
   let mk ?(loc = !default_loc) ?(attrs = []) ?(docs = empty_docs)
-        ?(prim = []) name typ =
+         name typ =
     {
      pval_name = name;
      pval_type = typ;
      pval_attributes = add_docs_attrs docs attrs;
      pval_loc = loc;
-     pval_prim = prim;
+    }
+end
+
+module Prim = struct
+  let mk_decl ?(loc = !default_loc) ?(attrs = []) ?(docs = empty_docs)
+         ~prim name typ =
+    {
+     pprim_name = name;
+     pprim_kind = Pprim_decl (typ, prim);
+     pprim_attributes = add_docs_attrs docs attrs;
+     pprim_loc = loc;
+    }
+
+  let mk_alias ?(loc = !default_loc) ?(attrs = []) ?(docs = empty_docs)
+         name typ id =
+    {
+     pprim_name = name;
+     pprim_kind = Pprim_alias (typ, id);
+     pprim_attributes = add_docs_attrs docs attrs;
+     pprim_loc = loc;
     }
 end
 
