--- ocaml_505/parsing/ast_mapper.ml	2026-07-07 01:48:43.509116732 +0200
+++ ocaml_506_typing_recovery/parsing/ast_mapper.ml	2026-07-07 02:21:01.013665561 +0200
@@ -69,6 +69,8 @@
   package_type: mapper -> package_type -> package_type;
   pat: mapper -> pattern -> pattern;
   payload: mapper -> payload -> payload;
+  primitive_description: mapper -> primitive_description
+                         -> primitive_description;
   signature: mapper -> signature -> signature;
   signature_item: mapper -> signature_item -> signature_item;
   structure: mapper -> structure -> structure;
@@ -365,6 +367,7 @@
     let loc = sub.location sub loc in
     match desc with
     | Psig_value vd -> value ~loc (sub.value_description sub vd)
+    | Psig_primitive pd -> primitive ~loc (sub.primitive_description sub pd)
     | Psig_type (rf, l) ->
         type_ ~loc rf (List.map (sub.type_declaration sub) l)
     | Psig_typesubst l ->
@@ -422,7 +425,8 @@
         let attrs = sub.attributes sub attrs in
         eval ~loc ~attrs (sub.expr sub x)
     | Pstr_value (r, vbs) -> value ~loc r (List.map (sub.value_binding sub) vbs)
-    | Pstr_primitive vd -> primitive ~loc (sub.value_description sub vd)
+    | Pstr_val vd -> val_ ~loc (sub.value_description sub vd)
+    | Pstr_primitive pd -> primitive ~loc (sub.primitive_description sub pd)
     | Pstr_type (rf, l) -> type_ ~loc rf (List.map (sub.type_declaration sub) l)
     | Pstr_typext te -> type_extension ~loc (sub.type_extension sub te)
     | Pstr_exception ed -> exception_ ~loc (sub.type_exception sub ed)
@@ -706,14 +710,30 @@
     extension_constructor = T.map_extension_constructor;
     package_type = T.map_package_type;
     value_description =
-      (fun this {pval_name; pval_type; pval_prim; pval_loc;
-                 pval_attributes} ->
+      (fun this {pval_name; pval_type; pval_loc; pval_attributes} ->
         Val.mk
           (map_loc map_string this pval_name)
           (this.typ this pval_type)
           ~attrs:(this.attributes this pval_attributes)
           ~loc:(this.location this pval_loc)
-          ~prim:pval_prim
+      );
+    primitive_description =
+      (fun this {pprim_name; pprim_kind; pprim_loc; pprim_attributes} ->
+         match pprim_kind with
+         | Pprim_decl (pprim_type, pprim_prim) ->
+           Prim.mk_decl
+             (map_loc map_string this pprim_name)
+             (this.typ this pprim_type)
+             ~attrs:(this.attributes this pprim_attributes)
+             ~loc:(this.location this pprim_loc)
+             ~prim:pprim_prim
+         | Pprim_alias (pprim_type, pprim_ident) ->
+           Prim.mk_alias
+             (map_loc map_string this pprim_name)
+             (Option.map (this.typ this) pprim_type)
+             (map_loc_lid this pprim_ident)
+             ~attrs:(this.attributes this pprim_attributes)
+             ~loc:(this.location this pprim_loc)
       );
 
     pat = P.map;
