--- ocaml_505/parsing/depend.ml	2026-07-07 01:48:43.509116732 +0200
+++ ocaml_506_typing_recovery/parsing/depend.ml	2026-07-07 02:21:01.022762554 +0200
@@ -385,6 +385,13 @@
   match item.psig_desc with
     Psig_value vd ->
       add_type bv vd.pval_type; (bv, m)
+  | Psig_primitive pd ->
+      (match pd.pprim_kind with
+       | Pprim_decl (pprim_type, _) -> add_type bv pprim_type
+       | Pprim_alias (pprim_type, pprim_ident) ->
+         Option.iter (add_type bv) pprim_type;
+         add bv pprim_ident);
+      (bv, m)
   | Psig_type (_, dcls)
   | Psig_typesubst dcls->
       List.iter (add_type_declaration bv) dcls; (bv, m)
@@ -523,8 +530,15 @@
       add_expr bv e; (bv, m)
   | Pstr_value(rf, pel) ->
       let bv = add_bindings rf bv pel in (bv, m)
-  | Pstr_primitive vd ->
+  | Pstr_val vd ->
       add_type bv vd.pval_type; (bv, m)
+  | Pstr_primitive pd ->
+      (match pd.pprim_kind with
+       | Pprim_decl (pprim_type, _) -> add_type bv pprim_type
+       | Pprim_alias (pprim_type, pprim_ident) ->
+         Option.iter (add_type bv) pprim_type;
+         add bv pprim_ident);
+      (bv, m)
   | Pstr_type (_, dcls) ->
       List.iter (add_type_declaration bv) dcls; (bv, m)
   | Pstr_typext te ->
