--- ocaml_505/parsing/parsetree.mli	2026-07-07 01:48:43.511538860 +0200
+++ ocaml_506_typing_recovery/parsing/parsetree.mli	2026-07-07 02:21:01.057865822 +0200
@@ -113,7 +113,7 @@
            Invariant: [n >= 2].
         *)
   | Ptyp_constr of Longident.t loc * core_type list
-      (** [Ptyp_constr(lident, l)] represents:
+      (** [Ptyp_constr(tconstr, l)] represents:
             - [tconstr]               when [l=[]],
             - [T tconstr]             when [l=[T]],
             - [(T1, ..., Tn) tconstr] when [l=[T1 ; ... ; Tn]].
@@ -174,6 +174,10 @@
 
            - As the {{!value_description.pval_type}[pval_type]} field of a
            {!value_description}.
+
+           - As the {!core_type} in the
+           {{!primitive_description.pprim_kind}[pprim_kind]} field of a
+           {!primitive_description}.
          *)
   | Ptyp_package of package_type  (** [(module S)]. *)
   | Ptyp_open of Longident.t loc * core_type (** [M.(T)] *)
@@ -526,17 +530,35 @@
     {
      pval_name: string loc;
      pval_type: core_type;
-     pval_prim: string list;
      pval_attributes: attributes;  (** [... [\@\@id1] [\@\@id2]] *)
      pval_loc: Location.t;
     }
-(** Values of type {!value_description} represents:
-    - [val x: T],
-            when {{!value_description.pval_prim}[pval_prim]} is [[]]
-    - [external x: T = "s1" ... "sn"]
-            when {{!value_description.pval_prim}[pval_prim]} is [["s1";..."sn"]]
+(** Values of type {!value_description} represent [val x: T]. *)
+
+(** {2 Primitive descriptions} *)
+
+and primitive_description =
+  {
+    pprim_name: string loc;
+    pprim_kind: primitive_kind;
+    pprim_attributes: attributes;
+    pprim_loc: Location.t
+  }
+(** Values of type {!primitive_description} represent:
+  - [external x: T = "s1" ... "sn"] when
+    {{!primitive_description.pprim_kind}[pprim_kind]} is
+    [Pprim_decl (T, ["s1";..."sn"])].
+  - [external x: T = M.y] when {{!primitive_description.pprim_kind}[pprim_kind]}
+    is [Pprim_alias (Some T, M.y)]
+  - [external x = M.y] when {{!primitive_description.pprim_kind}[pprim_kind]}
+    is [Pprim_alias (None, M.y)]
 *)
 
+and primitive_kind =
+  | Pprim_decl of core_type * string list
+  | Pprim_alias of core_type option * Longident.t loc
+(** See the comment on {{!primitive_description}[primitive_description]}. *)
+
 (** {2 Type declarations} *)
 
 and type_declaration =
@@ -918,9 +940,11 @@
 
 and signature_item_desc =
   | Psig_value of value_description
-      (** - [val x: T]
-            - [external x: T = "s1" ... "sn"]
-         *)
+      (** [val x: T] *)
+  | Psig_primitive of primitive_description
+      (** - [external x: T = "s1" ... "sn" ]
+          - [external x = y]
+          - [external x: T = y] *)
   | Psig_type of rec_flag * type_declaration list
       (** [type t1 = ... and ... and tn  = ...] *)
   | Psig_typesubst of type_declaration list
@@ -1068,9 +1092,12 @@
             - [let rec P1 = E1 and ... and Pn = EN ]
                 when [rec] is {{!Asttypes.rec_flag.Recursive}[Recursive]}.
         *)
-  | Pstr_primitive of value_description
-      (** - [val x: T]
-            - [external x: T = "s1" ... "sn" ]*)
+  | Pstr_val of value_description
+      (** [val x: T] *)
+  | Pstr_primitive of primitive_description
+      (** - [external x: T = "s1" ... "sn" ]
+          - [external x = y]
+          - [external x: T = y] *)
   | Pstr_type of rec_flag * type_declaration list
       (** [type t1 = ... and ... and tn = ...] *)
   | Pstr_typext of type_extension  (** [type t1 += ...] *)
