--- ocaml_505/parsing/parser.mly	2026-07-07 01:48:43.510538861 +0200
+++ ocaml_506_typing_recovery/parsing/parser.mly	2026-07-07 02:21:01.055068551 +0200
@@ -66,8 +66,10 @@
   (Pstr_attribute body, None)
 let pstr_typext (te, ext) =
   (Pstr_typext te, ext)
-let pstr_primitive (vd, ext) =
-  (Pstr_primitive vd, ext)
+let pstr_val (vd, ext) =
+  (Pstr_val vd, ext)
+let pstr_primitive (pd, ext) =
+  (Pstr_primitive pd, ext)
 let pstr_type ((nr, ext), tys) =
   (Pstr_type (nr, tys), ext)
 let pstr_exception (te, ext) =
@@ -95,6 +97,8 @@
   (Psig_typext te, ext)
 let psig_value (vd, ext) =
   (Psig_value vd, ext)
+let psig_primitive (pd, ext) =
+  (Psig_primitive pd, ext)
 let psig_type ((nr, ext), tys) =
   (Psig_type (nr, tys), ext)
 let psig_typesubst ((nr, ext), tys) =
@@ -714,12 +718,17 @@
     | _ ->
         err pmty.pmty_loc Not_with_type
   in
-  match pmty with
-  | {pmty_desc = Pmty_ident lid} -> (lid, [], pmty.pmty_attributes)
-  | {pmty_desc = Pmty_with({pmty_desc = Pmty_ident lid}, cstrs)} ->
-      (lid, List.map map_cstr cstrs, pmty.pmty_attributes)
-  | _ ->
-      err pmty.pmty_loc Neither_identifier_nor_with_type
+  let rec flatten attributes cstrs desc =
+    match desc with
+    | Pmty_ident lid -> (lid, cstrs, attributes)
+    | Pmty_with (pmty, more_cstrs) ->
+       let attributes = pmty.pmty_attributes @ attributes in
+       let cstrs = List.map map_cstr more_cstrs @ cstrs in
+       flatten attributes cstrs pmty.pmty_desc
+    | _ ->
+       err pmty.pmty_loc Neither_identifier_nor_with_type
+  in
+  flatten pmty.pmty_attributes [] pmty.pmty_desc
 
 let mk_directive_arg ~loc k =
   { pdira_desc = k;
@@ -1556,10 +1565,10 @@
         { pstr_extension $1 (add_docs_attrs (symbol_docs $sloc) $2) }
     | floating_attribute
         { pstr_attribute $1 }
-    | primitive_declaration
+    | primitive_description
         { pstr_primitive $1 }
     | value_description
-        { pstr_primitive $1 }
+        { pstr_val $1 }
     | type_declarations
         { pstr_type $1 }
     | str_type_extension
@@ -1796,8 +1805,8 @@
         { psig_attribute $1 }
     | value_description
         { psig_value $1 }
-    | primitive_declaration
-        { psig_value $1 }
+    | primitive_description
+        { psig_primitive $1 }
     | type_declarations
         { psig_type $1 }
     | type_subst_declarations
@@ -2935,8 +2944,14 @@
   xs = rev(reversed_labeled_tuple_body)
     { xs }
 ;
+record_update_expr:
+| simple_expr nonempty_llist(labeled_simple_expr)
+    { mkexp ~loc:$sloc (Pexp_apply ($1, $2)) }
+| simple_expr
+    { $1 }
+;
 record_expr_content:
-  eo = ioption(terminated(simple_expr, WITH))
+  eo = ioption(terminated(record_update_expr, WITH))
   fields = separated_or_terminated_nonempty_list(SEMI, record_expr_field)
     { eo, fields }
 ;
@@ -3246,23 +3261,49 @@
       ext }
 ;
 
-/* Primitive declarations */
+/* Primitive descriptions */
 
-primitive_declaration:
-  EXTERNAL
-  ext = ext
-  attrs1 = attributes
-  id = mkrhs(val_ident)
-  COLON
-  ty = possibly_poly(core_type)
-  EQUAL
-  prim = raw_string+
-  attrs2 = post_item_attributes
-    { let attrs = attrs1 @ attrs2 in
-      let loc = make_loc $sloc in
-      let docs = symbol_docs $sloc in
-      Val.mk id ty ~prim ~attrs ~loc ~docs,
-      ext }
+primitive_description:
+  | EXTERNAL
+    ext = ext
+    attrs1 = attributes
+    id = mkrhs(val_ident)
+    COLON
+    ty = possibly_poly(core_type)
+    EQUAL
+    prim = raw_string+
+    attrs2 = post_item_attributes
+      { let attrs = attrs1 @ attrs2 in
+        let loc = make_loc $sloc in
+        let docs = symbol_docs $sloc in
+        Prim.mk_decl id ty ~prim ~attrs ~loc ~docs,
+        ext }
+  | EXTERNAL
+    ext = ext
+    attrs1 = attributes
+    id = mkrhs(val_ident)
+    EQUAL
+    lid = mkrhs(val_longident)
+    attrs2 = post_item_attributes
+      { let attrs = attrs1 @ attrs2 in
+        let loc = make_loc $sloc in
+        let docs = symbol_docs $sloc in
+        Prim.mk_alias id None lid ~attrs ~loc ~docs,
+        ext }
+  | EXTERNAL
+    ext = ext
+    attrs1 = attributes
+    id = mkrhs(val_ident)
+    COLON
+    ty = possibly_poly(core_type)
+    EQUAL
+    lid = mkrhs(val_longident)
+    attrs2 = post_item_attributes
+      { let attrs = attrs1 @ attrs2 in
+        let loc = make_loc $sloc in
+        let docs = symbol_docs $sloc in
+        Prim.mk_alias id (Some ty) lid ~attrs ~loc ~docs,
+        ext }
 ;
 
 (* Type declarations and type substitutions. *)
