--- ocaml_505/parsing/printast.ml	2026-07-07 01:48:43.511538860 +0200
+++ ocaml_506_typing_recovery/parsing/printast.ml	2026-07-07 02:21:01.063870354 +0200
@@ -439,7 +439,23 @@
        x.pval_name fmt_location x.pval_loc;
   attributes i ppf x.pval_attributes;
   core_type (i+1) ppf x.pval_type;
-  list (i+1) string ppf x.pval_prim
+
+and primitive_description i ppf x =
+  line i ppf "primitive_description %a %a\n" fmt_string_loc
+       x.pprim_name fmt_location x.pprim_loc;
+  attributes i ppf x.pprim_attributes;
+  let i = i + 1 in
+  match x.pprim_kind with
+  | Pprim_decl (pprim_type, pprim_prim) ->
+    line i ppf "Pprim_decl\n";
+    let i = i + 1 in
+    core_type i ppf pprim_type;
+    list i string ppf pprim_prim
+  | Pprim_alias (pprim_type, pprim_ident) ->
+    line i ppf "Pprim_alias\n";
+    let i = i + 1 in
+    option i core_type ppf pprim_type;
+    longident_loc i ppf pprim_ident
 
 and type_parameter i ppf (x, _variance) = core_type i ppf x
 
@@ -726,7 +742,7 @@
       line i ppf "Pmty_typeof\n";
       module_expr i ppf m;
   | Pmty_extension (s, arg) ->
-      line i ppf "Pmod_extension \"%s\"\n" s.txt;
+      line i ppf "Pmty_extension \"%s\"\n" s.txt;
       payload i ppf arg
 
 and signature i ppf x = list i signature_item ppf x
@@ -738,6 +754,9 @@
   | Psig_value vd ->
       line i ppf "Psig_value\n";
       value_description i ppf vd;
+  | Psig_primitive pd ->
+      line i ppf "Psig_primitive\n";
+      primitive_description i ppf pd;
   | Psig_type (rf, l) ->
       line i ppf "Psig_type %a\n" fmt_rec_flag rf;
       list i type_declaration ppf l;
@@ -867,9 +886,12 @@
   | Pstr_value (rf, l) ->
       line i ppf "Pstr_value %a\n" fmt_rec_flag rf;
       list i value_binding ppf l;
-  | Pstr_primitive vd ->
-      line i ppf "Pstr_primitive\n";
+  | Pstr_val vd ->
+      line i ppf "Pstr_val\n";
       value_description i ppf vd;
+  | Pstr_primitive pd ->
+      line i ppf "Pstr_primitive\n";
+      primitive_description i ppf pd;
   | Pstr_type (rf, l) ->
       line i ppf "Pstr_type %a\n" fmt_rec_flag rf;
       list i type_declaration ppf l;
