--- ocaml_505/utils/config_boot.ml	2026-07-07 01:48:43.521971910 +0200
+++ ocaml_506_typing_recovery/utils/config_boot.ml	2026-07-07 02:21:40.468041548 +0200
@@ -114,7 +114,7 @@
    from values in ../build-aux/ocaml_version.m4 - do not edit these lines
    directly. *)
 let is_official_release = false
-let release_number = 21
+let release_number = 22
 
 external standard_library_default : unit -> string = "%standard_library_default"
 
@@ -148,18 +148,18 @@
   else
     target_bindir
 
-let exec_magic_number = {magic|Caml1999X037|magic}
+let exec_magic_number = {magic|Caml1999X038|magic}
     (* exec_magic_number is duplicated in runtime/caml/exec.h *)
-and cmi_magic_number = {magic|Caml1999I037|magic}
-and cmo_magic_number = {magic|Caml1999O037|magic}
-and cma_magic_number = {magic|Caml1999A037|magic}
-and cmx_magic_number = {magic|Caml1999Y037|magic}
-and cmxa_magic_number = {magic|Caml1999Z037|magic}
-and ast_impl_magic_number = {magic|Caml1999M037|magic}
-and ast_intf_magic_number = {magic|Caml1999N037|magic}
-and cmxs_magic_number = {magic|Caml1999D037|magic}
-and cmt_magic_number = {magic|Caml1999T037|magic}
-and linear_magic_number = {magic|Caml1999L037|magic}
+and cmi_magic_number = {magic|Caml1999I038|magic}
+and cmo_magic_number = {magic|Caml1999O038|magic}
+and cma_magic_number = {magic|Caml1999A038|magic}
+and cmx_magic_number = {magic|Caml1999Y038|magic}
+and cmxa_magic_number = {magic|Caml1999Z038|magic}
+and ast_impl_magic_number = {magic|Caml1999M038|magic}
+and ast_intf_magic_number = {magic|Caml1999N038|magic}
+and cmxs_magic_number = {magic|Caml1999D038|magic}
+and cmt_magic_number = {magic|Caml1999T038|magic}
+and linear_magic_number = {magic|Caml1999L038|magic}
 
 let safe_string = true
 let default_safe_string = true
@@ -272,6 +272,7 @@
   p_bool "native_dynlink" native_dynlink;
   p_bool "naked_pointers" naked_pointers;
   p_bool "with_codegen_invariants" with_codegen_invariants;
+  p_int "reserved_header_bits" reserved_header_bits;
 
   p "exec_magic_number" exec_magic_number;
   p "cmi_magic_number" cmi_magic_number;
