--- ocaml_505/utils/domainstate.ml.c	2026-07-07 01:48:43.521971910 +0200
+++ ocaml_506_typing_recovery/utils/domainstate.ml.c	2026-07-07 02:21:40.477991720 +0200
@@ -15,12 +15,12 @@
 /**************************************************************************/
 
 #define CAML_CONFIG_H_NO_TYPEDEFS
-#include "config.h"
+#include "caml/config.h"
 let stack_ctx_words = Stack_ctx_words
 
 type t =
 #define DOMAIN_STATE(type, name) | Domain_##name
-#include "domain_state.tbl"
+#include "caml/domain_state.tbl"
 #undef DOMAIN_STATE
 
 let idx_of_field =
@@ -28,11 +28,11 @@
 #define DOMAIN_STATE(type, name) \
   let idx__##name = curr in \
   let curr = curr + 1 in
-#include "domain_state.tbl"
+#include "caml/domain_state.tbl"
 #undef DOMAIN_STATE
   let _ = curr in
   function
 #define DOMAIN_STATE(type, name) \
   | Domain_##name -> idx__##name
-#include "domain_state.tbl"
+#include "caml/domain_state.tbl"
 #undef DOMAIN_STATE
