diff -Naur a/Makefile b/Makefile
--- a/Makefile	2015-12-07 21:58:29.000000000 +0000
+++ b/Makefile	2017-07-21 14:19:28.720009800 +0100
@@ -3,6 +3,7 @@
 
 NATDYNLINK := $(shell if [ -f `ocamlfind ocamlc -where`/dynlink.cmxa ]; \
                       then echo YES; else echo NO; fi)
+EXT_OBJ = $(shell ocamlc -config | sed -ne "s/ext_obj: //p" | tr -d '\r')
 
 ifeq "${NATDYNLINK}" "YES"
 CMXS=easy_format.cmxs
@@ -53,7 +54,7 @@
 	caml2html easy_format_example.ml -t -o easy_format_example.html
 
 soft-clean:
-	rm -f *.cm[iox] *.cmxs *.o *.annot \
+	rm -f *.cm[iox] *.cmxs *.o *.obj *.annot \
 		test_easy_format lambda_example simple_example \
 		bytecode nativecode
 
@@ -65,7 +66,7 @@
 
 COMMON_INSTALL_FILES = META easy_format.cmi easy_format.mli
 BC_INSTALL_FILES = easy_format.cmo
-NC_INSTALL_FILES = easy_format.cmx easy_format.o $(CMXS)
+NC_INSTALL_FILES = easy_format.cmx easy_format$(EXT_OBJ) $(CMXS)
 
 install:
 	echo "version = \"$(VERSION)\"" > META; cat META.tpl >> META
