--- ocaml_504/typing/btype.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/btype.ml	2026-02-03 16:48:24
@@ -155,6 +155,7 @@
   newty3 ~level ~scope:Ident.lowest_scope desc
 
 let newgenty desc      = newty2 ~level:generic_level desc
+let newgenmono desc    = newgenty (Tpoly(desc, []))
 let newgenvar ?name () = newgenty (Tvar name)
 let newgenstub ~scope  = newty3 ~level:generic_level ~scope (Tvar None)
 
@@ -163,6 +164,7 @@
 let is_Tvar ty = match get_desc ty with Tvar _ -> true | _ -> false
 let is_Tunivar ty = match get_desc ty with Tunivar _ -> true | _ -> false
 let is_Tconstr ty = match get_desc ty with Tconstr _ -> true | _ -> false
+let is_Tpoly ty = match get_desc ty with Tpoly _ -> true | _ -> false
 let is_poly_Tpoly ty =
   match get_desc ty with Tpoly (_, _ :: _) -> true | _ -> false
 let type_kind_is_abstract decl =
@@ -170,10 +172,37 @@
 let type_origin decl =
   match decl.type_kind with
   | Type_abstract origin -> origin
-  | Type_variant _ | Type_record _ | Type_open -> Definition
+  | Type_variant _ | Type_record _ | Type_open | Type_external _ -> Definition
 
 let dummy_method = "*dummy method*"
 
+                  (********************************)
+                  (*  Utilities for poly types    *)
+                  (********************************)
+
+let tpoly_is_mono ty =
+  match get_desc ty with
+  | Tpoly(_, []) -> true
+  | Tpoly(_, _ :: _) -> false
+  | _ -> assert false
+
+let tpoly_get_poly ty =
+  match get_desc ty with
+  | Tpoly(ty, vars) -> (ty, vars)
+  | _ -> assert false
+
+let tpoly_get_mono ty =
+  match get_desc ty with
+  | Tpoly(ty, []) -> ty
+  | _ -> assert false
+
+let tpoly_get_mono_opt ty =
+  match get_desc ty with
+  | Tpoly(ty, []) -> Some ty
+  | Tpoly _ -> None
+  | _ -> assert false
+
+
 (**** Representative of a type ****)
 
 let merge_fixed_explanation fixed1 fixed2 =
@@ -254,7 +283,7 @@
   let l = String.length s in
   (* PR#10661: when l=4 and s is "#row", this is not a row name
      but the valid #-type name of a class named "row". *)
-  l > 4 && String.sub s (l-4) 4 = "#row"
+  l > 4 && String.ends_with ~suffix:"#row" s
 
 let is_constr_row ~allow_ident t =
   match get_desc t with
@@ -323,14 +352,20 @@
       let result = f init ty1 in
       f result ty2
   | Tnil                -> init
-  | Tlink _
-  | Tsubst _            -> assert false
+  | Tlink _             -> assert false
+  | Tsubst _            -> init
   | Tunivar _           -> init
   | Tpoly (ty, tyl)     ->
-    let result = f init ty in
-    List.fold_left f result tyl
+      let result = f init ty in
+      List.fold_left f result tyl
   | Tpackage pack ->
-    List.fold_left (fun result (_n, ty) -> f result ty) init pack.pack_cstrs
+      List.fold_left
+        (fun result (_n, ty) -> f result ty) init pack.pack_constraints
+  | Tfunctor (_, _, pack, ty) ->
+      let res =
+        List.fold_left (fun result (_n, ty) -> f result ty) init
+          pack.pack_constraints in
+      f res ty
 
 let iter_type_expr f ty =
   fold_type_expr (fun () v -> f v) () ty
@@ -360,9 +395,15 @@
         cstrs
   | Type_record(lbls, _) ->
       List.iter (fun d -> f d.ld_type) lbls
-  | Type_open ->
-      ()
+  | Type_open -> ()
+  | Type_external _ -> ()
 
+let map_pack map_path map_type { pack_path; pack_constraints } =
+  {
+    pack_path = map_path pack_path;
+    pack_constraints = List.map (Pair.map_snd map_type) pack_constraints;
+  }
+
                   (**********************************)
                   (*     Utilities for marking      *)
                   (**********************************)
@@ -476,6 +517,7 @@
     match get_desc ty with
       Tconstr (p, _, _)
     | Tobject (_, {contents=Some (p, _)})
+    | Tfunctor (_, _, {pack_path = p}, _)
     | Tpackage {pack_path = p} ->
         it.it_path p
     | Tvariant row ->
@@ -531,8 +573,12 @@
       let tyl = List.map f tyl in
       Tpoly (f ty, tyl)
   | Tpackage pack       ->
-      Tpackage {pack with
-        pack_cstrs = List.map (fun (n, ty) -> (n, f ty)) pack.pack_cstrs}
+      let pack_constraints =
+        List.map (fun (n, ty) -> (n, f ty)) pack.pack_constraints in
+      Tpackage {pack with pack_constraints}
+  | Tfunctor _ ->
+      (* doing this would break unicity of unscoped binding in Tfunctor *)
+      assert false
 
 (* TODO: rename to [module Copy_scope] *)
 module For_copy : sig
