--- ocaml_504/typing/datarepr.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/datarepr.ml	2026-02-03 16:48:24
@@ -237,11 +237,11 @@
   match decl.type_kind with
   | Type_variant (cstrs,rep) ->
      constructor_descrs ~current_unit ty_path decl cstrs rep
-  | Type_record _ | Type_abstract _ | Type_open -> []
+  | Type_record _ | Type_abstract _ | Type_open | Type_external _ -> []
 
 let labels_of_type ty_path decl =
   match decl.type_kind with
   | Type_record(labels, rep) ->
       label_descrs (newgenconstr ty_path decl.type_params)
         labels rep decl.type_private
-  | Type_variant _ | Type_abstract _ | Type_open -> []
+  | Type_variant _ | Type_abstract _ | Type_open | Type_external _ -> []
