--- ocaml_504/typing/envaux.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/envaux.ml	2026-02-03 16:48:24
@@ -68,12 +68,12 @@
           | Error `Functor -> assert false
           | Error `Not_found -> raise (Error (Module_not_found path'))
           end
-      | Env_functor_arg(Env_module(s, id, pres, desc), id')
+      | Env_not_aliasable(Env_module(s, id, pres, desc), id')
             when Ident.same id id' ->
           Env.add_module_declaration ~check:false
             id pres (Subst.module_declaration Keep subst desc)
-            ~arg:true (env_from_summary s subst)
-      | Env_functor_arg _ -> assert false
+            ~noalias:true (env_from_summary s subst)
+      | Env_not_aliasable _ -> assert false
       | Env_constraints(s, map) ->
           Path.Map.fold
             (fun path info ->
