--- ocaml_504/typing/errortrace.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/errortrace.ml	2026-02-03 16:48:24
@@ -19,11 +19,17 @@
 open Format_doc
 
 type position = First | Second
+type order = Less | Equal | More
 
 let swap_position = function
   | First -> Second
   | Second -> First
 
+let swap_order = function
+  | Less -> More
+  | Equal -> Equal
+  | More -> Less
+
 let print_pos ppf = function
   | First -> fprintf ppf "first"
   | Second -> fprintf ppf "second"
@@ -49,6 +55,7 @@
      we keep a [type_expr] to track renaming in {!Printtyp} *)
   | Self
   | Module_type of Path.t
+  | Module of Ident.t
   | Equation of 'a
   | Constraint
 
@@ -59,7 +66,8 @@
 let map_escape f esc =
   {esc with kind = match esc.kind with
      | Equation eq -> Equation (f eq)
-     | (Constructor _ | Univ _ | Self | Module_type _ | Constraint) as c -> c}
+     | (Constructor _ | Univ _ | Self | Module_type _
+        | Module _ | Constraint) as c -> c}
 
 let explain trace f =
   let rec explain = function
@@ -103,6 +111,10 @@
     | Package_inclusion of Format_doc.doc
     | Package_coercion of Format_doc.doc
 
+type univar =
+  | Var_mismatch of { order:order; diff:type_expr diff }
+  | Quantification_mismatch of type_expr list
+
 type ('a, 'variety) elt =
   (* Common *)
   | Diff : 'a diff -> ('a, _) elt
@@ -114,6 +126,7 @@
   | Incompatible_fields : { name:string; diff: type_expr diff } -> ('a, _) elt
       (* Could move [Incompatible_fields] into [obj] *)
   | First_class_module: first_class_module -> ('a,_) elt
+  | Univar of univar
   (* Unification & Moregen; included in Equality for simplicity *)
   | Rec_occur : type_expr * type_expr -> ('a, _) elt
 
@@ -126,11 +139,13 @@
   | Diff x -> Diff (map_diff f x)
   | Escape {kind = Equation x; context} ->
       Escape { kind = Equation (f x); context }
-  | Escape {kind = (Univ _ | Self | Constructor _ | Module_type _ | Constraint);
+  | Escape {kind = (Univ _ | Self | Constructor _
+      | Module_type _ | Module _ | Constraint);
             _}
   | Variant _ | Obj _ | Function_label_mismatch _ | Tuple_label_mismatch _
   | Incompatible_fields _
   | Rec_occur (_, _) | First_class_module _  as x -> x
+  | Univar _  as x -> x
 
 let map f t = List.map (map_elt f) t
 
@@ -147,6 +162,12 @@
     Variant (Fixed_row(swap_position pos,k,f))
   | Variant (No_tags(pos,f)) ->
     Variant (No_tags(swap_position pos,f))
+  | Univar (Var_mismatch d) ->
+      Univar (Var_mismatch {
+        order = swap_order d.order;
+        diff = swap_diff d.diff
+      })
+  | Univar (Quantification_mismatch _) as x -> x
   | x -> x
 
 let swap_trace e = List.map swap_elt e
