--- ocaml_504/typing/errortrace.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/errortrace.mli	2026-02-03 16:48:24
@@ -18,10 +18,13 @@
 open Types
 
 type position = First | Second
+type order = Less | Equal | More
 
 val swap_position : position -> position
 val print_pos : position Format_doc.printer
 
+val swap_order: order -> order
+
 type expanded_type = { ty: type_expr; expanded: type_expr }
 
 (** [trivial_expansion ty] creates an [expanded_type] whose expansion is also
@@ -44,6 +47,7 @@
      we keep a [type_expr] to track renaming in {!Printtyp} *)
   | Self
   | Module_type of Path.t
+  | Module of Ident.t
   | Equation of 'a
   | Constraint
 
@@ -89,6 +93,10 @@
     | Package_inclusion of Format_doc.doc
     | Package_coercion of Format_doc.doc
 
+type univar =
+  | Var_mismatch of { order:order; diff:type_expr diff }
+  | Quantification_mismatch of type_expr list
+
 type ('a, 'variety) elt =
   (* Common *)
   | Diff : 'a diff -> ('a, _) elt
@@ -99,6 +107,7 @@
   | Tuple_label_mismatch of string option diff
   | Incompatible_fields : { name:string; diff: type_expr diff } -> ('a, _) elt
   | First_class_module: first_class_module -> ('a,_) elt
+  | Univar of univar
   (* Unification & Moregen; included in Equality for simplicity *)
   | Rec_occur : type_expr * type_expr -> ('a, _) elt
 
