--- ocaml_504/typing/errortrace_report.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/errortrace_report.ml	2026-02-03 16:48:24
@@ -117,10 +117,14 @@
 let print_tags ppf tags  =
   Fmt.(pp_print_list ~pp_sep:comma) print_tag ppf tags
 
-let is_unit env ty =
-  match Types.get_desc (Ctype.expand_head env ty) with
-  | Tconstr (p, _, _) -> Path.same p Predef.path_unit
-  | _ -> false
+let is_unit_param env ty =
+  let ty, vars = Btype.tpoly_get_poly ty in
+  if vars <> [] then false
+  else begin
+    match Types.get_desc (Ctype.expand_head env ty) with
+    | Tconstr (p, _, _) -> Path.same p Predef.path_unit
+    | _ -> false
+  end
 
 let unifiable env ty1 ty2 =
   let snap = Btype.snapshot () in
@@ -134,13 +138,13 @@
 let explanation_diff env t3 t4 =
   match Types.get_desc t3, Types.get_desc t4 with
   | Tarrow (_, ty1, ty2, _), _
-    when is_unit env ty1 && unifiable env ty2 t4 ->
+    when is_unit_param env ty1 && unifiable env ty2 t4 ->
       Some (doc_printf
           "@,@[@{<hint>Hint@}: Did you forget to provide %a as argument?@]"
           Style.inline_code "()"
         )
   | _, Tarrow (_, ty1, ty2, _)
-    when is_unit env ty1 && unifiable env t3 ty2 ->
+    when is_unit_param env ty1 && unifiable env t3 ty2 ->
       Some (doc_printf
           "@,@[@{<hint>Hint@}: Did you forget to wrap the expression using \
            %a?@]"
@@ -233,6 +237,11 @@
         "%a@,@[The module type@;<1 2>%a@ would escape its scope@]"
         pp_doc pre pp_path p
     )
+  | Errortrace.Module id -> Some(
+      doc_printf
+        "%a@,@[The module@;<1 2>%a@ would escape its scope@]"
+        pp_doc pre pp_path (Path.Pident id)
+    )
   | Errortrace.Equation Errortrace.{ty = _; expanded = t} ->
       Variable_names.reserve t;
       Some(
@@ -307,6 +316,58 @@
   | Errortrace.Package_coercion pr ->
       Some(doc_printf "@,@[%a@]" Fmt.pp_doc pr)
 
+let explain_univar prev = function
+  | Errortrace.Var_mismatch { diff; order} ->
+      let prev = match prev with
+        | Some (Errortrace.Incompatible_fields f) ->
+            explain_incompatible_fields f.name f.diff
+        | _ -> Fmt.Doc.empty
+      in
+      add_type_to_preparation diff.got;
+      add_type_to_preparation diff.expected;
+      let more = match order with
+        | Equal ->  Fmt.Doc.empty
+        | Less ->
+          Fmt.doc_printf
+            "@ The first type variable %a was introduced in@ an@ earlier@ \
+             universal@ quantification."
+              (Style.as_inline_code prepared_type_expr) diff.got
+        | More ->
+            Fmt.doc_printf
+              "@ The second type variable %a was introduced in@ an@ earlier@ \
+               universal@ quantification."
+              (Style.as_inline_code prepared_type_expr) diff.expected
+      in
+      doc_printf
+        "%a@,@[The universal variables@ %a and@ %a@ are distinct.%a@]"
+        Fmt.pp_doc prev
+        (Style.as_inline_code prepared_type_expr) diff.got
+        (Style.as_inline_code prepared_type_expr) diff.expected
+        pp_doc more
+  | Errortrace.Quantification_mismatch delta ->
+      let qp ppf x = Style.as_inline_code prepared_type_expr ppf x in
+      let pp ppf ty =
+        add_type_to_preparation ty;
+        match Types.get_desc ty with
+        | Tunivar None -> ()
+        | Tunivar (Some name) ->
+            Fmt.fprintf ppf
+              "@,@[The universal type variable %a in the first@ type@ matches@ \
+               multiple@ distinct@ variables in the second type.@]"
+              Style.inline_code ("'" ^ name)
+        | Tvar _ ->
+              Fmt.fprintf ppf
+                "@,@[The type variable %a is not generalizable@ to@ an@ \
+                 universal@ type variable.@]"
+                qp ty
+        | _ ->
+              Fmt.fprintf ppf
+                "@,@[The type %a is not a type variable.@]"
+                qp ty
+      in
+      let pp_sep _ () = () in
+      doc_printf "%a" (pp_print_list ~pp_sep pp) delta
+
 let explanation (type variety) intro prev env
   : (Errortrace.expanded_type, variety) Errortrace.elt -> _ = function
   | Errortrace.Diff {got; expected} ->
@@ -369,6 +430,7 @@
              {[ The type int occurs inside int list -> 'a |}
         *)
     end
+  | Univar um -> Some (explain_univar prev um)
 
 let mismatch intro env trace =
   Errortrace.explain trace (fun ~prev h -> explanation intro prev env h)
@@ -389,6 +451,11 @@
               "@,@[<hov>Type %a was considered abstract@ when checking\
                @ constraints@ in this@ recursive type definition.@]"
               pp_path p
+        | Approx_recmod ->
+            fprintf ppf
+              "@,@[<hov>Type %a was considered abstract@ when checking\
+               @ constraints@ in this@ recursive module definition.@]"
+              pp_path p
         | Definition | Existential _ -> ()
       end
   | _ -> ()
@@ -440,8 +507,8 @@
       let tr = match mis, last with
         | None, Some elt -> tr @ [elt]
         | Some _, _ | _, None -> tr
-       in
-       fprintf ppf
+      in
+      fprintf ppf
         "@[<v>\
           @[%a%a@]%a%a\
          @]"
@@ -451,8 +518,8 @@
         (pp_print_option pp_doc) mis;
       if env <> Env.empty
       then warn_on_missing_defs env ppf head;
-       Internal_names.print_explanations env ppf;
-       Ident_conflicts.err_print ppf
+      Internal_names.print_explanations env ppf;
+      Ident_conflicts.err_print ppf
     )
 
 let report_error trace_format ppf mode env tr
