--- ocaml_504/typing/gprinttyp.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/gprinttyp.ml	2026-02-03 16:48:24
@@ -199,7 +199,7 @@
           Hashtbl.replace id_map.tbl id last;
           Main last
 
-  (** Generate color from the node id to keep the color stable inbetween
+  (** Generate color from the node id to keep the color stable in between
       different calls to the typechecker on the same input. *)
   let colorize_id params id =
     if not params.colorize then None
@@ -667,6 +667,11 @@
     snd @@ List.fold_left
       (labeled_edge params id0)
       (0,gh) l
+  and package_constraint params id0 gh (l, ty) =
+    let l = labelr "%a =" Pp.longident (Option.get @@ Longident.unflatten l) in
+    edge params id0 l ty gh
+  and package_constraints params id0 l gh =
+    List.fold_left (package_constraint params id0) gh l
   and node params color ~lvl id tynode desc dg =
     let add_tynode l = add_node l color ~lvl id tynode dg in
     let mk fmt = labelk (fun l -> add_tynode (Decoration.make l)) fmt in
@@ -677,6 +682,12 @@
     | Types.Tvar name -> mk "%a" Pp.pretty_var name
     | Types.Tarrow(l,t1,t2,_) ->
        mk "→%a" Pp.exponent_of_label l |> numbered [t1; t2]
+    | Types.Tfunctor(l,us,{pack_path; pack_constraints},t2) ->
+        mk "→%a (%s : %a)" Pp.exponent_of_label l
+                    (Ident.Unscoped.name us)
+                    pp_path pack_path
+          |> package_constraints params id pack_constraints
+          |> numbered [t2]
     | Types.Ttuple tl ->
         mk "*" |> labeled_edges params id tl
     | Types.Tconstr (p,tl,abbrevs) ->
@@ -742,15 +753,10 @@
             fields
         in
         { elts; graph = add_subgraph (labelr "polyvar", fields) main }
-    | Types.Tpackage {pack_path; pack_cstrs} ->
-        let types = List.map snd pack_cstrs in
-        let pp_cstrs ppf (l, _) =
-          Pp.longident ppf (Option.get @@ Longident.unflatten l)
-        in
-        mk "[mod %a with %a]"
+    | Types.Tpackage {pack_path; pack_constraints} ->
+        mk "[mod %a]"
           pp_path pack_path
-          Pp.(list ~sep:semi pp_cstrs) pack_cstrs
-        |> numbered types
+        |> package_constraints params id pack_constraints
   and variant params id0 (elts,main,fields) (name,rf)  =
     let id = Index.subnode ~name id0 in
     let fnode = Node id in
