--- ocaml_504/typing/ident.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/ident.ml	2026-02-03 16:48:24
@@ -19,6 +19,83 @@
 let highest_scope = 100_000_000
   (* assumed to fit in 27 bits, see Types.scope_field *)
 
+(* A stamp of 0 denotes a persistent identifier *)
+
+let currentstamp = s_ref 0
+let predefstamp = s_ref 0
+
+module Unscoped = struct
+
+  type desc = { name: string; stamp: int }
+
+  type state =
+    | Udesc of desc
+    | Ulink of t
+  and t = { mutable state : state }
+
+  let create s =
+    incr currentstamp;
+    { state = Udesc { name = s; stamp = !currentstamp } }
+
+  let rec get_desc us =
+    match us.state with
+      Udesc d -> d
+    | Ulink u -> get_desc u
+
+  let name us = (get_desc us).name
+
+  let refresh us = create (get_desc us).name
+
+  let equal us1 us2 = (get_desc us1).name = (get_desc us2).name
+
+  let stamp us = (get_desc us).stamp
+
+  let same us1 us2 = stamp us1 = stamp us2
+
+
+  type change = t * state
+
+  let change_log = ref (fun _ -> assert false)
+
+  let undo_change (us, us_d) =
+    us.state <- us_d
+
+  let rec repr us =
+    match us.state with
+    | Ulink us' ->
+      repr us'
+    | Udesc _ -> us
+
+  let link us1 us2 =
+    let us1 = repr us1 in
+    let us2 = repr us2 in
+    if us1 == us2 then () else begin
+    !change_log (us1, us1.state);
+    us1.state <- Ulink us2
+    end
+
+  let equiv id_pairs us1 us2 =
+    (*
+      The list [id_pairs] does not have a cannonical order of pairs
+      so we need to check if (us1, us2) or (us2, us1) occurs in the list.
+      The reason why we do not have a cannonical representation is because
+      unification identifiers are mutable, they can change once put inside the
+      list.
+    *)
+    let s1 = stamp us1 in
+    let s2 = stamp us2 in
+    s1 = s2
+    || List.exists (fun (i1, i2) -> (stamp i1 = s1 && stamp i2 = s2)
+                            || stamp i2 = s1 && stamp i1 = s2) id_pairs
+
+  module Ops = struct
+    type nonrec t = t
+    let compare us1 us2 = stamp us1 - stamp us2
+  end
+
+  module Set = Stdlib.Set.Make(Ops)
+end
+
 type t =
   | Local of { name: string; stamp: int }
   | Scoped of { name: string; stamp: int; scope: int }
@@ -26,12 +103,8 @@
   | Predef of { name: string; stamp: int }
       (* the stamp is here only for fast comparison, but the name of
          predefined identifiers is always unique. *)
+  | Unscoped of Unscoped.t
 
-(* A stamp of 0 denotes a persistent identifier *)
-
-let currentstamp = s_ref 0
-let predefstamp = s_ref 0
-
 let create_scoped ~scope s =
   incr currentstamp;
   Scoped { name = s; stamp = !currentstamp; scope }
@@ -40,6 +113,13 @@
   incr currentstamp;
   Local { name = s; stamp = !currentstamp }
 
+let of_unscoped u =
+  Unscoped u
+
+let find_unscoped = function
+    Unscoped u -> Some u
+  | _ -> None
+
 let create_predef s =
   incr predefstamp;
   Predef { name = s; stamp = !predefstamp }
@@ -52,18 +132,23 @@
   | Scoped { name; _ }
   | Global name
   | Predef { name; _ } -> name
+  | Unscoped us -> Unscoped.name us
 
 let rename = function
   | Local { name; stamp = _ }
   | Scoped { name; stamp = _; scope = _ } ->
       incr currentstamp;
       Local { name; stamp = !currentstamp }
+  | Unscoped us -> Unscoped (Unscoped.refresh us)
   | id ->
       Misc.fatal_errorf "Ident.rename %s" (name id)
 
 let unique_name = function
   | Local { name; stamp }
   | Scoped { name; stamp } -> name ^ "_" ^ Int.to_string stamp
+  | Unscoped us ->
+    let Unscoped.{ name; stamp } = Unscoped.get_desc us in
+    name ^ "_" ^ Int.to_string stamp
   | Global name ->
       (* we're adding a fake stamp, because someone could have named his unit
          [Foo_123] and since we're using unique_name to produce symbol names,
@@ -77,6 +162,9 @@
 let unique_toplevel_name = function
   | Local { name; stamp }
   | Scoped { name; stamp } -> name ^ "/" ^ Int.to_string stamp
+  | Unscoped us ->
+    let Unscoped.{ name; stamp } = Unscoped.get_desc us in
+    name ^ "/" ^ Int.to_string stamp
   | Global name
   | Predef { name; _ } -> name
 
@@ -90,6 +178,7 @@
   | Scoped { name = name1; _ }, Scoped { name = name2; _ }
   | Global name1, Global name2 ->
       name1 = name2
+  | Unscoped us1, Unscoped us2 -> Unscoped.equal us1 us2
   | Predef { stamp = s1; _ }, Predef { stamp = s2 } ->
       (* if they don't have the same stamp, they don't have the same name *)
       s1 = s2
@@ -102,6 +191,7 @@
   | Scoped { stamp = s1; _ }, Scoped { stamp = s2; _ }
   | Predef { stamp = s1; _ }, Predef { stamp = s2 } ->
       s1 = s2
+  | Unscoped us1, Unscoped us2 -> Unscoped.same us1 us2
   | Global name1, Global name2 ->
       name1 = name2
   | _ ->
@@ -110,8 +200,21 @@
 let stamp = function
   | Local { stamp; _ }
   | Scoped { stamp; _ } -> stamp
+  | Unscoped us -> Unscoped.stamp us
   | _ -> 0
 
+let equiv id_pairs i1 i2 =
+  match i1, i2 with
+  | Local { stamp = s1; _ }, Local { stamp = s2; _ }
+  | Scoped { stamp = s1; _ }, Scoped { stamp = s2; _ }
+  | Predef { stamp = s1; _ }, Predef { stamp = s2 } ->
+      s1 = s2
+  | Unscoped us1, Unscoped us2 -> Unscoped.equiv id_pairs us1 us2
+  | Global name1, Global name2 ->
+      name1 = name2
+  | _ ->
+      false
+
 let compare_stamp id1 id2 =
   compare (stamp id1) (stamp id2)
 
@@ -119,6 +222,7 @@
   | Scoped { scope; _ } -> scope
   | Local _ -> highest_scope
   | Global _ | Predef _ -> lowest_scope
+  | Unscoped _ -> lowest_scope
 
 let reinit_level = ref (-1)
 
@@ -129,6 +233,7 @@
 
 let global = function
   | Local _
+  | Unscoped _
   | Scoped _ -> false
   | Global _
   | Predef _ -> true
@@ -137,6 +242,10 @@
   | Predef _ -> true
   | _ -> false
 
+let is_unscoped = function
+  | Unscoped _ -> true
+  | _ -> false
+
 let canonical_stamps = s_table Hashtbl.create 0
 let next_canonical_stamp = s_table Hashtbl.create 0
 
@@ -176,6 +285,10 @@
   | Local { name; stamp } ->
       fprintf ppf "%a"
         pp_stamped (name, stamp)
+  | Unscoped us ->
+      let Unscoped.{ name; stamp } = Unscoped.get_desc us in
+      fprintf ppf "U%a"
+        pp_stamped (name, stamp)
   | Scoped { name; stamp; scope } ->
       fprintf ppf "%a%s"
         pp_stamped (name, stamp)
@@ -204,6 +317,8 @@
 
 and 'a data =
   { ident: t;
+    name : string;
+    stamp : int;
     data: 'a;
     previous: 'a data option }
 
@@ -215,6 +330,14 @@
  *   | Node(_,_,_,h) -> h
  *)
 
+(* Before returning the data we check that the corresponding binder
+  was not mutated since it has been added to the environment
+  (for example : Unscoped indentifiers and Unscoped.link).
+*)
+let get_data k =
+  assert (stamp k.ident = k.stamp);
+  k.data
+
 let mknode l d r =
   let hl = match l with Empty -> 0 | Node(_,_,_,h) -> h
   and hr = match r with Empty -> 0 | Node(_,_,_,h) -> h in
@@ -246,11 +369,23 @@
 
 let rec add id data = function
     Empty ->
-      Node(Empty, {ident = id; data = data; previous = None}, Empty, 1)
+      let k = {
+        ident = id;
+        name = name id;
+        stamp = stamp id;
+        data = data;
+        previous = None
+      } in Node(Empty, k, Empty, 1)
   | Node(l, k, r, h) ->
-      let c = String.compare (name id) (name k.ident) in
+      let c = String.compare (name id) k.name in
       if c = 0 then
-        Node(l, {ident = id; data = data; previous = Some k}, r, h)
+        let new_k = {
+          ident = id;
+          name = name id;
+          stamp = stamp id;
+          data = data;
+          previous = Some k
+        } in Node(l, new_k, r, h)
       else if c < 0 then
         balance (add id data l) k r
       else
@@ -278,7 +413,7 @@
     Empty ->
       Empty
   | (Node (l, k, r, h) as m) ->
-      let c = String.compare (name id) (name k.ident) in
+      let c = String.compare (name id) k.name in
       if c = 0 then
         match k.previous with
         | None -> merge l r
@@ -292,16 +427,16 @@
     None ->
       raise Not_found
   | Some k ->
-      if same id k.ident then k.data else find_previous id k.previous
+      if same id k.ident then get_data k else find_previous id k.previous
 
 let rec find_same id = function
     Empty ->
       raise Not_found
   | Node(l, k, r, _) ->
-      let c = String.compare (name id) (name k.ident) in
+      let c = String.compare (name id) k.name in
       if c = 0 then
         if same id k.ident
-        then k.data
+        then get_data k
         else find_previous id k.previous
       else
         find_same id (if c < 0 then l else r)
@@ -310,37 +445,37 @@
     Empty ->
       raise Not_found
   | Node(l, k, r, _) ->
-      let c = String.compare n (name k.ident) in
+      let c = String.compare n k.name in
       if c = 0 then
-        k.ident, k.data
+        k.ident, get_data k
       else
         find_name n (if c < 0 then l else r)
 
 let rec get_all = function
   | None -> []
-  | Some k -> (k.ident, k.data) :: get_all k.previous
+  | Some k -> (k.ident, get_data k) :: get_all k.previous
 
 let rec find_all n = function
     Empty ->
       []
   | Node(l, k, r, _) ->
-      let c = String.compare n (name k.ident) in
+      let c = String.compare n k.name in
       if c = 0 then
-        (k.ident, k.data) :: get_all k.previous
+        (k.ident, get_data k) :: get_all k.previous
       else
         find_all n (if c < 0 then l else r)
 
 let get_all_seq k () =
-  Seq.unfold (Option.map (fun k -> (k.ident, k.data), k.previous))
+  Seq.unfold (Option.map (fun k -> (k.ident, get_data k), k.previous))
     k ()
 
 let rec find_all_seq n tbl () =
   match tbl with
   | Empty -> Seq.Nil
   | Node(l, k, r, _) ->
-      let c = String.compare n (name k.ident) in
+      let c = String.compare n k.name in
       if c = 0 then
-        Seq.Cons((k.ident, k.data), get_all_seq k.previous)
+        Seq.Cons((k.ident, get_data k), get_all_seq k.previous)
       else
         find_all_seq n (if c < 0 then l else r) ()
 
@@ -354,12 +489,13 @@
   | Node(l, k, r, _) ->
       fold_aux f (l :: stack) (f k accu) r
 
-let fold_name f tbl accu = fold_aux (fun k -> f k.ident k.data) [] accu tbl
+let fold_name f tbl accu =
+  fold_aux (fun k -> f k.ident (get_data k)) [] accu tbl
 
 let rec fold_data f d accu =
   match d with
     None -> accu
-  | Some k -> f k.ident k.data (fold_data f k.previous accu)
+  | Some k -> f k.ident (get_data k) (fold_data f k.previous accu)
 
 let fold_all f tbl accu =
   fold_aux (fun k -> fold_data f (Some k)) [] accu tbl
@@ -369,7 +505,7 @@
 let rec iter f = function
     Empty -> ()
   | Node(l, k, r, _) ->
-      iter f l; f k.ident k.data; iter f r
+      iter f l; f k.ident (get_data k); iter f r
 
 (* Idents for sharing keys *)
 
@@ -405,6 +541,14 @@
   | Global _, _ -> 1
   | _, Global _ -> (-1)
   | Predef { stamp = s1; _ }, Predef { stamp = s2; _ } -> compare s1 s2
+  | Predef _, _ -> 1
+  | _, Predef _ -> (-1)
+  | Unscoped x, Unscoped y ->
+    let x = Unscoped.get_desc x in
+    let y = Unscoped.get_desc y in
+    let c = x.stamp - y.stamp in
+    if c <> 0 then c
+    else compare x.name y.name
 
 let output oc id = output_string oc (unique_name id)
 let hash i = (Char.code (name i).[0]) lxor (stamp i)
