--- ocaml_504/typing/includemod.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/includemod.mli	2026-02-03 16:48:24
@@ -59,7 +59,6 @@
     | Mt_core of core_module_type_symptom
     | Signature of signature_symptom
     | Functor of functor_symptom
-    | Invalid_module_alias of Path.t
     | After_alias_expansion of module_type_diff
 
 
@@ -84,10 +83,13 @@
   and signature_symptom = {
     env: Env.t;
     subst: Subst.t;
+    sig1: signature;
+    sig2: signature;
     missings: Types.signature_item list;
-    incompatibles: (Ident.t * sigitem_symptom) list;
+    incompatibles: (Types.signature_item * sigitem_symptom) list;
     oks: (int * Typedtree.module_coercion) list;
-    leftovers: ((Types.signature_item as 'it) * 'it * int) list
+    additions: signature_item list;
+    untypables: ((Types.signature_item as 'it) * 'it * int) list;
     (** signature items that could not be compared due to type divergence *)
   }
   and sigitem_symptom =
@@ -137,6 +139,7 @@
 module FieldMap: Map.S with type key = field_desc
 
 val item_ident_name: Types.signature_item -> Ident.t * Location.t * field_desc
+val item_subst: Ident.t -> Types.signature_item -> Subst.t -> Subst.t
 val is_runtime_component: Types.signature_item -> bool
 
 
@@ -181,7 +184,9 @@
 val check_modtype_equiv:
   loc:Location.t -> Env.t -> Ident.t -> module_type -> module_type -> unit
 
-val signatures: Env.t -> mark:bool -> signature -> signature -> module_coercion
+val signatures:
+  Env.t -> ?subst:Subst.t -> mark:bool ->
+  signature -> signature -> module_coercion
 
 (** Check an implementation against an interface *)
 val check_implementation: Env.t -> signature -> signature -> unit
@@ -216,7 +221,6 @@
       Ident.t * class_declaration * class_declaration *
       Ctype.class_match_failure list
   | Unbound_module_path of Path.t
-  | Invalid_module_alias of Path.t
 
 type pos =
   | Module of Ident.t
@@ -241,6 +245,8 @@
 
 val expand_module_alias: strengthen:bool -> Env.t -> Path.t -> Types.module_type
 
+(** Error message functions *)
+
 module Functor_inclusion_diff: sig
   module Defs: sig
     type left = Types.functor_parameter
@@ -270,3 +276,15 @@
     args:(Error.functor_arg_descr * Types.module_type) list ->
     Diffing.Define(Defs).patch
 end
+
+(* Typechecking with subst *)
+module Check: sig
+  type 'a compatibility_test = Env.t -> Subst.t -> 'a -> 'a -> bool
+  val values : Types.value_description compatibility_test
+  val types: Types.type_declaration compatibility_test
+  val class_types : Types.class_type_declaration compatibility_test
+  val classes: Types.class_declaration compatibility_test
+  val modules: Types.module_declaration compatibility_test
+  val module_types: Types.modtype_declaration compatibility_test
+  val extensions: Types.extension_constructor compatibility_test
+end
