--- ocaml_504/typing/oprint.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/oprint.ml	2026-02-03 16:48:24
@@ -309,6 +309,17 @@
       pp_print_space ppf ();
       print_out_type_1 ppf ty2;
       pp_close_box ppf ()
+  | Otyp_functor (lab, id, pack, ty) ->
+      pp_open_box ppf 0;
+      print_arg_label ppf lab;
+      pp_print_string ppf "(module ";
+      print_ident ppf id;
+      pp_print_string ppf " : ";
+      print_package ppf pack;
+      pp_print_string ppf ") ->";
+      pp_print_space  ppf ();
+      print_out_type_1 ppf ty;
+      pp_close_box ppf ()
   | ty -> print_out_type_2 ~arg:false ppf ty
 and print_out_type_2 ~arg ppf =
   function
@@ -339,8 +350,9 @@
       print_typargs ppf tyl;
       print_ident ppf id;
       pp_close_box ppf ()
-  | Otyp_object {fields; open_row} ->
-      fprintf ppf "@[<2>< %a >@]" (print_fields open_row) fields
+  | Otyp_object { fields; row} ->
+      fprintf ppf "@[<2>< %a >@]"
+        (print_object_fields row) fields
   | Otyp_stuff s -> pp_print_string ppf s
   | Otyp_var (non_gen, s) -> ty_var ~non_gen ppf s
   | Otyp_variant (row_fields, closed, tags) ->
@@ -362,13 +374,14 @@
          else if tags = None then "> " else "? ")
         print_fields row_fields
         print_present tags
-  | Otyp_alias _ | Otyp_poly _ | Otyp_arrow _ | Otyp_tuple _ as ty ->
+  | Otyp_alias _ | Otyp_poly _ | Otyp_arrow _
+  | Otyp_functor _ | Otyp_tuple _ as ty ->
       pp_open_box ppf 1;
       pp_print_char ppf '(';
       print_out_type ppf ty;
       pp_print_char ppf ')';
       pp_close_box ppf ()
-  | Otyp_abstract | Otyp_open
+  | Otyp_abstract | Otyp_open | Otyp_external _
   | Otyp_sum _ | Otyp_manifest (_, _) -> ()
   | Otyp_record lbls -> print_record_decl ppf lbls
   | Otyp_module pack ->
@@ -383,20 +396,25 @@
       let sep = if !first then (first := false; "with") else "and" in
       fprintf ppf " %s type %s = %a" sep s print_out_type t
     )
-    pack.opack_cstrs
+    pack.opack_constraints
 and print_record_decl ppf lbls =
   fprintf ppf "{%a@;<1 -2>}"
     (print_list_init print_out_label (fun ppf -> fprintf ppf "@ ")) lbls
-and print_fields open_row ppf =
+and print_object_fields row ppf =
   function
     [] ->
-      if open_row then fprintf ppf "..";
+      begin match row with
+      | Orow_closed -> ()
+      | Orow_open_anonymous -> fprintf ppf ".."
+      | Orow_open ty -> fprintf ppf ".. as %a" print_out_type ty
+      end
   | [s, t] ->
       fprintf ppf "%a : %a" print_lident s print_out_type t;
-      if open_row then fprintf ppf ";@ ";
-      print_fields open_row ppf []
+      if row <> Orow_closed then fprintf ppf ";@ ";
+      print_object_fields row ppf []
   | (s, t) :: l ->
-      fprintf ppf "%s : %a;@ %a" s print_out_type t (print_fields open_row) l
+      fprintf ppf "%s : %a;@ %a"
+        s print_out_type t (print_object_fields row) l
 and print_row_field ppf (l, opt_amp, tyl) =
   let pr_of ppf =
     if opt_amp then fprintf ppf " of@ &@ "
@@ -696,7 +714,7 @@
       (fun (ty1, ty2) ->
          fprintf ppf "@ @[<2>constraint %a =@ %a@]" !out_type ty1
            !out_type ty2)
-      td.otype_cstrs
+      td.otype_constraints
   in
   let type_defined ppf =
     match td.otype_params with
@@ -752,6 +770,8 @@
   | Otyp_open ->
       fprintf ppf " =%a .."
         print_private td.otype_private
+  | Otyp_external name ->
+      fprintf ppf " =@ external %S" name
   | ty ->
       fprintf ppf " =%a@;<1 2>%a"
         print_private td.otype_private
@@ -800,12 +820,12 @@
         [] -> fprintf ppf "%a" print_lident ext.oext_type_name
       | [ty_param] ->
         fprintf ppf "@[%a@ %a@]"
-          (print_type_parameter ~non_gen:false)
+          type_parameter
           ty_param
           print_lident ext.oext_type_name
       | _ ->
         fprintf ppf "@[(@[%a)@]@ %a@]"
-          (print_list print_type_parameter (fun ppf -> fprintf ppf ",@ "))
+          (print_list type_parameter (fun ppf -> fprintf ppf ",@ "))
           ext.oext_type_params
           print_lident ext.oext_type_name
   in
@@ -821,11 +841,11 @@
       [] -> fprintf ppf "%a" print_lident te.otyext_name
     | [param] ->
       fprintf ppf "@[%a@ %a@]"
-        (print_type_parameter ~non_gen:false) param
+        type_parameter param
         print_lident te.otyext_name
     | _ ->
         fprintf ppf "@[(@[%a)@]@ %a@]"
-          (print_list print_type_parameter (fun ppf -> fprintf ppf ",@ "))
+          (print_list type_parameter (fun ppf -> fprintf ppf ",@ "))
           te.otyext_params
           print_lident te.otyext_name
   in
