--- ocaml_504/typing/out_type.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/out_type.mli	2026-02-03 16:48:24
@@ -41,7 +41,7 @@
 
 val wrap_printing_env: error:bool -> Env.t -> (unit -> 'a) -> 'a
 (** Call the function using the environment for type path shortening
-    This affects all the printing and tree cration functions functions below
+    This affects all the printing and tree creation functions below
     Also, if [~error:true], then disable the loading of cmis *)
 
 
