--- ocaml_504/typing/outcometree.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/outcometree.mli	2026-02-03 16:48:24
@@ -72,7 +72,7 @@
   | Otyp_class of out_ident * out_type list
   | Otyp_constr of out_ident * out_type list
   | Otyp_manifest of out_type * out_type
-  | Otyp_object of { fields: (string * out_type) list; open_row:bool}
+  | Otyp_object of { fields: (string * out_type) list; row: out_row}
   | Otyp_record of out_label list
   | Otyp_stuff of string
   | Otyp_sum of out_constructor list
@@ -82,7 +82,14 @@
   | Otyp_poly of string list * out_type
   | Otyp_module of out_package
   | Otyp_attribute of out_type * out_attribute
+  | Otyp_external of string
+  | Otyp_functor of Asttypes.arg_label * out_ident * out_package * out_type
 
+and out_row =
+  | Orow_closed
+  | Orow_open_anonymous
+  | Orow_open of out_type
+
 and out_label = {
   olab_name: string;
   olab_mut: Asttypes.mutable_flag;
@@ -98,7 +105,7 @@
 
 and out_package = {
   opack_path: out_ident;
-  opack_cstrs: (string * out_type) list;
+  opack_constraints: (string * out_type) list;
 }
 
 and out_variant =
@@ -140,17 +147,17 @@
     otype_private: Asttypes.private_flag;
     otype_immediate: Type_immediacy.t;
     otype_unboxed: bool;
-    otype_cstrs: (out_type * out_type) list }
+    otype_constraints: (out_type * out_type) list }
 and out_extension_constructor =
   { oext_name: string;
     oext_type_name: string;
-    oext_type_params: string list;
+    oext_type_params: out_type_param list;
     oext_args: out_type list;
     oext_ret_type: out_type option;
     oext_private: Asttypes.private_flag }
 and out_type_extension =
   { otyext_name: string;
-    otyext_params: string list;
+    otyext_params: out_type_param list;
     otyext_constructors: out_constructor list;
     otyext_private: Asttypes.private_flag }
 and out_val_decl =
