--- ocaml_504/typing/parmatch.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/parmatch.ml	2026-02-03 16:48:24
@@ -863,7 +863,7 @@
             ty env]
       | _ -> [omega]
       end
-  | Typedecl (_, _, {type_kind = Type_abstract _ | Type_open})
+  | Typedecl (_, _, {type_kind = Type_abstract _ | Type_open | Type_external _})
   | May_have_typedecl -> [omega]
 
 let get_variant_constructors env ty =
@@ -1316,17 +1316,18 @@
      best way to make a non-fragile distinction between "all As" and
      "at least one B".
   *)
-  List.to_seq [Some p; None] |> Seq.flat_map
-    (function
-      | Some p ->
-          let sub_witnesses = exhaust ext [ps] (n - 1) in
-          Seq.map (fun row -> p :: row) sub_witnesses
-      | None ->
-          (* note: calling [exhaust] recursively of p would
-             result in an infinite loop in the case n=1 *)
-          let p_witnesses = specialize_and_exhaust ext [[p]] 1 in
-          Seq.map (fun p_row -> p_row @ omegas (n - 1)) p_witnesses
-    )
+  let sub_witnesses =
+    exhaust ext [ps] (n - 1)
+    |> Seq.map (fun row -> p :: row)
+  in
+  let p_witnesses () =
+    let omega_row = omegas (n - 1) in
+    (* note: calling [exhaust] recursively of p would
+       result in an infinite loop in the case n=1 *)
+    specialize_and_exhaust ext [[p]] 1
+    |> Seq.map (fun p_row -> p_row @ omega_row)
+  in
+  Seq.append sub_witnesses (Seq.delay p_witnesses)
 
 and specialize_and_exhaust ext pss n =
   let pss = simplify_first_col pss in
