--- ocaml_504/typing/shape.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/shape.mli	2026-02-03 16:48:24
@@ -58,12 +58,20 @@
   type t = private
     | Compilation_unit of string
     | Item of { comp_unit: string; id: int; from: Unit_info.intf_or_impl }
+    | Local_opaque_item of { comp_unit: string; id: int }
+      (** "Local_opaque_item" is used to give uids to usages of values of which
+          the definition is not known statically. This notably happens when
+          accessing first-class modules' items or usages of a functor
+          parameter's items inside the body of the functor itself. Having Uids
+          synthesized for these usages is useful for tools that provides
+          occurrences and renaming features. *)
     | Internal
     | Predef of string
 
   val reinit : unit -> unit
 
   val mk : current_unit:(Unit_info.t option) -> t
+  val mk_local_opaque : current_unit:(Unit_info.t option) -> t
   val of_compilation_unit_id : Ident.t -> t
   val of_predef_id : Ident.t -> t
   val internal_not_actually_unique : t
@@ -71,6 +79,15 @@
   val for_actual_declaration : t -> bool
 
   include Identifiable.S with type t := t
+
+  (* Dependencies between related Uids are recorded and written in CMT files *)
+  module Deps : sig
+    type kind = Definition_to_declaration | Declaration_to_declaration
+
+    val clear : unit -> unit
+    val get : unit -> (kind * t * t) list
+    val record_declaration_dependency: kind * t * t -> unit
+  end
 end
 
 module Sig_component_kind : sig
@@ -124,6 +141,7 @@
   | Abs of var * t
   | App of t * t
   | Struct of t Item.Map.t
+  | Pack of Ident.t
   | Alias of t
   | Leaf
   | Proj of t * Item.t
@@ -150,7 +168,7 @@
 val decompose_abs : t -> (var * t) option
 
 val for_persistent_unit : string -> t
-val leaf_for_unpack : t
+val leaf_for_unpack : unit -> t
 
 module Map : sig
   type shape = t
