--- ocaml_504/typing/shape_reduce.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/shape_reduce.mli	2026-02-03 16:48:24
@@ -19,9 +19,13 @@
 type result =
   | Resolved of Shape.Uid.t (** Shape reduction succeeded and a uid was found *)
   | Resolved_alias of Shape.Uid.t * result (** Reduction led to an alias *)
+  | Resolved_local_use of Shape.Uid.t
+    (** Shape reduction was stuck on a functor parameter or a first class module
+        and a new uid was generated linked to the item's declaration uid. *)
   | Unresolved of Shape.t (** Result still contains [Comp_unit] terms *)
   | Approximated of Shape.Uid.t option
     (** Reduction failed: it can arrive with first-class modules for example *)
+  | Missing_uid of Shape.t
   | Internal_error_missing_uid
     (** Reduction succeeded but no uid was found, this should never happen *)
 
@@ -59,4 +63,5 @@
 
 (** [local_reduce_for_uid] will not reduce shapes that require loading external
   compilation units. *)
-val local_reduce_for_uid : Env.t -> Shape.t -> result
+val local_reduce_for_uid :
+  Env.t -> namespace:Shape.Sig_component_kind.t -> Path.t -> Shape.t -> result
