--- ocaml_504/typing/tast_mapper.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/tast_mapper.ml	2026-02-03 16:48:24
@@ -222,21 +222,22 @@
   | Ttype_variant list -> Ttype_variant (List.map (constructor_decl sub) list)
   | Ttype_record list -> Ttype_record (List.map (label_decl sub) list)
   | Ttype_open -> Ttype_open
+  | Ttype_external name -> Ttype_external name
 
 let type_declaration sub x =
   let typ_loc = sub.location sub x.typ_loc in
   let typ_name = map_loc sub x.typ_name in
-  let typ_cstrs =
+  let typ_constraints =
     List.map
       (tuple3 (sub.typ sub) (sub.typ sub) (sub.location sub))
-      x.typ_cstrs
+      x.typ_constraints
   in
   let typ_kind = sub.type_kind sub x.typ_kind in
   let typ_manifest = Option.map (sub.typ sub) x.typ_manifest in
   let typ_params = List.map (tuple2 (sub.typ sub) id) x.typ_params in
   let typ_attributes = sub.attributes sub x.typ_attributes in
-  {x with typ_loc; typ_name; typ_cstrs; typ_kind; typ_manifest; typ_params;
-          typ_attributes}
+  {x with typ_loc; typ_name; typ_constraints; typ_kind; typ_manifest;
+          typ_params; typ_attributes}
 
 let type_declarations sub (rec_flag, list) =
   (rec_flag, List.map (sub.type_declaration sub) list)
@@ -278,7 +279,7 @@
   {x with ext_loc; ext_name; ext_kind; ext_attributes}
 
 let pat_extra sub = function
-  | Tpat_unpack as d -> d
+  | Tpat_unpack pty -> Tpat_unpack (Option.map (sub.package_type sub) pty)
   | Tpat_type (path,lid) -> Tpat_type (path, map_loc_lid sub lid)
   | Tpat_open (path,lid,env) ->
       Tpat_open (path, map_loc_lid sub lid, sub.env sub env)
@@ -478,19 +479,6 @@
           path,
           List.map (tuple3 id (map_loc sub) (sub.expr sub)) list
         )
-    | Texp_letmodule (id, s, pres, mexpr, exp) ->
-        Texp_letmodule (
-          id,
-          map_loc sub s,
-          pres,
-          sub.module_expr sub mexpr,
-          sub.expr sub exp
-        )
-    | Texp_letexception (cd, exp) ->
-        Texp_letexception (
-          sub.extension_constructor sub cd,
-          sub.expr sub exp
-        )
     | Texp_assert (exp, loc) ->
         Texp_assert (sub.expr sub exp, loc)
     | Texp_lazy exp ->
@@ -511,8 +499,8 @@
         Texp_unreachable
     | Texp_extension_constructor (lid, path) ->
         Texp_extension_constructor (map_loc_lid sub lid, path)
-    | Texp_open (od, e) ->
-        Texp_open (sub.open_declaration sub od, sub.expr sub e)
+    | Texp_struct_item (si, e) ->
+        Texp_struct_item (sub.structure_item sub si, sub.expr sub e)
   in
   let exp_attributes = sub.attributes sub x.exp_attributes in
   {x with exp_loc; exp_extra; exp_desc; exp_env; exp_attributes}
@@ -520,9 +508,9 @@
 
 let package_type sub x =
   let tpt_txt = map_loc_lid sub x.tpt_txt in
-  let tpt_cstrs = List.map
-    (tuple2 (map_loc_lid sub) (sub.typ sub)) x.tpt_cstrs in
-  {x with tpt_txt; tpt_cstrs}
+  let tpt_constraints = List.map
+    (tuple2 (map_loc_lid sub) (sub.typ sub)) x.tpt_constraints in
+  {x with tpt_txt; tpt_constraints}
 
 let binding_op sub x =
   let bop_loc = sub.location sub x.bop_loc in
@@ -805,6 +793,9 @@
         Ttyp_package (sub.package_type sub pack)
     | Ttyp_open (path, mod_ident, t) ->
         Ttyp_open (path, map_loc_lid sub mod_ident, sub.typ sub t)
+    | Ttyp_functor (label, id, pack, t) ->
+        Ttyp_functor (label, map_loc sub id,
+                      sub.package_type sub pack, sub.typ sub t)
   in
   let ctyp_attributes = sub.attributes sub x.ctyp_attributes in
   {x with ctyp_loc; ctyp_desc; ctyp_env; ctyp_attributes}
