--- ocaml_504/typing/tast_iterator.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/tast_iterator.ml	2026-02-03 16:48:24
@@ -196,6 +196,7 @@
   | Ttype_variant list -> List.iter (constructor_decl sub) list
   | Ttype_record list -> List.iter (label_decl sub) list
   | Ttype_open -> ()
+  | Ttype_external _ -> ()
 
 let type_declaration sub x =
   sub.item_declaration sub (Type x);
@@ -207,7 +208,7 @@
       sub.typ sub c1;
       sub.typ sub c2;
       sub.location sub loc)
-    x.typ_cstrs;
+    x.typ_constraints;
   sub.type_kind sub x.typ_kind;
   Option.iter (sub.typ sub) x.typ_manifest;
   List.iter (fun (c, _) -> sub.typ sub c) x.typ_params
@@ -244,7 +245,7 @@
   sub.attributes sub attrs;
   match e with
   | Tpat_type (_, lid) -> iter_loc_lid sub lid
-  | Tpat_unpack -> ()
+  | Tpat_unpack pty -> Option.iter (sub.package_type sub) pty
   | Tpat_open (_, lid, env) -> iter_loc_lid sub lid; sub.env sub env
   | Tpat_constraint ct -> sub.typ sub ct
 
@@ -379,13 +380,6 @@
       sub.expr sub exp
   | Texp_override (_, list) ->
       List.iter (fun (_, s, e) -> iter_loc sub s; sub.expr sub e) list
-  | Texp_letmodule (_, s, _, mexpr, exp) ->
-      iter_loc sub s;
-      sub.module_expr sub mexpr;
-      sub.expr sub exp
-  | Texp_letexception (cd, exp) ->
-      sub.extension_constructor sub cd;
-      sub.expr sub exp
   | Texp_assert (exp, _) -> sub.expr sub exp
   | Texp_lazy exp -> sub.expr sub exp
   | Texp_object (cl, _) -> sub.class_structure sub cl
@@ -396,13 +390,13 @@
       sub.case sub body
   | Texp_unreachable -> ()
   | Texp_extension_constructor (lid, _) -> iter_loc_lid sub lid
-  | Texp_open (od, e) ->
-      sub.open_declaration sub od;
+  | Texp_struct_item (si, e) ->
+      sub.structure_item sub si;
       sub.expr sub e
 
-
-let package_type sub {tpt_cstrs; tpt_txt; _} =
-  List.iter (fun (lid, p) -> iter_loc_lid sub lid; sub.typ sub p) tpt_cstrs;
+let package_type sub {tpt_constraints; tpt_txt; _} =
+  List.iter
+    (fun (lid, p) -> iter_loc_lid sub lid; sub.typ sub p) tpt_constraints;
   iter_loc_lid sub tpt_txt
 
 let binding_op sub {bop_loc; bop_op_name; bop_exp; _} =
@@ -614,6 +608,10 @@
   | Ttyp_open (_, mod_ident, t) ->
       iter_loc_lid sub mod_ident;
       sub.typ sub t
+  | Ttyp_functor (_, id, pack, ct) ->
+      iter_loc sub id;
+      sub.package_type sub pack;
+      sub.typ sub ct
 
 let class_structure sub {cstr_self; cstr_fields; _} =
   sub.pat sub cstr_self;
