--- ocaml_504/typing/types.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/types.ml	2026-02-03 16:48:24
@@ -45,10 +45,11 @@
   | Tunivar of string option
   | Tpoly of type_expr * type_expr list
   | Tpackage of package
+  | Tfunctor of arg_label * Ident.Unscoped.t * package * type_expr
 
 and package =
     { pack_path : Path.t;
-      pack_cstrs : (string list * type_expr) list }
+      pack_constraints : (string list * type_expr) list }
 
 and row_desc =
     { row_fields: (label * row_field) list;
@@ -89,6 +90,12 @@
   | Cunknown : [> `none] commutable_gen
   | Cvar : {mutable commu: any commutable_gen} -> [> `var] commutable_gen
 
+type tfunctor = {
+  id_us : Ident.Unscoped.t;
+  pack : package;
+  ty : type_expr;
+}
+
 module TransientTypeOps = struct
   type t = type_expr
   let compare t1 t2 = t1.id - t2.id
@@ -170,6 +177,7 @@
      type 'a t = A of (('a -> unit) -> unit) : pos
      type +'a p = ..  : may_pos + inj
      type 'a t = A    : inj
+     type 'a t = external "t" : may_pos + may_neg + inj
  *)
 
 module Variance = struct
@@ -273,10 +281,12 @@
   | Type_record of 'lbl list * record_representation
   | Type_variant of 'cstr list * variant_representation
   | Type_open
+  | Type_external of string
 
 and type_origin =
     Definition
   | Rec_check_regularity
+  | Approx_recmod
   | Existential of string
 
 and record_representation =
@@ -446,7 +456,19 @@
   | Sig_class (id, _, _, _)
   | Sig_class_type (id, _, _, _)
     -> id
+
+  let classify_signature_item =
+    let open Shape.Sig_component_kind in
+    function
+    | Sig_value(id, v, _) -> Value, id, v.val_loc
+    | Sig_type (id, td, _, _) -> Type, id, td.type_loc
+    | Sig_typext (id, te, _, _) -> Extension_constructor, id, te.ext_loc
+    | Sig_module (id, _, md, _, _) -> Module, id, md.md_loc
+    | Sig_modtype (id, mtd, _) -> Module_type, id, mtd.mtd_loc
+    | Sig_class (id, c, _, _) -> Class, id, c.cty_loc
+    | Sig_class_type (id, ct, _, _) -> Class_type, id, ct.clty_loc
 
+
 (**** Definitions for backtracking ****)
 
 type change =
@@ -460,6 +482,7 @@
   | Ckind of [`var] field_kind_gen
   | Ccommu of [`var] commutable_gen
   | Cuniv of type_expr option ref * type_expr option
+  | Cuident of Ident.Unscoped.change
 
 type changes =
     Change of change * changes ref
@@ -472,6 +495,9 @@
   let r' = ref Unchanged in
   !trail := Change (ch, r');
   trail := r'
+
+let () =
+    Ident.Unscoped.change_log := (fun change -> log_change (Cuident change))
 
 (* constructor and accessors for [field_kind] *)
 
@@ -651,6 +677,11 @@
   let row = row_repr_no_fields row in
   {row with row_fields; row_name}
 
+let subst_row_name_path id_map row =
+  match row_name row with
+  | Some (p, tl) -> set_row_name row (Some (Path.subst id_map p, tl))
+  | None -> row
+
 type row_desc_repr =
     Row of { fields: (label * row_field) list;
              more:type_expr;
@@ -759,6 +790,7 @@
   | Ckind  (FKvar r) -> r.field_kind <- FKprivate
   | Ccommu (Cvar r)  -> r.commu <- Cunknown
   | Cuniv  (r, v)    -> r := v
+  | Cuident change    -> Ident.Unscoped.undo_change change
 
 type snapshot = changes ref * int
 let last_snapshot = Local_store.s_ref 0
