--- ocaml_504/typing/typeopt.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/typeopt.ml	2026-02-03 16:48:24
@@ -22,6 +22,11 @@
 open Lambda
 
 let scrape_ty env ty =
+  let ty =
+    match get_desc ty with
+    | Tpoly(ty, _) -> ty
+    | _ -> ty
+  in
   match get_desc ty with
   | Tconstr _ ->
       let ty = Ctype.expand_head_opt env ty in
@@ -106,7 +111,7 @@
       | Some #Predef.data_type_constr | None ->
         try
           match (Env.find_type p env).type_kind with
-          | Type_abstract _ ->
+          | Type_abstract _ | Type_external _ ->
               Any
           | Type_record _ | Type_variant _ | Type_open ->
               Addr
@@ -116,8 +121,8 @@
              Maybe we should emit a warning. *)
           Any
       end
-  | Tarrow _ | Ttuple _ | Tpackage _ | Tobject _ | Tnil | Tvariant _ ->
-      Addr
+  | Tarrow _ | Ttuple _ | Tpackage _ | Tobject _
+  | Tnil | Tvariant _ | Tfunctor _ -> Addr
   | Tlink _ | Tsubst _ | Tpoly _ | Tfield _ ->
       assert false
 
