--- ocaml_504/typing/typeclass.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/typeclass.ml	2026-02-03 16:48:24
@@ -107,6 +107,7 @@
   | No_overriding of string * string
   | Duplicate of string * string
   | Closing_self_type of class_signature
+  | Polymorphic_class_parameter
 
 exception Error of Location.t * Env.t * error
 exception Error_forward of Location.error
@@ -192,6 +193,7 @@
   | Cty_signature _ ->
       constr
   | Cty_arrow (l, ty, cty) ->
+      let ty = Ctype.newmono ty in
       Ctype.newty (Tarrow (l, ty, constructor_type constr cty, commu_ok))
 
                 (***********************************)
@@ -773,12 +775,11 @@
                match get_desc ty with
                | Tvar _ ->
                    let ty' = Ctype.newvar () in
-                   Ctype.unify val_env (Ctype.newty (Tpoly (ty', []))) ty;
-                   Ctype.unify val_env (type_approx val_env sbody) ty'
+                   Ctype.unify val_env (Ctype.newmono ty') ty;
+                   type_approx val_env sbody ty'
                | Tpoly (ty1, tl) ->
-                   let _, ty1' = Ctype.instance_poly ~fixed:false tl ty1 in
-                   let ty2 = type_approx val_env sbody in
-                   Ctype.unify val_env ty2 ty1'
+                   let ty1' = Ctype.instance_poly tl ty1 in
+                   type_approx val_env sbody ty1'
                | _ -> assert false
              with Ctype.Unify err ->
                raise(Error(loc, val_env,
@@ -907,10 +908,10 @@
       Warnings.with_state warning_state
         (fun () ->
            let ty = Btype.method_type label.txt sign in
-           let self_type = sign.Types.csig_self in
+           let self_param_type = Btype.newgenmono sign.Types.csig_self in
            let meth_type =
              mk_expected
-               (Btype.newgenty (Tarrow(Nolabel, self_type, ty, commu_ok)))
+               (Btype.newgenty (Tarrow(Nolabel, self_param_type, ty, commu_ok)))
            in
            let texp =
              Ctype.with_raised_nongen_level
@@ -925,10 +926,10 @@
       Warnings.with_state warning_state
         (fun () ->
            let unit_type = Ctype.instance Predef.type_unit in
-           let self_type = sign.Types.csig_self in
+           let self_param_type = Ctype.newmono sign.Types.csig_self in
            let meth_type =
-             mk_expected
-               (Ctype.newty (Tarrow (Nolabel, self_type, unit_type, commu_ok)))
+             mk_expected (Ctype.newty
+              (Tarrow (Nolabel, self_param_type, unit_type, commu_ok)))
            in
            let texp =
              Ctype.with_raised_nongen_level
@@ -1117,6 +1118,8 @@
           cl_attributes = scl.pcl_attributes;
          }
   | Pcl_fun (l, Some default, spat, sbody) ->
+      if has_poly_constraint spat then
+        raise(Error(spat.ppat_loc, val_env, Polymorphic_class_parameter));
       let loc = default.pexp_loc in
       let open Ast_helper in
       let scases = [
@@ -1150,6 +1153,8 @@
       in
       class_expr cl_num val_env met_env virt self_scope sfun
   | Pcl_fun (l, None, spat, scl') ->
+      if has_poly_constraint spat then
+        raise(Error(spat.ppat_loc, val_env, Polymorphic_class_parameter));
       let (pat, pv, val_env', met_env) =
         Ctype.with_local_level_generalize_structure_if_principal
           (fun () ->
@@ -1424,6 +1429,7 @@
       let arg =
         if Btype.is_optional l then Ctype.instance var_option
         else Ctype.newvar () in
+      let arg = Ctype.newmono arg in
       Ctype.newty (Tarrow (l, arg, approx_declaration cl, commu_ok))
   | Pcl_let (_, _, cl) ->
       approx_declaration cl
@@ -1437,6 +1443,7 @@
       let arg =
         if Btype.is_optional l then Ctype.instance var_option
         else Ctype.newvar () in
+      let arg = Ctype.newmono arg in
       Ctype.newty (Tarrow (l, arg, approx_description ct, commu_ok))
   | _ -> Ctype.newvar ()
 
@@ -2104,24 +2111,24 @@
       Errortrace_report.unification ppf env err
         (msg  "The type parameter")
         (msg "does not meet its constraint: it should be")
-  | Bad_parameters (id, params, cstrs) ->
-      Out_type.prepare_for_printing (params @ cstrs);
+  | Bad_parameters (id, params, constraints) ->
+      Out_type.prepare_for_printing (params @ constraints);
       fprintf ppf
         "@[The abbreviation %a@ is used with parameter(s)@ %a@ \
            which are incompatible with constraint(s)@ %a@]"
         (Style.as_inline_code Printtyp.ident) id
         pp_args params
-        pp_args cstrs
-  | Bad_class_type_parameters (id, params, cstrs) ->
+        pp_args constraints
+  | Bad_class_type_parameters (id, params, constraints) ->
       let pp_hash ppf id = fprintf ppf "#%a" Printtyp.ident id in
-      Out_type.prepare_for_printing (params @ cstrs);
+      Out_type.prepare_for_printing (params @ constraints);
       fprintf ppf
         "@[The class type %a@ is used with parameter(s)@ %a,@ \
            whereas the class type definition@ constrains@ \
            those parameters to be@ %a@]"
         (Style.as_inline_code pp_hash) id
        pp_args params
-       pp_args cstrs
+       pp_args constraints
   | Class_match_failure error ->
       Includeclass.report_error_doc Type ppf error
   | Unbound_val lab ->
@@ -2204,6 +2211,9 @@
        it has been unified with the self type of a class that is not yet@ \
        completely defined.@]"
       (Style.as_inline_code Printtyp.type_scheme) sign.csig_self
+  | Polymorphic_class_parameter ->
+      fprintf ppf
+        "Class parameters cannot be polymorphic."
 
 let report_error_doc env ppf err =
   Printtyp.wrap_printing_env ~error:true
