--- ocaml_504/typing/typedtree.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/typedtree.mli	2026-02-03 16:48:24
@@ -67,9 +67,9 @@
                            branches of [tconst].
          *)
   | Tpat_open of Path.t * Longident.t loc * Env.t
-  | Tpat_unpack
-        (** (module P)     { pat_desc  = Tpat_var "P"
-                           ; pat_extra = (Tpat_unpack, _, _) :: ... }
+  | Tpat_unpack of package_type option
+        (** (module P : ?S)     { pat_desc  = Tpat_var "P"
+                           ; pat_extra = (Tpat_unpack ?S, _, _) :: ... }
             (module _)     { pat_desc  = Tpat_any
             ; pat_extra = (Tpat_unpack, _, _) :: ... }
          *)
@@ -284,10 +284,6 @@
   | Texp_instvar of Path.t * Path.t * string loc
   | Texp_setinstvar of Path.t * Path.t * string loc * expression
   | Texp_override of Path.t * (Ident.t * string loc * expression) list
-  | Texp_letmodule of
-      Ident.t option * string option loc * Types.module_presence * module_expr *
-        expression
-  | Texp_letexception of extension_constructor * expression
   | Texp_assert of expression * Location.t
   | Texp_lazy of expression
   | Texp_object of class_structure * string list
@@ -301,8 +297,7 @@
     }
   | Texp_unreachable
   | Texp_extension_constructor of Longident.t loc * Path.t
-  | Texp_open of open_declaration * expression
-        (** let open[!] M in e *)
+  | Texp_struct_item of structure_item * expression
 
 and meth =
     Tmeth_name of string
@@ -687,11 +682,12 @@
   | Ttyp_poly of string list * core_type
   | Ttyp_package of package_type
   | Ttyp_open of Path.t * Longident.t loc * core_type
+  | Ttyp_functor of arg_label * Ident.t loc * package_type * core_type
 
 and package_type = {
   tpt_path : Path.t;
-  tpt_cstrs : (Longident.t loc * core_type) list;
-  tpt_type : Types.module_type;
+  tpt_constraints : (Longident.t loc * core_type) list;
+  tpt_type : Types.package;
   tpt_txt : Longident.t loc;
 }
 
@@ -731,7 +727,7 @@
     typ_name: string loc;
     typ_params: (core_type * (variance * injectivity)) list;
     typ_type: Types.type_declaration;
-    typ_cstrs: (core_type * core_type * Location.t) list;
+    typ_constraints: (core_type * core_type * Location.t) list;
     typ_kind: type_kind;
     typ_private: private_flag;
     typ_manifest: core_type option;
@@ -744,6 +740,7 @@
   | Ttype_variant of constructor_declaration list
   | Ttype_record of label_declaration list
   | Ttype_open
+  | Ttype_external of string
 
 and label_declaration =
     {
@@ -946,3 +943,7 @@
 
 val map_apply_arg:
   ('a -> ' b) -> ('a, 'omitted) arg_or_omitted ->  ('b, 'omitted) arg_or_omitted
+
+val path_of_module : module_expr -> Path.t option
+
+val remove_module_constraint : module_expr -> module_expr
