--- ocaml_504/typing/typetexp.mli	2025-11-07 15:50:07
+++ ocaml_505/typing/typetexp.mli	2026-02-03 16:48:24
@@ -93,6 +93,8 @@
   | Opened_object of Path.t option
   | Not_an_object of type_expr
   | Repeated_tuple_label of string
+  | Polymorphic_optional_param of string
+  | Functor_optional_param of string
 
 exception Error of Location.t * Env.t * error
 
@@ -101,7 +103,15 @@
     (Location.t -> Env.t -> Longident.t -> Path.t) ref
 val transl_modtype: (* from Typemod *)
     (Env.t -> Parsetree.module_type -> Typedtree.module_type) ref
-val check_package_with_type_constraints: (* from Typemod *)
-    (Location.t -> Env.t -> Types.module_type ->
-     (Longident.t Asttypes.loc * Typedtree.core_type) list ->
-     Types.module_type) ref
+
+type 'a maybe_compute_mty =
+  | ComputeMType : Types.module_type maybe_compute_mty
+  | NoMType : unit maybe_compute_mty
+type forward_decl = {
+  mutable check_package_with_type_constraints : (* from Typemod *)
+    'a. Location.t -> Env.t -> Types.module_type ->
+        'a maybe_compute_mty ->
+        (Longident.t Asttypes.loc * Typedtree.core_type) list -> 'a;
+}
+
+val forward_decl : forward_decl
