--- ocaml_504/typing/untypeast.ml	2025-11-07 15:50:07
+++ ocaml_505/typing/untypeast.ml	2026-02-03 16:48:24
@@ -216,11 +216,11 @@
   let attrs = sub.attributes sub decl.typ_attributes in
   Type.mk ~loc ~attrs
     ~params:(List.map (type_parameter sub) decl.typ_params)
-    ~cstrs:(
+    ~constraints:(
       List.map
         (fun (ct1, ct2, loc) ->
            (sub.typ sub ct1, sub.typ sub ct2, sub.location sub loc))
-        decl.typ_cstrs)
+        decl.typ_constraints)
     ~kind:(sub.type_kind sub decl.typ_kind)
     ~priv:decl.typ_private
     ?manifest:(Option.map (sub.typ sub) decl.typ_manifest)
@@ -233,6 +233,7 @@
   | Ttype_record list ->
       Ptype_record (List.map (sub.label_declaration sub) list)
   | Ttype_open -> Ptype_open
+  | Ttype_external name -> Ptype_external name
 
 let constructor_arguments sub = function
    | Cstr_tuple l -> Pcstr_tuple (List.map (sub.typ sub) l)
@@ -286,16 +287,19 @@
   let attrs = sub.attributes sub pat.pat_attributes in
   let desc =
   match pat with
-      { pat_extra=[Tpat_unpack, loc, _attrs]; pat_desc = Tpat_any; _ } ->
-        Ppat_unpack { txt = None; loc  }
-    | { pat_extra=[Tpat_unpack, _, _attrs];
+      { pat_extra=[Tpat_unpack pty, loc, _attrs]; pat_desc = Tpat_any; _ } ->
+        Ppat_unpack({ txt = None; loc  }, Option.map (sub.package_type sub) pty)
+    | { pat_extra=[Tpat_unpack pty, _, _attrs];
         pat_desc = Tpat_var (_,name, _); _ } ->
-        Ppat_unpack { name with txt = Some name.txt }
+        Ppat_unpack ({ name with txt = Some name.txt },
+          Option.map (sub.package_type sub) pty)
     | { pat_extra=[Tpat_type (_path, lid), _, _attrs]; _ } ->
         Ppat_type (map_loc sub lid)
     | { pat_extra= (Tpat_constraint ct, _, _attrs) :: rem; _ } ->
         Ppat_constraint (sub.pat sub { pat with pat_extra=rem },
                          sub.typ sub ct)
+    | { pat_extra = (Tpat_open (_path, lid, _env), _, _attrs) :: rem; _ } ->
+        Ppat_open (lid, sub.pat sub { pat with pat_extra=rem })
     | _ ->
     match pat.pat_desc with
       Tpat_any -> Ppat_any
@@ -303,18 +307,10 @@
         begin
           match (Ident.name id).[0] with
             'A'..'Z' ->
-              Ppat_unpack { name with txt = Some name.txt}
+              Ppat_unpack ({ name with txt = Some name.txt}, None)
           | _ ->
               Ppat_var name
         end
-
-    (* We transform (_ as x) in x if _ and x have the same location.
-       The compiler transforms (x:t) into (_ as x : t).
-       This avoids transforming a warning 27 into a 26.
-     *)
-    | Tpat_alias ({pat_desc = Tpat_any; pat_loc}, _id, name, _, _ty)
-         when pat_loc = pat.pat_loc ->
-       Ppat_var name
 
     | Tpat_alias (pat, _id, name, _, _ty) ->
         Ppat_alias (sub.pat sub pat, name)
@@ -548,12 +544,6 @@
         Pexp_override (List.map (fun (_path, lid, exp) ->
               (map_loc sub lid, sub.expr sub exp)
           ) list)
-    | Texp_letmodule (_id, name, _pres, mexpr, exp) ->
-        Pexp_letmodule (name, sub.module_expr sub mexpr,
-          sub.expr sub exp)
-    | Texp_letexception (ext, exp) ->
-        Pexp_letexception (sub.extension_constructor sub ext,
-                           sub.expr sub exp)
     | Texp_assert (exp, _) -> Pexp_assert (sub.expr sub exp)
     | Texp_lazy exp -> Pexp_lazy (sub.expr sub exp)
     | Texp_object (cl, _) ->
@@ -575,8 +565,8 @@
                         PStr [ Str.eval ~loc
                                  (Exp.construct ~loc (map_loc sub lid) None)
                              ])
-    | Texp_open (od, exp) ->
-        Pexp_open (sub.open_declaration sub od, sub.expr sub exp)
+    | Texp_struct_item (si, exp) ->
+        Pexp_struct_item (sub.structure_item sub si, sub.expr sub exp)
   in
   List.fold_right (exp_extra sub) exp.exp_extra
     (Exp.mk ~loc ~attrs desc)
@@ -590,7 +580,8 @@
 
 let package_type sub pack =
   { ppt_path = map_loc sub pack.tpt_txt;
-    ppt_cstrs = List.map (fun (s, ct) -> (s, sub.typ sub ct)) pack.tpt_cstrs;
+    ppt_constraints =
+      List.map (fun (s, ct) -> (s, sub.typ sub ct)) pack.tpt_constraints;
     ppt_attrs = [];
     ppt_loc = sub.location sub pack.tpt_txt.loc }
 
@@ -844,6 +835,9 @@
         Ptyp_poly (list, sub.typ sub ct)
     | Ttyp_package pack -> Ptyp_package (sub.package_type sub pack)
     | Ttyp_open (_path, mod_ident, t) -> Ptyp_open (mod_ident, sub.typ sub t)
+    | Ttyp_functor (label, name, pack, ct) ->
+        let name = Location.mkloc (Ident.name name.txt) name.loc in
+        Ptyp_functor (label, name, sub.package_type sub pack, sub.typ sub ct)
   in
   Typ.mk ~loc ~attrs desc
 
