--- ocaml_504/utils/build_path_prefix_map.ml	2025-11-07 15:50:07
+++ ocaml_505/utils/build_path_prefix_map.ml	2026-02-03 16:49:03
@@ -116,3 +116,14 @@
   match rewrite_first prefix_map path with
   | None -> path
   | Some path -> path
+
+let make_source path : pair option -> path option = function
+  | None -> None
+  | Some { target; source } ->
+    if String.starts_with ~prefix:target path then
+      Some (source ^ (String.sub path (String.length target)
+                        (String.length path - String.length target)))
+    else None
+
+let invert_all prefix_map path =
+  List.filter_map (make_source path) (List.rev prefix_map)
