--- ocaml_504/utils/clflags.ml	2025-11-07 15:50:07
+++ ocaml_505/utils/clflags.ml	2026-02-03 16:49:03
@@ -38,9 +38,11 @@
   end
 end)
 
-let objfiles = ref ([] : string list)   (* .cmo and .cma files *)
-and ccobjs = ref ([] : string list)     (* .o, .a, .so and -cclib -lxxx *)
-and dllibs = ref ([] : string list)     (* .so and -dllib -lxxx *)
+let objfiles = ref ([] : string list)         (* .cmo and .cma files *)
+and ccobjs = ref ([] : string list)           (* .o, .a, .so and -cclib -lxxx *)
+and dllibs = ref ([] : (suffixed:bool * string) list)
+                                              (* .so, -dllib -lxxx and
+                                                 -dllib-suffixed -lxxx *)
 
 let cmi_file = ref None
 
@@ -48,6 +50,7 @@
 and output_name = ref (None : string option) (* -o *)
 and include_dirs = ref ([] : string list) (* -I *)
 and hidden_include_dirs = ref ([] : string list) (* -H *)
+and standard_library_default = ref None (* -set-runtime-default *)
 and no_std_include = ref false          (* -nostdlib *)
 and no_cwd = ref false                  (* -nocwd *)
 and print_types = ref false             (* -i *)
@@ -86,6 +89,12 @@
 and open_modules = ref []               (* -open *)
 and use_prims = ref ""                  (* -use-prims ... *)
 and use_runtime = ref ""                (* -use-runtime ... *)
+and target_bindir =                     (* -launch-method ... *)
+  ref Config.target_bindir
+and launch_method =
+  ref Config.launch_method
+and search_method =                     (* -search-method ... *)
+  ref Config.search_method
 and plugin = ref false                  (* -plugin ... *)
 and principal = ref false               (* -principal *)
 and real_paths = ref true               (* -short-paths *)
