--- ocaml_504/utils/config.common.ml.in	2025-11-07 15:50:07
+++ ocaml_505/utils/config.common.ml.in	2026-02-03 16:49:03
@@ -20,6 +20,28 @@
 (* The main OCaml version string has moved to ../build-aux/ocaml_version.m4 *)
 let version = Sys.ocaml_version
 
+(* is_official_release and release_number are automatically updated autoconf
+   from values in ../build-aux/ocaml_version.m4 - do not edit these lines
+   directly. *)
+let is_official_release = false
+let release_number = 21
+
+external standard_library_default : unit -> string = "%standard_library_default"
+
+let standard_library_default_raw = standard_library_default ()
+
+external stdlib_dirs : string -> string * string option
+   = "caml_sys_get_stdlib_dirs"
+
+let standard_library_default, relative_root_dir =
+  stdlib_dirs standard_library_default_raw
+
+let standard_library_relative =
+  if relative_root_dir = None then
+    None
+  else
+    Some standard_library_default_raw
+
 let standard_library =
   try
     Sys.getenv "OCAMLLIB"
@@ -29,6 +51,13 @@
   with Not_found ->
     standard_library_default
 
+let bindir = Option.value ~default:bindir relative_root_dir
+let target_bindir =
+  if target_bindir = Filename.current_dir_name then
+    Filename.dirname Sys.executable_name
+  else
+    target_bindir
+
 let exec_magic_number = {magic|@EXEC_MAGIC_NUMBER@|magic}
     (* exec_magic_number is duplicated in runtime/caml/exec.h *)
 and cmi_magic_number = {magic|@CMI_MAGIC_NUMBER@|magic}
@@ -46,6 +75,21 @@
 let default_safe_string = true
 let naked_pointers = false
 
+type launch_method = Executable | Shebang of string option
+type search_method = Disable | Fallback | Enable
+
+let launch_method =
+  match launch_method with
+  | "exe" -> Executable
+  | "sh" -> Shebang None
+  | _ -> Shebang (Some launch_method)
+
+let search_method =
+  match search_method with
+  | "enable" -> Enable
+  | "fallback" -> Fallback
+  | _ -> Disable
+
 let interface_suffix = ref ".mli"
 
 let max_tag = 243
@@ -57,11 +101,16 @@
 let max_young_wosize = 256
 let stack_threshold = 32 (* see runtime/caml/config.h *)
 let stack_safety_margin = 6
+let target_unix = (target_os_type = "Unix")
+let target_win32 = (target_os_type = "Win32")
+let target_cygwin = (target_os_type = "Cygwin")
 let default_executable_name =
-  match target_os_type with
-    "Unix" -> "a.out"
-  | "Win32" | "Cygwin" -> "camlprog.exe"
-  | _ -> "camlprog"
+  if target_unix then
+    "a.out"
+  else if target_win32 || target_cygwin then
+    "camlprog.exe"
+  else
+    "camlprog"
 type configuration_value =
   | String of string
   | Int of int
@@ -71,9 +120,13 @@
   let p x v = (x, String v) in
   let p_int x v = (x, Int v) in
   let p_bool x v = (x, Bool v) in
+  let standard_library_relative =
+    Option.value ~default:"" standard_library_relative
+  in
 [
   p "version" version;
   p "standard_library_default" standard_library_default;
+  p "standard_library_relative" standard_library_relative;
   p "standard_library" standard_library;
   p "ccomp_type" ccomp_type;
   p "c_compiler" c_compiler;
@@ -113,6 +166,8 @@
   p_bool "systhread_supported" systhread_supported;
   p "host" host;
   p "target" target;
+  p "bytecode_runtime_id" bytecode_runtime_id;
+  p "native_runtime_id" native_runtime_id;
   p_bool "flambda" flambda;
   p_bool "safe_string" safe_string;
   p_bool "default_safe_string" default_safe_string;
