--- ocaml_504/utils/config.fixed.ml	2025-11-07 15:50:07
+++ ocaml_505/utils/config.fixed.ml	2026-02-03 16:49:03
@@ -21,12 +21,14 @@
 let boot_cannot_call s = "/ The boot compiler should not call " ^ s
 
 let bindir = "/tmp"
-let standard_library_default = "/tmp"
+let target_bindir = bindir
 let ccomp_type = "n/a"
 let c_compiler = boot_cannot_call "the C compiler"
+let c_compiler_vendor = ""
 let c_output_obj = ""
 let c_has_debug_prefix_map = false
 let as_has_debug_prefix_map = false
+let as_is_cc = false
 let bytecode_cflags = ""
 let bytecode_cppflags = ""
 let native_cflags = ""
@@ -57,6 +59,8 @@
 let align_int64 = true
 let function_sections = false
 let afl_instrument = false
+let bytecode_runtime_id = ""
+let native_runtime_id = ""
 let native_compiler = false
 let tsan = false
 let architecture = "none"
@@ -67,6 +71,7 @@
 let asm = boot_cannot_call "the assembler"
 let asm_cfi_supported = false
 let asm_size_type_directives = false
+let asm_dwarf_version = None
 let with_frame_pointers = false
 let reserved_header_bits = 0
 let ext_exe = ".ex_The boot compiler should not be using Config.ext_exe"
@@ -79,3 +84,7 @@
 let systhread_supported = false
 let flexdll_dirs = []
 let ar_supports_response_files = true
+let shebangscripts = false
+let suffixing = false
+let launch_method = "sh"
+let search_method = "always"
