--- ocaml_502/parsing/ast_invariants.ml	2024-06-27 15:42:08.724127244 +0200
+++ ocaml_503/parsing/ast_invariants.ml	2024-09-17 01:16:30.152541842 +0200
@@ -23,6 +23,8 @@
 let no_args loc = err loc "Function application with no argument."
 let empty_let loc = err loc "Let with no bindings."
 let empty_type loc = err loc "Type declarations cannot be empty."
+let empty_poly_binder loc =
+  err loc "Explicit universal type quantification cannot be empty."
 let complex_id loc = err loc "Functor application not allowed here."
 let module_type_substitution_missing_rhs loc =
   err loc "Module type substitution with no right hand side"
@@ -53,6 +55,7 @@
     | Ptyp_tuple ([] | [_]) -> invalid_tuple loc
     | Ptyp_package (_, cstrs) ->
       List.iter (fun (id, _) -> simple_longident id) cstrs
+    | Ptyp_poly([],_) -> empty_poly_binder loc
     | _ -> ()
   in
   let pat self pat =
