--- ocaml_502/parsing/ast_iterator.ml	2024-06-27 15:42:08.724127244 +0200
+++ ocaml_503/parsing/ast_iterator.ml	2024-09-17 01:16:30.152541842 +0200
@@ -493,6 +493,7 @@
     | Ppat_type s -> iter_loc sub s
     | Ppat_lazy p -> sub.pat sub p
     | Ppat_unpack s -> iter_loc sub s
+    | Ppat_effect (p1,p2) -> sub.pat sub p1; sub.pat sub p2
     | Ppat_exception p -> sub.pat sub p
     | Ppat_extension x -> sub.extension sub x
     | Ppat_open (lid, p) ->
