--- ocaml_502/parsing/pprintast.mli	2024-06-27 15:42:08.724127244 +0200
+++ ocaml_503/parsing/pprintast.mli	2024-09-17 01:16:30.152541842 +0200
@@ -59,3 +59,13 @@
   (** Print a type variable name as a valid identifier, taking care of the
       special treatment required for the single quote character in second
       position, or for keywords by escaping them with \#. No-op on "_". *)
+
+(** {!Format_doc} functions for error messages *)
+module Doc:sig
+  val longident: Longident.t Format_doc.printer
+  val tyvar: string Format_doc.printer
+
+  (** Returns a format document if the expression reads nicely as the subject
+      of a sentence in a error message. *)
+  val nominal_exp : Parsetree.expression -> Format_doc.t option
+end
